{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0",
   "metadata": {},
   "source": [
    "# Collect CG characteristics! !\n",
    "\n",
    "## Frac diffs, Lasso ML values, ..\n",
    "\n",
    "Highlight CpG regions (promoter, gene body)?\n",
    "\n",
    "Add actual genomic positions if you provide them?\n",
    "\n",
    "Generate a heatmap or volcano plot?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "\n",
    "def current_time():\n",
    "    \"\"\"Returns the current date and time as a formatted string.\"\"\"\n",
    "    return datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\") \n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    print(\"Current Date and Time:\", current_time())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare mC fractures across conditions\n",
    "import sys\n",
    "from pathlib import Path\n",
    "from datetime import datetime\n",
    "from matplotlib import pyplot as plt \n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "# import time\n",
    "# import shap\n",
    "\n",
    "\n",
    "# Add the folder containing 'my_functions.py' to sys.path\n",
    "sys.path.append(\"/home/michalula/code/epiCausality/epiCode/utils/\") # str(Path(__file__).parent / 'utils'))\n",
    "\n",
    "# Import the module or specific functions\n",
    "# from /home/michalula/code/epiCausality/epiCode/utils/funcs_extract_mC_profiles_from_BAMs.py\n",
    "# from funcs_extract_mC_profiles_from_BAMs import system_info, extract_from_bam\n",
    "from funcs_extract_mC_profiles_from_BAMs import (\n",
    "    system_info,\n",
    "    get_reference_sequence,\n",
    "    create_output_directory,\n",
    "    extract_from_bam,\n",
    "    process_extracted_reads,\n",
    "    visualize_data,\n",
    "    create_padded_reads,\n",
    "    plot_padded_reads,\n",
    "    save_padded_reads,\n",
    "    remove_low_methylated_reads,\n",
    "    process_extracted_reads_no_fully_unmethylated,\n",
    "    create_padded_reads_no_fully_unmethylated,\n",
    "    current_time\n",
    "    # main,\n",
    ")\n",
    "from funcs_analize_forward_reverse_extracted_mC_reads import analize_forward_reverse_CGs_pipeline, load_padded_reads\n",
    "\n",
    "def get_fractions(CGs_pd):\n",
    "    CGs_sums = np.nansum(CGs_pd, axis=0) \n",
    "    CGs_fracs = CGs_sums  / len(CGs_pd) \n",
    "    return CGs_fracs\n",
    "\n",
    "# Day 6 post ElectroPoration (EP)\n",
    "## Unedited D6\n",
    "save_folder_path = \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/unedited/analyze_single_reads/dimelo_v2_output/\"\n",
    "save_padded_reads_name_np = save_folder_path + \"CG_137_padded_reads_day6_unedited_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD490_numRVS644.npy\"\n",
    "\n",
    "CGs_D6_unedited_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "\n",
    "CGs_D6_unedited_mc07_fracs = get_fractions(CGs_D6_unedited_mc07)\n",
    "print(\"CGs_D6_unedited_mc07 shape: \", CGs_D6_unedited_mc07.shape)\n",
    "print(\"CGs_D6_unedited_mc07_fracs shape: \", CGs_D6_unedited_mc07_fracs.shape)\n",
    "save_folder_path = \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/unedited/analyze_single_reads/dimelo_v2_output/\"\n",
    "save_padded_reads_name_np = save_folder_path + \"CG_137_padded_reads_day6_unedited_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD489_numRVS638.npy\"\n",
    "CGs_D6_unedited_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "\n",
    "CGs_D6_unedited_mc0995_fracs = get_fractions(CGs_D6_unedited_mc0995)\n",
    "print(\"CGs_D6_unedited_mc0995 shape: \", CGs_D6_unedited_mc0995.shape)\n",
    "print(\"CGs_D6_unedited_mc0995_fracs shape: \", CGs_D6_unedited_mc0995_fracs.shape)   \n",
    "\n",
    "\n",
    "## CRISPRoff Day 6 \n",
    "save_padded_reads_name_np =  \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/croff/analyze_single_reads/dimelo_v2_output/CG_137_padded_reads_day6_CRoff_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD802_numRVS1480.npy\"\n",
    "CGs_D6_CRoff_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "\n",
    "CGs_D6_CRoff_mc07_fracs = get_fractions(CGs_D6_CRoff_mc07)\n",
    "print(\"CGs_D6_CRoff_mc07 shape: \", CGs_D6_CRoff_mc07.shape)\n",
    "print(\"CGs_D6_CRoff_mc07_fracs shape: \", CGs_D6_CRoff_mc07_fracs.shape)\n",
    "\n",
    "\n",
    "save_padded_reads_name_np = \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/croff/analyze_single_reads/dimelo_v2_output/\" +\\\n",
    "     'CG_137_padded_reads_day6_CRoff_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD793_numRVS1449.npy'\n",
    "# \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/croff/analyze_single_reads/dimelo_v2_output/CG_137_padded_reads_day6_CRoff_Tcells_Thr0.995_NoFullyUnmethylated_T2Tv2_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-09-13_units_combined_numFWD652_numRVS1125.npy\"\n",
    "CGs_D6_CRoff_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "\n",
    "CGs_D6_CRoff_mc0995_fracs = get_fractions(CGs_D6_CRoff_mc0995)\n",
    "print(\"CGs_D6_CRoff_mc0995 shape: \", CGs_D6_CRoff_mc0995.shape)\n",
    "print(\"CGs_D6_CRoff_mc0995_fracs shape: \", CGs_D6_CRoff_mc0995_fracs.shape)     \n",
    "\n",
    "# Day 28: \n",
    "# Unedited Cells \n",
    "\n",
    "# Day 28 post EP\n",
    "## Unedited Cells \n",
    "folder_path = \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_28/unedited/analyze_single_reads/dimelo_v2_output/\"\n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day28_unedited_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD6_numRVS17.npy\"\n",
    "CGs_D28_unedited_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "\n",
    "CGs_D28_unedited_mc07_fracs = get_fractions(CGs_D28_unedited_mc07)\n",
    "print(\"CGs_D28_unedited_mc07 shape: \", CGs_D28_unedited_mc07.shape)\n",
    "print(\"CGs_D28_unedited_mc07_fracs shape: \", CGs_D28_unedited_mc07_fracs.shape)  \n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day28_unedited_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD6_numRVS18.npy\"\n",
    "CGs_D28_unedited_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "\n",
    "CGs_D28_unedited_mc0995_fracs = get_fractions(CGs_D28_unedited_mc0995)\n",
    "print(\"CGs_D28_unedited_mc0995 shape: \", CGs_D28_unedited_mc0995.shape)\n",
    "print(\"CGs_D28_unedited_mc0995_fracs shape: \", CGs_D28_unedited_mc0995_fracs.shape) \n",
    "# CRoff day 28\n",
    "folder_path = \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_28/croff/analyze_single_reads/dimelo_v2_output/\"\n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day28_CRoff_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD6_numRVS6.npy\"\n",
    "CGs_D28_CRoff_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "\n",
    "CGs_D28_CRoff_mc07_fracs = get_fractions(CGs_D28_CRoff_mc07)\n",
    "print('CGs_D28_CRoff_mc07 shape: ', CGs_D28_CRoff_mc07.shape) \n",
    "print(\"CGs_D28_CRoff_mc07_fracs shape: \", CGs_D28_CRoff_mc07_fracs.shape)\n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day28_CRoff_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD6_numRVS6.npy\"\n",
    "CGs_D28_CRoff_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "CGs_D28_CRoff_mc0995_fracs = get_fractions(CGs_D28_CRoff_mc0995)\n",
    "print(\"CGs_D28_CRoff_mc0995 shape: \", CGs_D28_CRoff_mc0995.shape)\n",
    "print(\"CGs_D28_CRoff_mc0995_fracs shape: \", CGs_D28_CRoff_mc0995_fracs.shape)   \n",
    "# Day 35: CROff\n",
    "# Day 35\n",
    "folder_path = \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_35/croff/analyze_single_reads/dimelo_v2_output/\"\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day35_CRoff_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD104_numRVS222.npy\"\n",
    "CGs_D35_CRoff_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "CGs_D35_CRoff_mc07_fracs = get_fractions(CGs_D35_CRoff_mc07)\n",
    "print(\"CGs_D35_CRoff_mc07 shape\", CGs_D35_CRoff_mc07.shape)\n",
    "print(\"CGs_D35_CRoff_mc07_fracs shape: \", CGs_D35_CRoff_mc07_fracs.shape)   \n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day35_CRoff_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD104_numRVS222.npy\"\n",
    "CGs_D35_CRoff_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "CGs_D35_CRoff_mc0995_fracs = get_fractions(CGs_D35_CRoff_mc0995)\n",
    "print(\"CGs_D35_CRoff_mc0995 shape:\", CGs_D35_CRoff_mc0995.shape)\n",
    "print(\"CGs_D35_CRoff_mc0995_fracs shape: \", CGs_D35_CRoff_mc0995_fracs.shape)   \n",
    "# Day 35: unedited; 37hours sequenced\n",
    "folder_path= \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_35/unedited/part1_37h_sequenced/dimelo_v2_output/\"\n",
    "\n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day35_CRoff_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-10-05_units_combined_numFWD86_numRVS115.npy\"\n",
    "CGs_D35_unedited_mc07_lib1 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "CGs_D35_unedited_mc07_frac_lib1 = get_fractions(CGs_D35_unedited_mc07_lib1)\n",
    "print(\"CGs_D35_unedited_mc07_lib1 shape:\", CGs_D35_unedited_mc07_lib1.shape)\n",
    "print(\"CGs_D35_unedited_mc07_fracs_lib1 shape: \", CGs_D35_unedited_mc07_frac_lib1.shape)   \n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day35_CRoff_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-10-05_units_combined_numFWD86_numRVS114.npy\"\n",
    "CGs_D35_unedited_mc0995_lib1 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "CGs_D35_unedited_mc0995_fracs_lib1 = get_fractions(CGs_D35_unedited_mc0995_lib1)\n",
    "print(\"CGs_D35_unedited_mc0995_lib1 shape:\", CGs_D35_unedited_mc0995_lib1.shape)\n",
    "print(\"CGs_D35_unedited_mc0995_fracs_lib1 shape: \", CGs_D35_unedited_mc0995_fracs_lib1.shape)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(16, 5))\n",
    "plt.bar(np.arange(len(CGs_D6_CRoff_mc0995_fracs)), CGs_D6_CRoff_mc0995_fracs, snap=False, alpha=0.5, label='mC > 0.995 CRISPRoff', color='deepskyblue')\n",
    "plt.bar(np.arange(len(CGs_D6_unedited_mc0995_fracs)), CGs_D6_unedited_mc0995_fracs, snap=False, alpha=0.5, label='mC > 0.995 Unedited', color='navy')\n",
    "# mediumblue.'turquoise', 'aquamarine', 'deepskyblue', 'dodgerblue', 'cornflowerblue', 'royalblue', 'blue', 'mediumblue', 'darkblue', 'navy', 'midnightblue'\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('CpG units')\n",
    "plt.ylabel('Fraction of mC')\n",
    "plt.title('Fraction of mC for CRISPROff and Unedited T cells CGs day 6 \\n mC threshold 99.5%')\n",
    "plt.legend()\n",
    "# Show the plot\n",
    "plt.show()\n",
    "\n",
    "\n",
    "# DIFFERENCES\n",
    "\n",
    "dif_D6_mCthresh_995 =   CGs_D6_CRoff_mc0995_fracs  - CGs_D6_unedited_mc0995_fracs\n",
    "\n",
    "plt.figure(figsize=(16, 5))\n",
    "plt.bar(np.arange(len(dif_D6_mCthresh_995)), dif_D6_mCthresh_995, snap=False, alpha=0.5, label='CRoff - Unedited, D6 with 99.5% mC', color='turquoise'  )\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('CpG units')\n",
    "plt.ylabel('Fraction of mC difference')\n",
    "plt.title('Fraction Differences between mC Unedited T cells Day 6 \\n confidence threshold 99.5%')\n",
    "plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# LASSO regression to find most important CGs for distinguishing CRISPRoff vs Unedited T cells day 6\n",
    "\n",
    "lasso_abs_coefs = np.load(\"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/model_data/lasso_abs_coefs_silenced_unedited_Tcells_day6_mCthresh0.995_2025-10-21.npy\") \n",
    "# /home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/model_data/lasso_abs_coefs_silenced_unedited_Tcells_day6_mCthresh0.92025-10-14.npy\")\n",
    "# \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/model_data/lasso_abs_coefs_silenced_unedited_Tcells_day6_mCthresh0.92025-10-14.npy\")\n",
    "print(\"lasso_abs_coefs shape: \", lasso_abs_coefs.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(20, 5))  # <-- Correct way to set figure size\n",
    "shifts = [-.4,-.15] #,.1]\n",
    "\n",
    "# for i in range(1,2):\n",
    "    # plt.bar(X_axis+shifts[i], np.abs(model.coef_[i]), .25,label=celltypes[i]) \n",
    "X_axis = np.arange(1, lasso_abs_coefs.shape[0]+1) # +1 # np.arange(10)+1    \n",
    "plt.bar(X_axis,lasso_abs_coefs) #, .25) #  model.coef_) # np.abs(model.coef_)), .25) # ,label=celltypes[i])\n",
    "\n",
    "plt.legend()\n",
    "plt.title(\"Absolute Value of LASSO Coefficient\")\n",
    "plt.xlabel(\"CpG Site\")\n",
    "plt.xticks(X_axis)\n",
    "plt.xticks(X_axis, fontsize=4)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plt.figure(figsize=(16, 5))\n",
    "plt.bar(np.arange(len(dif_D6_mCthresh_995)), dif_D6_mCthresh_995, snap=False, alpha=0.5, label='CRoff - Unedited, D6 with 99.5% mC', color='mediumblue'  )\n",
    "plt.bar(np.arange(len(lasso_abs_coefs)), lasso_abs_coefs, snap=False, alpha=0.5, label='LASSO Abs values from LR on CRoff vs Unedited, D6 with 99.5% mC', color='magenta'  )\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('CpG units')\n",
    "plt.ylabel('Fraction of mC difference')\n",
    "plt.title('Fraction Differences between mC Unedited T cells Day 6 \\n confidence threshold 99.5%')\n",
    "plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# plt.figure(figsize=(16, 5))\n",
    "# plt.bar(np.arange(len(dif_D6_mCthresh_995)), dif_D6_mCthresh_995, snap=False, alpha=0.5, label='CRoff - Unedited, D6 with 99.5% mC', color='mediumblue'  )\n",
    "# plt.bar(np.arange(len(lasso_abs_coefs)), lasso_abs_coefs, snap=False, alpha=0.5, label='LASSO Abs values from LR on CRoff vs Unedited, D6 with 99.5% mC', color='magenta'  )\n",
    "\n",
    "# # Add labels and title\n",
    "# plt.xlabel('CpG units')\n",
    "# plt.ylabel('Fraction of mC difference')\n",
    "# plt.title('Fraction Differences between mC Unedited T cells Day 6 \\n confidence threshold 99.5%')\n",
    "# plt.legend()\n",
    "\n",
    "# # Show the plot\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalize the lasso coefficients to a 0-1 range\n",
    "lasso_abs_coefs_norm = (lasso_abs_coefs - np.min(lasso_abs_coefs)) / (np.max(lasso_abs_coefs) - np.min(lasso_abs_coefs))\n",
    "print(\"Normalized LASSO coefficients:\", lasso_abs_coefs_norm) \n",
    "\n",
    "# plot normalized coefficients\n",
    "plt.figure(figsize=(16, 5))\n",
    "plt.bar(np.arange(len(lasso_abs_coefs_norm)), lasso_abs_coefs_norm, snap=False, alpha=0.5, label='Normalized LASSO Abs values from LR on CRoff vs Unedited, D6 with 99.5% mC', color='magenta'  )\n",
    "plt.xlabel('CpG units')\n",
    "plt.ylabel('Normalized LASSO Coefficient (0-1)')\n",
    "plt.title('Normalized LASSO Coefficients for CpG sites distinguishing CRoff vs Unedited T cells Day 6')\n",
    "plt.legend()\n",
    "plt.show()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Normalize the lasso coefficients to a 0-1 range\n",
    "# lasso_abs_coefs_norm = (lasso_abs_coefs - np.min(lasso_abs_coefs)) / (np.max(lasso_abs_coefs) - np.min(lasso_abs_coefs))\n",
    "# print(\"Normalized LASSO coefficients:\", lasso_abs_coefs_norm) \n",
    "\n",
    "# # plot normalized coefficients\n",
    "# plt.figure(figsize=(16, 5))\n",
    "# plt.bar(np.arange(len(lasso_abs_coefs_norm)), lasso_abs_coefs_norm, snap=False, alpha=0.5, label='Normalized LASSO Abs values from LR on CRoff vs Unedited, D6 with 99.5% mC', color='magenta'  )\n",
    "# plt.xlabel('CpG units')\n",
    "# plt.ylabel('Normalized LASSO Coefficient (0-1)')\n",
    "# plt.title('Normalized LASSO Coefficients for CpG sites distinguishing CRoff vs Unedited T cells Day 6')\n",
    "# plt.legend()\n",
    "# plt.show()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plt.figure(figsize=(16, 5))\n",
    "plt.bar(np.arange(len(dif_D6_mCthresh_995)), dif_D6_mCthresh_995, snap=False, alpha=0.5, label='CRoff - Unedited, D6 with 99.5% mC', color='mediumblue'  )\n",
    "plt.bar(np.arange(len(lasso_abs_coefs_norm)), lasso_abs_coefs_norm, snap=False, alpha=0.5, label='LASSO Norm Abs values from LR on CRoff vs Unedited, D6 with 99.5% mC', color='magenta'  )\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('CpG units')\n",
    "plt.ylabel('Fraction of mC difference')\n",
    "plt.title('Fraction Differences between mC Unedited T cells Day 6 \\n confidence threshold 99.5%')\n",
    "plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# plt.figure(figsize=(16, 5))\n",
    "# plt.bar(np.arange(len(dif_D6_mCthresh_995)), dif_D6_mCthresh_995, snap=False, alpha=0.5, label='CRoff - Unedited, D6 with 99.5% mC', color='mediumblue'  )\n",
    "# plt.bar(np.arange(len(lasso_abs_coefs_norm)), lasso_abs_coefs_norm, snap=False, alpha=0.5, label='LASSO Norm Abs values from LR on CRoff vs Unedited, D6 with 99.5% mC', color='magenta'  )\n",
    "\n",
    "# # Add labels and title\n",
    "# plt.xlabel('CpG units')\n",
    "# plt.ylabel('Fraction of mC difference')\n",
    "# plt.title('Fraction Differences between mC Unedited T cells Day 6 \\n confidence threshold 99.5%')\n",
    "# plt.legend()\n",
    "\n",
    "# # Show the plot\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normolize dif_D6_mCthresh_995 to max 1\n",
    "dif_D6_mCthresh_995_max1 = dif_D6_mCthresh_995 / np.max(np.abs(dif_D6_mCthresh_995))\n",
    "\n",
    "# Normalized dif_D6_mCthresh_995 to be in range 0-1\n",
    "dif_D6_mCthresh_995_norm01 = (dif_D6_mCthresh_995 - np.min(dif_D6_mCthresh_995)) / (np.max(dif_D6_mCthresh_995) - np.min(dif_D6_mCthresh_995))\n",
    "print(\"Normalized dif_D6_mCthresh_995:\", dif_D6_mCthresh_995_norm01)      \n",
    "\n",
    "plt.figure(figsize=(16, 5))\n",
    "plt.bar(np.arange(len(dif_D6_mCthresh_995)), dif_D6_mCthresh_995, snap=False, alpha=0.5, label='CRoff - Unedited, D6 with 99.5% mC')#, color='mediumblue'  )\n",
    "plt.bar(np.arange(len(dif_D6_mCthresh_995_max1)), dif_D6_mCthresh_995_max1, snap=False, alpha=0.5, label='max1(CRoff - Unedited), D6 with 99.5% mC')#, color='mediumblue'  )\n",
    "plt.bar(np.arange(len(dif_D6_mCthresh_995_norm01)), dif_D6_mCthresh_995_norm01, snap=False, alpha=0.5, label='norm01(CRoff - Unedited) -min/(max-min), D6 with 99.5% mC')#, color='mediumblue'  )\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('CpG units')\n",
    "plt.ylabel('Fraction of mC difference')\n",
    "plt.title('Fraction Differences between mC Unedited T cells Day 6 \\n confidence threshold 99.5%')\n",
    "plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normolize dif_D6_mCthresh_995 to max 1\n",
    "dif_D6_mCthresh_995_max1 = dif_D6_mCthresh_995 / np.max(np.abs(dif_D6_mCthresh_995))\n",
    "\n",
    "# Normalized dif_D6_mCthresh_995 to be in range 0-1\n",
    "dif_D6_mCthresh_995_norm01 = (dif_D6_mCthresh_995 - np.min(dif_D6_mCthresh_995)) / (np.max(dif_D6_mCthresh_995) - np.min(dif_D6_mCthresh_995))\n",
    "print(\"Normalized dif_D6_mCthresh_995:\", dif_D6_mCthresh_995_norm01)      \n",
    "\n",
    "plt.figure(figsize=(16, 5))\n",
    "plt.bar(np.arange(len(dif_D6_mCthresh_995)), dif_D6_mCthresh_995, snap=False, alpha=0.5, label='CRoff - Unedited, D6 with 99.5% mC')#, color='mediumblue'  )\n",
    "plt.bar(np.arange(len(dif_D6_mCthresh_995_max1)), dif_D6_mCthresh_995_max1, snap=False, alpha=0.5, label='max1(CRoff - Unedited), D6 with 99.5% mC')#, color='mediumblue'  )\n",
    "# plt.bar(np.arange(len(dif_D6_mCthresh_995_norm01)), dif_D6_mCthresh_995_norm01, snap=False, alpha=0.5, label='norm01(CRoff - Unedited) -min/(max-min), D6 with 99.5% mC')#, color='mediumblue'  )\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('CpG units')\n",
    "plt.ylabel('Fraction of mC difference')\n",
    "plt.title('Fraction Differences between mC Unedited T cells Day 6 \\n confidence threshold 99.5%')\n",
    "plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plt.figure(figsize=(16, 5))\n",
    "plt.bar(np.arange(len(dif_D6_mCthresh_995)), dif_D6_mCthresh_995, snap=False, alpha=0.5, label='CRoff - Unedited, D6 with 99.5% mC', color='mediumblue'  )\n",
    "plt.bar(np.arange(len(lasso_abs_coefs_norm)), lasso_abs_coefs_norm, snap=False, alpha=0.5, label='LASSO Norm Abs values from LR on CRoff vs Unedited, D6 with 99.5% mC', color='magenta'  )\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('CpG units')\n",
    "plt.ylabel('Fraction of mC difference')\n",
    "plt.title('Fraction Differences between mC Unedited T cells Day 6 \\n confidence threshold 99.5%')\n",
    "plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# plt.figure(figsize=(16, 5))\n",
    "# plt.bar(np.arange(len(dif_D6_mCthresh_995_max1)), dif_D6_mCthresh_995_max1, snap=False, alpha=0.5, label='max1(CRoff - Unedited), D6 with 99.5% mC', color='deepskyblue'  )\n",
    "# plt.bar(np.arange(len(lasso_abs_coefs_norm)), lasso_abs_coefs_norm, snap=False, alpha=0.5, label='LASSO Norm Abs values from LR on CRoff vs Unedited, D6 with 99.5% mC', color='magenta'  )\n",
    "\n",
    "# # Add labels and title\n",
    "# plt.xlabel('CpG units')\n",
    "# plt.ylabel('Fraction of mC difference')\n",
    "# plt.title('Fraction Differences between mC Unedited T cells Day 6 \\n confidence threshold 99.5%')\n",
    "# plt.legend()\n",
    "\n",
    "# # Show the plot\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plt.figure(figsize=(16, 5))\n",
    "plt.bar(np.arange(len(dif_D6_mCthresh_995)), dif_D6_mCthresh_995, snap=False, alpha=0.5, label='CRoff - Unedited, D6 with 99.5% mC', color='deepskyblue'  )\n",
    "plt.bar(np.arange(len(lasso_abs_coefs_norm)), lasso_abs_coefs_norm, snap=False, alpha=0.5, label='LASSO Norm Abs values from LR on CRoff vs Unedited, D6 with 99.5% mC', color='magenta'  )\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('CpG units')\n",
    "plt.ylabel('Fraction of mC difference')\n",
    "plt.title('Fraction Differences between mC Unedited T cells Day 6 \\n confidence threshold 99.5%')\n",
    "plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# plt.figure(figsize=(16, 5))\n",
    "# plt.bar(np.arange(len(dif_D6_mCthresh_995)), dif_D6_mCthresh_995, snap=False, alpha=0.5, label='CRoff - Unedited, D6 with 99.5% mC', color='deepskyblue'  )\n",
    "# plt.bar(np.arange(len(lasso_abs_coefs_norm)), lasso_abs_coefs_norm, snap=False, alpha=0.5, label='LASSO Norm Abs values from LR on CRoff vs Unedited, D6 with 99.5% mC', color='magenta'  )\n",
    "\n",
    "# # Add labels and title\n",
    "# plt.xlabel('CpG units')\n",
    "# plt.ylabel('Fraction of mC difference')\n",
    "# plt.title('Fraction Differences between mC Unedited T cells Day 6 \\n confidence threshold 99.5%')\n",
    "# plt.legend()\n",
    "\n",
    "# # Show the plot\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18",
   "metadata": {},
   "source": [
    "# Load SHAP coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19",
   "metadata": {},
   "outputs": [],
   "source": [
    "shap_sumabs_coefs = np.load(\"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/model_data/shap_sumabs_coefs_silenced_unedited_Tcells_day6_mCthresh0.995_2025-10-21.npy\")\n",
    "    # \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/model_data/shap_sumabs_coefs_silenced_unedited_Tcells_day6_mCthresh0.92025-10-14.npy\")\n",
    "shap_sumabs_coefs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# shap_norm_sumabs_coefs="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "22",
   "metadata": {},
   "source": [
    "# Generate real CpG pair coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# --- STEP 1: Load genomic sequence ---\n",
    "# Paste your hg38 DNA sequence here as a string (no line breaks)\n",
    "# Example placeholder sequence (replace with real one)\n",
    "genomic_sequence = \"\"\"\n",
    "GTCATTGTATGGTTCCTGGGATAAGAGTAAGAACCGTGGTCCCTTGCTCTGTCAAAATCATAAACCTATTTTCTTTGTATTTTGACTTAATTACATAAGCTTGTTTTGGATTGGAAGGTCTATAGAGACTATTTTAGTGCCAAATGCTATTTAATCAATGCAGAAATAAGCCCTTGAGAGGTGCAGTGACTCAGCTGAGGACACACTGATAGTTTCTGGTAGATAAGTATGAAGCCTTGGAACTACTTCTCATATCATAAGCCCCTGGGTTCCCAAATCACCTCCCCAAGTACCTAATTCATCATCCTCTGCATACAAGAAATTTTTGCTCATTCCAATAAAAAGAATGTAGTACGGCAGGACAGGAGAAGTACAGGTTTTGGAATCAAACTAGGTTTAGATTCTGACCTGGTCGCTAGTTAGTGCATGGGCTTTGGTAAACAGCCTCTCAGCCTGTTTCCTTGTCTGCAAAATATGGATAGTTATCTTCACTTCATGAAGTTGTGTTGGTGAAATGGTATAATGTGTATAAAGCAACTGGAACACAACAAATGAGGCCATTTATTTAAAATAATCTTTAGTAAGTTATAAATTCACATATTTGCTTTTTTCTGAGGAATGCTTAGATTTTATCATAGATCTCTTTTCTAAATTGAGTAAATACATAGTGGATTTCTTTGTTTCCTTCCCCTCAGAATGTTGAAATTTGGCTGATTGTAACTATTACTTAAGACTCTATCCAGTAATTGTTCGTAAAATTTCTTTTCTTTTTTTTTTTTTGAGACGGAGTTTTGCTCTCGTCCCCCAGGCTGGAGTGCAATGGCGCGATCTTGGCTCACTGCAACCTCCGCCTCCCAGGTTCAAGTGATTCTCCCGCCTCAGCCTCCCAAGTAGCTGGGATTACAGGCACCTGCCACAACGCCCAGCCAATTTTTGTATTTTTAGTAGAGACGGGGTTTCACCATGTTGGCCAGGCTAGTCTGGAACTCCTGACCTCAGGTGATCCTCCTGCCTTGGCCTCCCAAAGTGCTGAGATTACAGGCGTGAGCCACCACGCCCGGCCAAAATTTCCTTTTAAAAAAAGTTATTGCCAATAACTAAGTATGCTGGAAGGTGAATGCTTAGGACATTTAGTGTCACTGAGCTGTGTGATAGAGATAATATTTTTGACACTTTGCAGAGCTCATTGTTTTCTCTCCATATAAAGCTATGACATTCACATAATAAACATGTAGAGAGCTTTAACAGACTTCAATGTTAAGTCTTCTAACAGCTGCATAACTGGTCTTGCTTGAACCTGTCTTTTTCATAATGCTACCTTGCCTTTCACACCAATTCTGTGAGCCCAAGCAGGAATGAAGATTACCTATCTCTAATGGCAGGACAGAAGGGAGTGGTGGTTAAGGTTACTGGAGGTTATTGGGGGAATAAATGAAGACCTCATGCCACTTCTAAACACTGCCAGGACCTAACATACCTTGTGTTTTTATATCCAATATGGGAGTAAAGGTTAGATATTTTAACATCAAATATGAGGATGACACTAGATGGTTTTAAAAGCCCCCCACCATAACTTCTAAACTCCACAATGTTCGAATATTGTTAGGTTTCAGTGAAATCCTACAGTAGTCCCTTGCTTGGAATCACTGTAGGATTGGCTCCAGCAATGGGCAACACAAGGTTTGGTTCTATACAGTAGGTATGGGTCAAAGAAAATGCTATGATTGAAAAATTGCTGTTTTCCTCCTCAAAAAGCTCATACCAACCAGGAAAGAGCTATCAGAAAACCTTCAGGAAAGAATATAATGATAGGATATAACAAATACCCCCACAGCTGGTCCATAGACCAAAGTATAATTCCAAGTAGGGAATTTCACTTAACGAGGTAAGGCATTTAAGGGAAAAGGAATCTAGGCTAGGACTCTGCTAGCCAGACCCAGATAAATTATTACAGGGAAGCATGTTCTTGAGGCTATTCTCTATGAAGAAGGGCAACCCACCTTGCCATCTATATCCTTTCTGTTTTCGACAGGGGTGGTCCTAAGCTGGATAGATCTTTACTCTGACCAGTACCATATAGGTCTGTGGTTCTCAAACCTGGCTGCACAGTGTGGCACTTTAAGTAATCCTGATACCCAGTCCCATCCCCAAAGATGCTGATTTAATTTGTCTAAGGTGCAGTTTGGGCCTCTGGATTGGAACCAACCCATATACCCATCAATGATAGACTGGATAAAGAAAATGTGGCACATATACACCATAGAATACTACACAGCCATAAAAAAGGATGAGTTCATGTCCTATGCAGGGACATGGATGAAGCTGGAAACCATCAACCTCAGCAAACTAACATAAGACCAGAAAACCAAACACTGCATGTTCTTACTCATAAGTGGGAGTTGAACAATGAGAATACACAGACACAGGGAGGGGAACATCACACACGGGGGCCTGTTGGAGGGTGGGGGTTAGGGGAGGGATAGCATTAGGAGAAATACCTAATGTAGATGTTGGGTTGATGGGTGCAGCAAACCACCATGGCACGTGTATACCTATGTAACAAACCTGCATGTTCTGCACATGTATTCCAGAACTTAAAGAATATATATTTATATATGTATAAATATATATGAATATATAAATATAAATATATATATAAGCTCCCCACGTGATTCTAATGTGTGGCCAGTCTTGCAAACCTCAGGCTATTTATTCTAAAGGGTTTGTAACAAGCCTTCCTCTTAAAATCCACCCCACCACCATTCAGCGTTTTGGGGTCTGTTTGTTTTTATTGTTATCCCACCCCACACCGCCCCGAGGCGCGCGCGCGCGCGCGCACACACACACACACACACACACACACACATACACACACGCACACTGGTGAATTTCTCTCTACAGTCAGTCTGGAGTAATCCCAAAGTGGTGTCTTTCGTAAATAAGGAGAACCCGGGTGAAGAAAATGACTCCCACCCGAACAAGGCATGAACAATGTTCACTCCCTACTGTGTTATTCAACCTGTTTCCCCAGGTCTCTGTTTTCACATTAGAGAGTGTTCTAGGAGATGACGCCCTTCCTCCTTAGTTATTTCCCCACCCTCGTGCTGGCCTTTGACAGACCTCCCAGTAGAGGGCCCAAGACGCGGGTAGAGCACCGCGTCTCAGCGCCTGAGTCTCAGCCCCCGAACTCCACCGCACCTGCAGGTCCCCTTGGCAGCACTCAAGCGCGGGGATGCTCCGCTTAGACGAACTCACGTGCGGGCAGCAAGGCCTGCGATACTTGAGCACCCCTCCCCCTCTCCCGTTTACACCCCGTTTGTGTTTACGTAGCGAGGAGATATTTAGGTTTCTAGAAGGCAGGTCATCGCAGGCCCCACCCAGCAGTGGAGAGAGTGAGTCCCAGAGGGTGTTGCCAGCGAGCTCCTCCTCCTTCCCCTCCCCACTCTCCCCGAGTCTAGGGCCCCCGGGGCGTATGACGCCGGAGCCCTCTGACCGCACCTCTGACCACAACAAACCCCTACTCCACCCGTCTTGTTTGTCCCACCCTTGGTGACGCAGAGCCCCAGCCCAGACCCCGCCCAAAGCACTCATTTAACTGGTATTGCGGAGCCACGAGGCTTCTGCTTACTGCAACTCGCTCCGGCCGCTGGGCGTAGCTGCGACTCGGCGGAGTCCCGGCGGCGCGTCCTTGTTCTAACCCGGCGCGCCATGACCGTCGCGCGGCCGAGCGTGCCCGCGGCGCTGCCCCTCCTCGGGGAGCTGCCCCGGCTGCTGCTGCTGGTGCTGTTGTGCCTGCCGGCCGTGTGGGGTGAGTAGGGGCCCGGCGGCCGGGGAAGCCCCTGGGCTGGGTGGGAGGTCCAAGTCGGTCTCTGAGACACGCACAGGGGCCGGCGACTTGGCAGGTGGGGAGCTTGGCCCGCGGTCGTGGTTCCCGCCGTCCTGTGCCTTTAAGGCTCTCGCCGCTCACCAGCATTTGGGGCTCCTGCTGTGTCGGCCCCCAGCTGACTTGGCTTTAGGGGTCGGCGTGGAGGGTTAAAGAGGCCCCGGCTGGGTTTGCGGAGCAGCCAAGCCTGGCAAAATCGAAAGGGAGGGCTCAAAGAGACTGTATCCTTAACCCCCAAAAAGCTGGTCTAAAAGGATGGGAGGCCAGACCGCTGACCGTTCCCCACTCTCGACAGAGTCCAGCCGTGTGGAGCACACGATGCTGCAAACTTGCATGTCATCTCTTTCAGGTGTGGCATTTCAAGGGGGCTTGTGTCTTGAAAACAGCAACTGTGAGGACACTTGATAGTCATTTCCTTCAGTTCTGCTTTTGTCTCCCTAGGTGACTGTGGCCTTCCCCCAGATGTACCTAATGCCCAGCCAGCTTTGGAAGGCCGTACAAGTTTTCCCGAGGATACTGTAATAACGTACAAATGTGAAGAAAGCTTTGTGAAAATTCCTGGCGAGAAGGACTCAGTGATCTGCCTTAAGGGCAGTCAATGGTCAGATATTGAAGAGTTCTGCAATCGTAAGTTCTTCATCTTTTTAGAAAAGTTCTGGGAATGGAATGTATCTTAAATTTATTTTTATATACCTTTGGAGTGACTAGTAATTGATAGTTCTCTAGCGTTACTAAACCCCAGGGTATACCCTGTTGGCACGTCACACTCCAGCTAATTGAAGACATTTACCACCCTGAGTAGGTCCTGCTGCTTCATTAAGAGTTTTTCTTCCTAACACTGTACCTCTTGAACGAACTAGTAGAGAAAATAAACAATAGGTATTGTTTAGGGAGTAAACATATAAAATATATTTGGCAATTGCCTTGAGACACATAGCGTAAATGATATGTTTTCTTAAGTATTTTATAAAGTAGGACTTACCAAATACAACTCTGTGGTGGCATTTTTAAAAGCTCTCAATATTTACTTCAAAATCTGTTATACTTGTACATATTTATAGGGTACATGTGATATTCTGGCACATGCATACAGTGTGTCATGATCAATCAGGGTAATTGGATAGCCAGTACCTCAAACACTTGGTGTTTGTGTGGAGAATATTCCAAATCTTCTCTTCTTGCTACTTTGAAAAAATATATGTATATATATAAATTGGGATATATATATAGGGAGAGAGATATATATAATTGGGATATATATATATAAGGAGATATATAATTGGGATATATATAAGGAGATATATATAGGGAGAGATATATATATATAGGGAGATATATATATATAGGGAGAGAGATATATATAGGGAGAGAGAGATGTATAAACTGGGATATAGACAGGTAGATAGGTTGATTGTTAAAAATGAATAAGAGGCCCAGTGAAGAGTTGTTTGGGGGCTAACACAGTAACCGTGTTTGATACTGATACCTAGTAGAATTAAAAGAAATGAAGATTTCTTGTTTTAGGTGAATATACATACAGTCATCCACACCTTATGTGCCCTTTGATTTCATAAAATATATATTGGCATTATCTTGGAGGAAAAAAATCTGTTACTTACATAGCCCTGTATTAGCTTTGAGATATATGAGTAGAGATAAAGATAAAGAGACCCGATAACCTGACTTTGAACAACTTAAATCAATAGAAGCTACTGTAAATACAGAAATTGGCACAAATACCTTTAGAAGCCATGCGTGAAATAGGCTTTAACTTTTTATTTTTATCCACAATAGAGGCAGTAGGATGAAGTACATAGTCTTTGCACATGTATCATGTTTTAACTCTTCCCCATGTTATTACTTTATCCAATCCTCCAAACAACCTTATGAGGTAAGTTGGGGCTCTTACACTCTTTTTATGGCGGAGGAGACAGACCCCAGTAGTTTAATAACTTCCTCAAGAGCAGCAGCTTTTACGCCAAAACCAAGGTGTAAACTTGGGTCTTCTGACTACAAATCCAATGATGACCCTTCCACTTGACCATGCCTCTCTAGATTCACATTTGTGGGCAATCTGTTTACAGATGTAAATACTTTATGTATACAATTTGTGGGCAATCTGTTTACAGGTGTAAATACTTTATGTATACAATTGACTCATCCTTGCCCCCAGGGCTCATGGGAATTAAATGAGTTTGTACGTAAATAAATATTAGCTGCTATTATTGTTGTGCAGGACATTAGTCCAGGGGTCTGGAAATACCCTATTCCTGTACTGTATGGTGCTGTGTAGTTCTTTAGGGCAGGGTAAGGAAACCAACATTTATTGAATGCCAGAACTGGGCCAGGCAACTTAAATTTTTTTTTTTTTTTAAGTGGGGATAGTGATATGTGTGTACCAACTCTACTTCCTAAGCCTCTAGTAACTAGGAAACAAAGTATACATGTGAATACATTTGCTACATTGCAAGTGCCATGGTGATATTAATATAGTTAAGTACTAAATATGCGCAAAGCAGTAAAGAAAGGGGGTTATTAGGGTCCAGATAATTAAATATAGATTATAAAACAAAAATTGATACTACATTTTTTGTTGCTGCTTTTGTTAATACTTTTA\n",
    "\"\"\".replace('\\n', '').upper()\n",
    "\n",
    "# Starting coordinate in hg38 for this sequence\n",
    "region_start_hg38 = 207318058\n",
    "region_chr = \"chr1\"\n",
    "\n",
    "# --- STEP 2: Scan for CpG sites ---\n",
    "cpg_records = []\n",
    "for i in range(len(genomic_sequence) - 1):\n",
    "    if genomic_sequence[i:i+2] == \"CG\":\n",
    "        chrom_start = region_start_hg38 + i        # 0-based start\n",
    "        chrom_end = chrom_start + 2                # BED-style: end is exclusive\n",
    "        cpg_records.append({\n",
    "            \"chrom\": region_chr,\n",
    "            \"start\": chrom_start,\n",
    "            \"end\": chrom_end,\n",
    "            \"name\": f\"CpG_{len(cpg_records) + 1}\",\n",
    "            \"score\": 0,\n",
    "            \"strand\": \"+\"\n",
    "        })\n",
    "\n",
    "# --- STEP 3: Create BED dataframe ---\n",
    "df_cpgs = pd.DataFrame(cpg_records)\n",
    "\n",
    "# Optional: Save to BED\n",
    "df_cpgs.to_csv(\"real_CpG_coords_hg38.bed\", sep=\"\\t\", index=False, header=False)\n",
    "\n",
    "# Preview\n",
    "print(df_cpgs.head())\n",
    "print(f\"\\nTotal CpGs found: {len(df_cpgs)}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24",
   "metadata": {},
   "source": [
    "Additions:\n",
    "\n",
    "T2T Coordinates â€” already present\n",
    "\n",
    "hg38 Coordinates â€” mapped via known offset\n",
    "\n",
    "TSS-relative distance (from TSS = center of region)\n",
    "\n",
    "Promoter/gene body/upstream region classification (already present)\n",
    "\n",
    "Î”mC values, significance\n",
    "\n",
    "CpG Index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# --- STEP 1: Load hg38 genomic sequence as a string ---\n",
    "# genomic_sequence = \"\"\"\n",
    "# <PASTE_FULL_SEQUENCE_HERE>\n",
    "# \"\"\".replace('\\n', '').upper()\n",
    "\n",
    "# --- STEP 2: Define genomic parameters ---\n",
    "region_chr = \"chr1\"\n",
    "region_start_hg38 = 207318058  # Start of your region in hg38\n",
    "cd55_tss_hg38 = 207321678           # Real CD55 TSS in hg38\n",
    "t2t_offset = -734704           # T2T = hg38 + offset\n",
    "cd55_tss_t2t= cd55_tss_hg38 + t2t_offset\n",
    "\n",
    "# --- STEP 3: Find CpG dinucleotides ---\n",
    "cpg_records = []\n",
    "for i in range(len(genomic_sequence) - 1):\n",
    "    if genomic_sequence[i:i+2] == \"CG\":\n",
    "        start_hg38 = region_start_hg38 + i\n",
    "        end_hg38 = start_hg38 + 2\n",
    "        start_t2t = start_hg38 + t2t_offset\n",
    "        end_t2t = end_hg38 + t2t_offset\n",
    "        rel_tss_hg38 = start_hg38 - cd55_tss_hg38\n",
    "        rel_tss_t2t = start_t2t - cd55_tss_t2t\n",
    "\n",
    "        cpg_records.append({\n",
    "            \"chrom\": region_chr,\n",
    "            \"start_hg38\": start_hg38,\n",
    "            \"end_hg38\": end_hg38,\n",
    "            \"start_T2T\": start_t2t,\n",
    "            \"end_T2T\": end_t2t,\n",
    "            \"rel_to_TSS_hg38\": rel_tss_hg38,\n",
    "            \"rel_to_TSS_T2T\": rel_tss_t2t,\n",
    "            \"name\": f\"CpG_{len(cpg_records) + 1}\",\n",
    "            \"strand\": \"+\"\n",
    "        })\n",
    "\n",
    "# --- STEP 4: Create DataFrame from CpG positions ---\n",
    "df_cpgs = pd.DataFrame(cpg_records)\n",
    "\n",
    "# --- STEP 5: Add Î”mC values ---\n",
    "# Replace this with your actual 1D array of 142 values\n",
    "# dif_D6_mCthresh_995 = np.linspace(-0.1, 0.5, len(df_cpgs))  # Simulated for structure\n",
    "\n",
    "if len(dif_D6_mCthresh_995) != len(df_cpgs):\n",
    "    raise ValueError(\"Î”mC array length does not match number of CpGs found!\")\n",
    "\n",
    "df_cpgs[\"Î”mC\"] = dif_D6_mCthresh_995\n",
    "df_cpgs[\"CpG Index\"] = np.arange(1, len(dif_D6_mCthresh_995)+1)\n",
    "df_cpgs[\"score\"] = (df_cpgs[\"Î”mC\"] * 1000).round().clip(lower=0, upper=1000).astype(int)\n",
    "\n",
    "# --- STEP 6: Annotate region (Promoter, Flank, Gene Body) ---\n",
    "def classify_region(pos, tss, window=1000):\n",
    "    if tss - window <= pos <= tss + window:\n",
    "        return \"Promoter\"\n",
    "    elif pos < tss - window:\n",
    "        return \"Upstream Flank\"\n",
    "    else:\n",
    "        return \"Gene Body / Downstream\"\n",
    "\n",
    "df_cpgs[\"Region\"] = df_cpgs[\"start_hg38\"].apply(lambda x: classify_region(x, cd55_tss_hg38))\n",
    "\n",
    "# --- STEP 7: Final column ordering ---\n",
    "df_cpgs = df_cpgs[[\n",
    "    \"CpG Index\", \"chrom\", \"start_hg38\", \"end_hg38\",\n",
    "    \"start_T2T\", \"end_T2T\",\n",
    "    \"name\", \"Î”mC\", \"score\",\n",
    "    \"rel_to_TSS_hg38\", \"rel_to_TSS_T2T\",\n",
    "    \"Region\", \"strand\"\n",
    "]]\n",
    "\n",
    "# --- STEP 8: Save and preview ---\n",
    "df_cpgs.to_csv(\"CD55_CpGs_Annotated_T2T_hg38.csv\", index=False)\n",
    "print(df_cpgs.head())\n",
    "print(f\"\\nâœ… Total CpGs found: {len(df_cpgs)}\")\n",
    "\n",
    "df_cpgs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26",
   "metadata": {},
   "source": [
    "Absolutely â€” here's the code to export your annotated CpG DataFrame (df_cpgs) to a clean, UCSC-compatible BED file with:\n",
    "\n",
    "chr, start, end\n",
    "\n",
    "CpG name\n",
    "\n",
    "score scaled from Î”mC (0â€“1000)\n",
    "\n",
    "unstranded (.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- STEP 1: Create BED DataFrame from annotated CpG table ---\n",
    "df_bed = pd.DataFrame({\n",
    "    \"chrom\": df_cpgs[\"chrom\"],\n",
    "    \"start\": df_cpgs[\"start_hg38\"],\n",
    "    \"end\": df_cpgs[\"end_hg38\"],\n",
    "    \"name\": df_cpgs[\"name\"],\n",
    "    \"score\": df_cpgs[\"score\"],\n",
    "    \"strand\": df_cpgs[\"strand\"]\n",
    "})\n",
    "\n",
    "# --- STEP 2: Sort by genomic start (optional but recommended) ---\n",
    "df_bed = df_bed.sort_values(by=[\"chrom\", \"start\"])\n",
    "\n",
    "# --- STEP 3: Save as tab-delimited BED file (no header) ---\n",
    "df_bed.to_csv(\"CD55_CpGs_UCSC_hg38.bed\", sep=\"\\t\", index=False, header=False)\n",
    "\n",
    "print(\"âœ… UCSC BED file saved as: CD55_CpGs_UCSC_hg38.bed\")\n",
    "\n",
    "df_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import numpy as np\n",
    "# import pandas as pd\n",
    "\n",
    "# # --- STEP 0: Load your real Î”mC array ---\n",
    "# # Replace this with your real array\n",
    "# # dif_D6_mCthresh_995 = np.array([\n",
    "# #     # Your full Î”mC array goes here (142 values)\n",
    "# # ])\n",
    "\n",
    "# # # --- STEP 1: Define region and reference info ---\n",
    "# # region_chr = \"chr1\"\n",
    "# # region_start_T2T = 206583354\n",
    "# # region_end_T2T = 206589854\n",
    "# # region_length = region_end_T2T - region_start_T2T\n",
    "# # tss_T2T = region_start_T2T + (region_length // 2)  # Assume midpoint as TSS\n",
    "\n",
    "# # # T2T â†’ hg38 coordinate offset\n",
    "# # offset_hg38_vs_t2t = +734704\n",
    "# # region_start_hg38 = region_start_T2T + offset_hg38_vs_t2t\n",
    "# # region_end_hg38 = region_end_T2T + offset_hg38_vs_t2t\n",
    "# # cd55_tss_hg38 = tss_T2T + offset_hg38_vs_t2t\n",
    "\n",
    "# # --- STEP 2: Generate CpG positions ---\n",
    "# cpg_count = len(dif_D6_mCthresh_995)\n",
    "# cpg_indices = np.arange(1, cpg_count + 1)\n",
    "# cpg_positions_T2T = np.linspace(region_start_T2T, region_end_T2T, cpg_count, dtype=int)\n",
    "# cpg_positions_hg38 = cpg_positions_T2T + offset_hg38_vs_t2t\n",
    "# rel_to_TSS_T2T = cpg_positions_T2T - tss_T2T\n",
    "# rel_to_TSS_hg38 = cpg_positions_hg38 - cd55_tss_hg38\n",
    "\n",
    "# # --- STEP 3: Region classifier ---\n",
    "# def classify_region(pos, tss, window=1000):\n",
    "#     if tss - window <= pos <= tss + window:\n",
    "#         return \"Promoter\"\n",
    "#     elif pos < tss - window:\n",
    "#         return \"Upstream Flank\"\n",
    "#     else:\n",
    "#         return \"Gene Body / Downstream\"\n",
    "\n",
    "# # --- STEP 4: Create annotation DataFrame ---\n",
    "# df_dmc = pd.DataFrame({\n",
    "#     \"Chromosome\": region_chr,\n",
    "#     \"CpG Index\": cpg_indices,\n",
    "#     \"Genomic Position (T2T v2.0)\": cpg_positions_T2T,\n",
    "#     \"Genomic Position (hg38)\": cpg_positions_hg38,\n",
    "#     \"Î”mC\": dif_D6_mCthresh_995,\n",
    "#     \"T2T_rel_to_TSS\": rel_to_TSS_T2T,\n",
    "#     \"hg38_rel_to_TSS\": rel_to_TSS_hg38,\n",
    "# })\n",
    "\n",
    "# # --- STEP 5: Annotate significance & region type ---\n",
    "# df_dmc[\"Significant\"] = (df_dmc[\"Î”mC\"] > 0.3) | (df_dmc[\"Î”mC\"] < -0.1)\n",
    "# df_dmc[\"Region\"] = df_dmc[\"Genomic Position (T2T v2.0)\"].apply(\n",
    "#     lambda x: classify_region(x, tss_T2T)\n",
    "# )\n",
    "\n",
    "# # --- STEP 6: Optional rearrange columns ---\n",
    "# df_dmc = df_dmc[[\n",
    "#     \"Chromosome\", \"CpG Index\", \n",
    "#     \"Genomic Position (T2T v2.0)\", \"T2T_rel_to_TSS\",\n",
    "#     \"Genomic Position (hg38)\", \"hg38_rel_to_TSS\",\n",
    "#     \"Î”mC\", \"Significant\", \"Region\"\n",
    "# ]]\n",
    "\n",
    "# # --- STEP 7: Save and preview ---\n",
    "# df_dmc.to_csv(\"Annotated_Delta_mC_T2T_hg38.csv\", index=False)\n",
    "# print(df_dmc.head())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # import numpy as np\n",
    "# # import pandas as pd\n",
    "\n",
    "# # # Î”mC array (replace with your actual data if loading from file)\n",
    "# # # dif_D6_mCthresh_995 = np.array([...])  # Your full Î”mC array goes here\n",
    "\n",
    "# # # Metadata from user\n",
    "# # region_chr = \"chr1\"\n",
    "# # region_start = 206583354\n",
    "# # region_end = 206589854\n",
    "# # region_length = region_end - region_start\n",
    "# # cpg_count = len(dif_D6_mCthresh_995)\n",
    "\n",
    "# # # Step 1: Generate CpG index and position\n",
    "# # cpg_indices = np.arange(1, cpg_count + 1)\n",
    "# # cpg_positions = np.linspace(region_start, region_end, cpg_count, dtype=int)\n",
    "\n",
    "# # # Step 2: Define TSS (midpoint) and promoter boundaries\n",
    "# # cd55_tss = region_start + (region_length // 2)  # Assuming TSS = center\n",
    "# # promoter_start = cd55_tss - 1000\n",
    "# # promoter_end = cd55_tss + 1000\n",
    "\n",
    "# # Step 3: Classify region\n",
    "# def classify_region(pos):\n",
    "#     if promoter_start <= pos <= promoter_end:\n",
    "#         return \"Promoter\"\n",
    "#     elif pos < promoter_start:\n",
    "#         return \"Upstream Flank\"\n",
    "#     else:\n",
    "#         return \"Gene Body / Downstream\"\n",
    "\n",
    "# # Step 4: Create DataFrame\n",
    "# df_dmc = pd.DataFrame({\n",
    "#     \"CpG Index\": cpg_indices,\n",
    "#     \"Î”mC\": dif_D6_mCthresh_995,\n",
    "#     \"Genomic Position (T2T v2.0)\": cpg_positions\n",
    "# })\n",
    "\n",
    "# # Step 5: Annotate significant Î”mC\n",
    "# df_dmc[\"Significant\"] = (df_dmc[\"Î”mC\"] > 0.3) | (df_dmc[\"Î”mC\"] < -0.1)\n",
    "\n",
    "# # Step 6: Annotate region type\n",
    "# df_dmc[\"Region\"] = df_dmc[\"Genomic Position (T2T v2.0)\"].apply(classify_region)\n",
    "\n",
    "# # Step 7: Add chromosome column\n",
    "# df_dmc[\"Chromosome\"] = region_chr\n",
    "\n",
    "# # Step 8: Rearrange columns\n",
    "# df_dmc = df_dmc[[\n",
    "#     \"Chromosome\", \"CpG Index\", \"Genomic Position (T2T v2.0)\",\n",
    "#     \"Î”mC\", \"Significant\", \"Region\"\n",
    "# ]]\n",
    "\n",
    "# # Optional: save to CSV\n",
    "# df_dmc.to_csv(\"Annotated_Delta_mC_T2T.csv\", index=False)\n",
    "\n",
    "# # Print preview\n",
    "# print(df_dmc.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas as pd\n",
    "\n",
    "# # --- STEP 1: Load genomic sequence (hg38) ---\n",
    "# genomic_sequence = \"\"\"\n",
    "# GTCATTGTATGGTTCCTGGGATAAGAGTAAGAACCGTGGTCCCTTGCTCTGTCAAAATCATAAACCTATTTTCTTTGT...\n",
    "# \"\"\".replace('\\n', '').upper()\n",
    "\n",
    "# # --- STEP 2: Define coordinates ---\n",
    "# region_chr = \"chr1\"\n",
    "# region_start_hg38 = 207318058\n",
    "# region_end_hg38 = 207324558\n",
    "\n",
    "# # TSS position (center of the region, or from annotation)\n",
    "# cd55_tss_hg38 = 207321308  # â† Use actual hg38 TSS here\n",
    "# tss_t2t  = 206586604  # â† Known T2T TSS from earlier\n",
    "\n",
    "# # Offset between hg38 and T2T\n",
    "# t2t_offset = -734704  # tss_t2t = cd55_tss_hg38 + offset\n",
    "\n",
    "# # --- STEP 3: Scan sequence for CpG dinucleotides ---\n",
    "# cpg_records = []\n",
    "# for i in range(len(genomic_sequence) - 1):\n",
    "#     if genomic_sequence[i:i+2] == \"CG\":\n",
    "#         start_hg38 = region_start_hg38 + i\n",
    "#         end_hg38 = start_hg38 + 2\n",
    "#         start_t2t = start_hg38 + t2t_offset\n",
    "#         end_t2t = end_hg38 + t2t_offset\n",
    "#         rel_to_tss_hg38 = start_hg38 - cd55_tss_hg38\n",
    "#         rel_to_tss_t2t = start_t2t - tss_t2t\n",
    "\n",
    "#         cpg_records.append({\n",
    "#             \"chrom\": region_chr,\n",
    "#             \"start_hg38\": start_hg38,\n",
    "#             \"end_hg38\": end_hg38,\n",
    "#             \"start_T2T\": start_t2t,\n",
    "#             \"end_T2T\": end_t2t,\n",
    "#             \"rel_to_TSS_hg38\": rel_to_tss_hg38,\n",
    "#             \"rel_to_TSS_T2T\": rel_to_tss_t2t,\n",
    "#             \"name\": f\"CpG_{len(cpg_records) + 1}\",\n",
    "#             \"score\": 0,\n",
    "#             \"strand\": \".\"\n",
    "#         })\n",
    "\n",
    "# # --- STEP 4: Create DataFrame ---\n",
    "# df_cpgs = pd.DataFrame(cpg_records)\n",
    "\n",
    "# # --- STEP 5: Export if needed ---\n",
    "# df_cpgs.to_csv(\"CpGs_hg38_T2T_with_TSS_relative.bed\", sep=\"\\t\", index=False)\n",
    "\n",
    "# # --- Preview ---\n",
    "# print(df_cpgs.head())\n",
    "# print(f\"Total CpGs found: {len(df_cpgs)}\")\n",
    "\n",
    "# df_cpgs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32",
   "metadata": {},
   "source": [
    "# use the tools samtools python to get the real corresponding hg38 DNA sequence for our ROI\n",
    "chr1:207318058â€“207324558 (hg38)\n",
    "ðŸ§¬ Requirements\n",
    "Ensure you have:\n",
    "\n",
    "The hg38 reference FASTA file: hg38.fa\n",
    "\n",
    "The .fai index file: hg38.fa.fai\n",
    "\n",
    "The samtools CLI tool installed\n",
    "\n",
    "The Python subprocess module (built-in)\n",
    "\n",
    "âœ… Python Code Using samtools faidx\n",
    "python\n",
    "Copy code\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# import subprocess\n",
    "\n",
    "# # --- Step 1: Define your input ---\n",
    "# fasta_path = \"hg38.fa\"  # Path to your local hg38 reference genome\n",
    "# region = \"chr1:207318058-207324558\"\n",
    "\n",
    "# # --- Step 2: Use samtools to extract sequence ---\n",
    "# def fetch_sequence_with_samtools(fasta, region):\n",
    "#     try:\n",
    "#         result = subprocess.run(\n",
    "#             [\"samtools\", \"faidx\", fasta, region],\n",
    "#             check=True,\n",
    "#             stdout=subprocess.PIPE,\n",
    "#             stderr=subprocess.PIPE,\n",
    "#             text=True\n",
    "#         )\n",
    "#         # Clean FASTA output: remove header and join lines\n",
    "#         lines = result.stdout.strip().split(\"\\n\")\n",
    "#         sequence = ''.join(lines[1:]).upper()\n",
    "#         return sequence\n",
    "#     except subprocess.CalledProcessError as e:\n",
    "#         print(\"Error running samtools:\", e.stderr)\n",
    "#         return None\n",
    "\n",
    "# # --- Step 3: Fetch sequence ---\n",
    "# genomic_sequence = fetch_sequence_with_samtools(fasta_path, region)\n",
    "\n",
    "# # --- Step 4: Preview ---\n",
    "# if genomic_sequence:\n",
    "#     print(\"Extracted sequence length:\", len(genomic_sequence))\n",
    "#     print(\"First 100 bases:\", genomic_sequence[:100])\n",
    "# else:\n",
    "#     print(\"Failed to extract sequence.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cpgs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# # Step 2: Create a DataFrame\n",
    "# df_dmc = pd.DataFrame({\n",
    "#     \"CpG Index\": np.arange(1, len(dif_D6_mCthresh_995)+1),\n",
    "#     \"Î”mC\": dif_D6_mCthresh_995\n",
    "# })\n",
    "\n",
    "# Step 3: Define significance threshold\n",
    "# You can adjust the thresholds based on biological interest\n",
    "\n",
    "mC_signif_positive_thresh = 0.3\n",
    "mC_signif_negative_thresh = -0.1\n",
    "\n",
    "df_cpgs[\"Significant\"] = (df_cpgs[\"Î”mC\"] > mC_signif_positive_thresh) | (df_cpgs[\"Î”mC\"] < mC_signif_negative_thresh)\n",
    "\n",
    "# Step 4: Filter top differentially methylated CpGs\n",
    "top_dmc = df_cpgs[df_cpgs[\"Significant\"]].sort_values(by=\"Î”mC\", ascending=False)\n",
    "\n",
    "# Step 5 (optional): Save as CSV\n",
    "top_dmc.to_csv(\"Top_Differential_mCs.csv\", index=False)\n",
    "\n",
    "# Print or display\n",
    "# print(top_dmc)\n",
    "top_dmc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36",
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- STEP 1: Create BED DataFrame from annotated CpG table ---\n",
    "top_dmc_bed = pd.DataFrame({\n",
    "    \"chrom\": top_dmc[\"chrom\"],\n",
    "    \"start\": top_dmc[\"start_hg38\"],\n",
    "    \"end\": top_dmc[\"end_hg38\"],\n",
    "    \"name\": top_dmc[\"name\"],\n",
    "    \"score\": top_dmc[\"Î”mC\"],\n",
    "    \"strand\": top_dmc[\"strand\"]\n",
    "})\n",
    "\n",
    "# --- STEP 2: Sort by genomic start (optional but recommended) ---\n",
    "top_dmc_bed = top_dmc_bed.sort_values(by=[\"chrom\", \"start\"])\n",
    "\n",
    "# --- STEP 3: Save as tab-delimited BED file (no header) ---\n",
    "top_dmc_bed.to_csv(\"top_DMC_CD55_CpGs_UCSC_hg38.bed\", sep=\"\\t\", index=False, header=False)\n",
    "\n",
    "print(\"âœ… UCSC BED file saved as: CD55_CpGs_UCSC_hg38.bed\")\n",
    "\n",
    "top_dmc_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cpgs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cpgs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39",
   "metadata": {},
   "source": [
    "# Add ML coef scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(lasso_abs_coefs)\n",
    "plt.plot(lasso_abs_coefs_norm)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(shap_sumabs_coefs)\n",
    "# plt.plot(shap_sumabs_coefs_norm)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(shap_sumabs_coefs/20)\n",
    "plt.plot(lasso_abs_coefs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cpgs[\"lasso_abs_coefs\"] = lasso_abs_coefs\n",
    "df_cpgs[\"lasso_abs_coefs_norm\"] = lasso_abs_coefs_norm\n",
    "df_cpgs[\"shap_sumabs_coefs\"] = shap_sumabs_coefs\n",
    "\n",
    "df_cpgs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import datetime     \n",
    "today = datetime.date.today()\n",
    "print(\"Today's date:\", today)\n",
    "save_name=\"CD55_CpGs_Annotated_T2T_hg38_with_model_coefs_Tcells_day6_mCthresh0.995_\"+str(today)+\".csv\"\n",
    "\n",
    "df_cpgs.to_csv(save_name, index=True)\n",
    "save_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df_cpgs), len(df_cpgs)//10,  df_cpgs.start_hg38.values[::len(df_cpgs)//10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46",
   "metadata": {},
   "source": [
    "# Visualize CpG positions along the genomic region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Step 2: Define TSS (midpoint) and promoter boundaries\n",
    "# cd55_tss = region_start + (region_length // 2)  # Assuming TSS = center\n",
    "promoter_start_hg38 = cd55_tss_hg38 - 1000\n",
    "promoter_end_hg38 = cd55_tss_hg38 + 1000\n",
    "\n",
    "# Function to classify regions\n",
    "def classify_region(pos, promoter_start=promoter_start_hg38, promoter_end=promoter_end_hg38):\n",
    "    if promoter_start <= pos <= promoter_end:\n",
    "        return \"Promoter\"\n",
    "    elif pos < promoter_start:\n",
    "        return \"Upstream Flank\"\n",
    "    else:\n",
    "        return \"Gene Body / Downstream\"\n",
    "\n",
    "# # --- STEP 2: CREATE ANNOTATED DATAFRAME ---\n",
    "# df_dmc = pd.DataFrame({\n",
    "#     \"CpG Index\": cpg_indices,\n",
    "#     \"Î”mC\": dif_D6_mCthresh_995,\n",
    "#     \"Genomic_Position\": cpg_positions\n",
    "# })\n",
    "\n",
    "# df_dmc[\"Significant\"] = (df_dmc[\"Î”mC\"] > 0.3) | (df_dmc[\"Î”mC\"] < -0.1)\n",
    "# df_dmc[\"Region\"] = df_dmc[\"Genomic_Position\"].apply(classify_region)\n",
    "# df_dmc[\"Chromosome\"] = region_chr\n",
    "\n",
    "df_cpgs[\"Region\"] = df_cpgs[\"start_hg38\"].apply(classify_region) #, args=(promoter_start_hg38, promoter_end_hg38))\n",
    "df_cpgs[\"Chromosome\"] = region_chr\n",
    "\n",
    "# --- STEP 3: PLOT BARPLOT WITH CLEAN X-TICKS ---\n",
    "sns.set(style=\"whitegrid\")\n",
    "\n",
    "plt.figure(figsize=(16, 6))\n",
    "ax = sns.barplot(\n",
    "    data=df_cpgs,\n",
    "    x=\"start_hg38\",\n",
    "    y=\"Î”mC\",\n",
    "    hue=\"Region\",\n",
    "    dodge=False,\n",
    "    palette=\"Set2\"\n",
    ")\n",
    "\n",
    "# Clean x-axis: show only 10 genomic positions\n",
    "# tick_positions = np.linspace(0, len(df_cpgs) - 1, 10, dtype=int)\n",
    "# tick_positions = [207318092, 207319112, 207320882, 207321201, 207321360, 207321611,\n",
    "#         207321739, 207321797, 207321977, 207322217, 207323409]\n",
    "# tick_positions = df_cpgs.start_hg38.values[::len(df_cpgs)//10]\n",
    "tick_positions = np.linspace(0, len(df_cpgs) - 1, 10, dtype=int)\n",
    "tick_labels = df_cpgs.loc[tick_positions, \"start_hg38\"]\n",
    "ax.set_xticks(tick_positions)\n",
    "ax.set_xticklabels(tick_labels, rotation=45, ha='right') #, fontsize=1) #  ha='right',\n",
    "\n",
    "# Add reference line and labels\n",
    "plt.axhline(0, color='gray', linestyle='--')\n",
    "plt.title(\"Î”mC vs Genomic Position (hg38)\", fontsize=14)\n",
    "plt.xlabel(\"Genomic Position (chr1)\")\n",
    "plt.ylabel(\"Î”mC (CRISPRoff - Unedited)\")\n",
    "plt.legend(title=\"Genomic Region\", bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd55_tss_t2t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_coord_columtn = \"start_T2T\"\n",
    "\n",
    "# Step 2: Define TSS (midpoint) and promoter boundaries\n",
    "# cd55_tss = region_start + (region_length // 2)  # Assuming TSS = center\n",
    "promoter_start_t2t = cd55_tss_t2t - 1000\n",
    "promoter_end_t2t = cd55_tss_t2t + 1000\n",
    "\n",
    "# Function to classify regions\n",
    "def classify_region(pos, promoter_start, promoter_end):\n",
    "    if promoter_start <= pos <= promoter_end:\n",
    "        return \"Promoter\"\n",
    "    elif pos < promoter_start:\n",
    "        return \"Upstream Flank\"\n",
    "    else:\n",
    "        return \"Gene Body / Downstream\"\n",
    "\n",
    "# # --- STEP 2: CREATE ANNOTATED DATAFRAME ---\n",
    "# df_dmc = pd.DataFrame({\n",
    "#     \"CpG Index\": cpg_indices,\n",
    "#     \"Î”mC\": dif_D6_mCthresh_995,\n",
    "#     \"Genomic_Position\": cpg_positions\n",
    "# })\n",
    "\n",
    "# df_dmc[\"Significant\"] = (df_dmc[\"Î”mC\"] > 0.3) | (df_dmc[\"Î”mC\"] < -0.1)\n",
    "# df_dmc[\"Region\"] = df_dmc[\"Genomic_Position\"].apply(classify_region)\n",
    "# df_dmc[\"Chromosome\"] = region_chr\n",
    "\n",
    "df_cpgs[\"Region\"] = df_cpgs[start_coord_columtn].apply(classify_region, args=(promoter_start_t2t, promoter_end_t2t))\n",
    "df_cpgs[\"Chromosome\"] = region_chr\n",
    "\n",
    "# --- STEP 3: PLOT BARPLOT WITH CLEAN X-TICKS ---\n",
    "sns.set(style=\"whitegrid\")\n",
    "\n",
    "plt.figure(figsize=(16, 6))\n",
    "ax = sns.barplot(\n",
    "    data=df_cpgs,\n",
    "    x=start_coord_columtn,\n",
    "    y=\"Î”mC\",\n",
    "    hue=\"Region\",\n",
    "    dodge=False,\n",
    "    palette=\"Set2\"\n",
    ")\n",
    "\n",
    "# Clean x-axis: show only 10 genomic positions\n",
    "# tick_positions = np.linspace(0, len(df_cpgs) - 1, 10, dtype=int)\n",
    "# tick_positions = [207318092, 207319112, 207320882, 207321201, 207321360, 207321611,\n",
    "#         207321739, 207321797, 207321977, 207322217, 207323409]\n",
    "# tick_positions = df_cpgs.start_hg38.values[::len(df_cpgs)//10]\n",
    "tick_positions = np.linspace(0, len(df_cpgs) - 1, 10, dtype=int)\n",
    "tick_labels = df_cpgs.loc[tick_positions, start_coord_columtn]\n",
    "ax.set_xticks(tick_positions)\n",
    "ax.set_xticklabels(tick_labels, rotation=45, ha='right') #, fontsize=1) #  ha='right',\n",
    "\n",
    "# Add reference line and labels\n",
    "plt.axhline(0, color='gray', linestyle='--')\n",
    "plt.title(\"Î”mC vs Genomic Position (T2Tv2.0))\", fontsize=14)\n",
    "plt.xlabel(\"Genomic Position (chr1)\")\n",
    "plt.ylabel(\"Î”mC (CRISPRoff - Unedited)\")\n",
    "plt.legend(title=\"Genomic Region\", bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cpgs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# --- STEP 2: SCATTER PLOT ---\n",
    "\n",
    "plt.figure(figsize=(16, 6))\n",
    "sns.scatterplot(\n",
    "    data=df_cpgs,\n",
    "    x=\"CpG Index\",\n",
    "    y=\"Î”mC\",\n",
    "    hue=\"Region\",\n",
    "    style=\"Significant\",\n",
    "    palette=\"Set2\",\n",
    "    s=80,\n",
    "    edgecolor='black'\n",
    ")\n",
    "plt.axhline(0, color='gray', linestyle='--')\n",
    "plt.title(\"Î”mC by CpG Index (CRISPRoff - Unedited)\", fontsize=14)\n",
    "plt.xlabel(\"CpG Index\")\n",
    "plt.ylabel(\"Î”mC\")\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cpgs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cpgs.start_hg38.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# # Step 2: Create a DataFrame\n",
    "# df_dmc = pd.DataFrame({\n",
    "#     \"CpG Index\": np.arange(1, len(dif_D6_mCthresh_995)+1),\n",
    "#     \"Î”mC\": dif_D6_mCthresh_995\n",
    "# })\n",
    "\n",
    "# # Step 3: Define significance threshold\n",
    "# # You can adjust the thresholds based on biological interest\n",
    "# df_dmc[\"Significant\"] = (df_dmc[\"Î”mC\"] > 0.3) | (df_dmc[\"Î”mC\"] < -0.1)\n",
    "\n",
    "# # Step 4: Filter top differentially methylated CpGs\n",
    "# top_dmc = df_dmc[df_dmc[\"Significant\"]].sort_values(by=\"Î”mC\", ascending=False)\n",
    "\n",
    "# # Step 5 (optional): Save as CSV\n",
    "# top_dmc.to_csv(\"Top_Differential_mCs.csv\", index=False)\n",
    "\n",
    "# # Print or display\n",
    "# print(top_dmc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Î”mC array (replace with your actual data if loading from file)\n",
    "# dif_D6_mCthresh_995 = np.array([...])  # Your full Î”mC array goes here\n",
    "\n",
    "# Metadata from user\n",
    "region_chr = \"chr1\"\n",
    "region_start = 206583354\n",
    "region_end = 206589854\n",
    "region_length = region_end - region_start\n",
    "cpg_count = len(dif_D6_mCthresh_995)\n",
    "\n",
    "# Step 1: Generate CpG index and position\n",
    "cpg_indices = np.arange(1, cpg_count + 1)\n",
    "cpg_positions = np.linspace(region_start, region_end, cpg_count, dtype=int)\n",
    "\n",
    "# Step 2: Define TSS (midpoint) and promoter boundaries\n",
    "cd55_tss = region_start + (region_length // 2)  # Assuming TSS = center\n",
    "promoter_start = cd55_tss - 1000\n",
    "promoter_end = cd55_tss + 1000\n",
    "\n",
    "# Step 3: Classify region\n",
    "def classify_region(pos):\n",
    "    if promoter_start <= pos <= promoter_end:\n",
    "        return \"Promoter\"\n",
    "    elif pos < promoter_start:\n",
    "        return \"Upstream Flank\"\n",
    "    else:\n",
    "        return \"Gene Body / Downstream\"\n",
    "\n",
    "# Step 4: Create DataFrame\n",
    "df_dmc = pd.DataFrame({\n",
    "    \"CpG Index\": cpg_indices,\n",
    "    \"Î”mC\": dif_D6_mCthresh_995,\n",
    "    \"Genomic Position (T2T v2.0)\": cpg_positions\n",
    "})\n",
    "\n",
    "# Step 5: Annotate significant Î”mC\n",
    "df_dmc[\"Significant\"] = (df_dmc[\"Î”mC\"] > 0.3) | (df_dmc[\"Î”mC\"] < -0.1)\n",
    "\n",
    "# Step 6: Annotate region type\n",
    "df_dmc[\"Region\"] = df_dmc[\"Genomic Position (T2T v2.0)\"].apply(classify_region)\n",
    "\n",
    "# Step 7: Add chromosome column\n",
    "df_dmc[\"Chromosome\"] = region_chr\n",
    "\n",
    "# Step 8: Rearrange columns\n",
    "df_dmc = df_dmc[[\n",
    "    \"Chromosome\", \"CpG Index\", \"Genomic Position (T2T v2.0)\",\n",
    "    \"Î”mC\", \"Significant\", \"Region\"\n",
    "]]\n",
    "\n",
    "# Optional: save to CSV\n",
    "df_dmc.to_csv(\"Annotated_Delta_mC_T2T.csv\", index=False)\n",
    "\n",
    "# Print preview\n",
    "print(df_dmc.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dmc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# --- STEP 1: DATA PREP ---\n",
    "\n",
    "# # Paste your actual Î”mC array here\n",
    "# dif_D6_mCthresh_995 = np.array([\n",
    "#     # paste all 142 values here\n",
    "# ])\n",
    "\n",
    "# # Metadata\n",
    "# region_chr = \"chr1\"\n",
    "# region_start = 206583354\n",
    "# region_end = 206589854\n",
    "# region_length = region_end - region_start\n",
    "# cpg_count = len(dif_D6_mCthresh_995)\n",
    "\n",
    "# Generate CpG indices and positions\n",
    "cpg_indices = np.arange(1, cpg_count + 1)\n",
    "cpg_positions = np.linspace(region_start, region_end, cpg_count, dtype=int)\n",
    "\n",
    "# Define TSS (center) and promoter boundaries\n",
    "cd55_tss = region_start + (region_length // 2)\n",
    "promoter_start = cd55_tss - 1000\n",
    "promoter_end = cd55_tss + 1000\n",
    "\n",
    "# Classify region\n",
    "def classify_region(pos):\n",
    "    if promoter_start <= pos <= promoter_end:\n",
    "        return \"Promoter\"\n",
    "    elif pos < promoter_start:\n",
    "        return \"Upstream Flank\"\n",
    "    else:\n",
    "        return \"Gene Body / Downstream\"\n",
    "\n",
    "# Create dataframe\n",
    "df_dmc = pd.DataFrame({\n",
    "    \"CpG Index\": cpg_indices,\n",
    "    \"Î”mC\": dif_D6_mCthresh_995,\n",
    "    \"Genomic Position (T2T v2.0)\": cpg_positions\n",
    "})\n",
    "\n",
    "df_dmc[\"Significant\"] = (df_dmc[\"Î”mC\"] > 0.3) | (df_dmc[\"Î”mC\"] < -0.1)\n",
    "df_dmc[\"Region\"] = df_dmc[\"Genomic Position (T2T v2.0)\"].apply(classify_region)\n",
    "df_dmc[\"Chromosome\"] = region_chr\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# --- STEP 2: SCATTER PLOT ---\n",
    "\n",
    "plt.figure(figsize=(16, 6))\n",
    "sns.scatterplot(\n",
    "    data=df_dmc,\n",
    "    x=\"CpG Index\",\n",
    "    y=\"Î”mC\",\n",
    "    hue=\"Region\",\n",
    "    style=\"Significant\",\n",
    "    palette=\"Set2\",\n",
    "    s=80,\n",
    "    edgecolor='black'\n",
    ")\n",
    "plt.axhline(0, color='gray', linestyle='--')\n",
    "plt.title(\"Î”mC by CpG Index (CRISPRoff - Unedited)\", fontsize=14)\n",
    "plt.xlabel(\"CpG Index\")\n",
    "plt.ylabel(\"Î”mC\")\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "# --- STEP 3: HEATMAP ---\n",
    "\n",
    "# Reshape data for heatmap (sorted by region for clarity)\n",
    "df_sorted = df_dmc.sort_values(by=[\"Region\", \"CpG Index\"])\n",
    "heatmap_data = df_sorted.pivot_table(index=\"CpG Index\", values=\"Î”mC\")\n",
    "\n",
    "plt.figure(figsize=(12, 10))\n",
    "sns.heatmap(\n",
    "    heatmap_data,\n",
    "    cmap=\"coolwarm\",\n",
    "    center=0,\n",
    "    cbar_kws={'label': 'Î”mC'}\n",
    ")\n",
    "plt.title(\"Î”mC Heatmap by CpG Site\")\n",
    "plt.ylabel(\"CpG Index (Sorted by Region)\")\n",
    "plt.xlabel(\"Î”mC\")\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "# --- STEP 4: VOLCANO/BAR PLOT BY POSITION ---\n",
    "\n",
    "plt.figure(figsize=(16, 6))\n",
    "sns.barplot(\n",
    "    data=df_dmc,\n",
    "    x=\"Genomic Position (T2T v2.0)\",\n",
    "    y=\"Î”mC\",\n",
    "    hue=\"Region\",\n",
    "    dodge=False,\n",
    "    palette=\"Set2\"\n",
    ")\n",
    "plt.xticks(rotation=90)\n",
    "plt.axhline(0, color='gray', linestyle='--')\n",
    "plt.title(\"Î”mC vs Genomic Position (T2T v2.0)\")\n",
    "plt.xlabel(\"Genomic Position (chr1)\")\n",
    "plt.ylabel(\"Î”mC\")\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Function to classify regions\n",
    "def classify_region(pos):\n",
    "    if promoter_start <= pos <= promoter_end:\n",
    "        return \"Promoter\"\n",
    "    elif pos < promoter_start:\n",
    "        return \"Upstream Flank\"\n",
    "    else:\n",
    "        return \"Gene Body / Downstream\"\n",
    "\n",
    "# --- STEP 2: CREATE ANNOTATED DATAFRAME ---\n",
    "df_dmc = pd.DataFrame({\n",
    "    \"CpG Index\": cpg_indices,\n",
    "    \"Î”mC\": dif_D6_mCthresh_995,\n",
    "    \"Genomic_Position\": cpg_positions\n",
    "})\n",
    "\n",
    "df_dmc[\"Significant\"] = (df_dmc[\"Î”mC\"] > 0.3) | (df_dmc[\"Î”mC\"] < -0.1)\n",
    "df_dmc[\"Region\"] = df_dmc[\"Genomic_Position\"].apply(classify_region)\n",
    "df_dmc[\"Chromosome\"] = region_chr\n",
    "\n",
    "# --- STEP 3: PLOT BARPLOT WITH CLEAN X-TICKS ---\n",
    "sns.set(style=\"whitegrid\")\n",
    "\n",
    "plt.figure(figsize=(16, 6))\n",
    "ax = sns.barplot(\n",
    "    data=df_dmc,\n",
    "    x=\"Genomic_Position\",\n",
    "    y=\"Î”mC\",\n",
    "    hue=\"Region\",\n",
    "    dodge=False,\n",
    "    palette=\"Set2\"\n",
    ")\n",
    "\n",
    "# Clean x-axis: show only 10 genomic positions\n",
    "tick_positions = np.linspace(0, len(df_dmc) - 1, 10, dtype=int)\n",
    "tick_labels = df_dmc.loc[tick_positions, \"Genomic_Position\"]\n",
    "ax.set_xticks(tick_positions)\n",
    "ax.set_xticklabels(tick_labels, rotation=45, ha='right')\n",
    "\n",
    "# Add reference line and labels\n",
    "plt.axhline(0, color='gray', linestyle='--')\n",
    "plt.title(\"Î”mC vs Genomic Position (T2T v2.0)\", fontsize=14)\n",
    "plt.xlabel(\"Genomic Position (chr1)\")\n",
    "plt.ylabel(\"Î”mC (CRISPRoff - Unedited)\")\n",
    "plt.legend(title=\"Genomic Region\", bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.savefig(\"delta_mC_vs_genomic_position_T2T.png\", dpi=300)\n",
    "# or for vector graphics:\n",
    "plt.savefig(\"delta_mC_vs_genomic_position_T2T.pdf\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63",
   "metadata": {},
   "source": [
    "# Add Hg38 coords!\n",
    "\n",
    "ðŸ§  Why This Works\n",
    "\n",
    "This assumes:\n",
    "\n",
    "No rearrangements or insertions in the corresponding T2T vs hg38 region\n",
    "\n",
    "The coordinates align linearly (which is true for many protein-coding loci like CD55)\n",
    "\n",
    "If you later need precision mapping with alignment files, you can use liftOver with chain files between T2T and hg38."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64",
   "metadata": {},
   "outputs": [],
   "source": [
    "hg38_start     = 207318058  \n",
    "t2t_start      = 206583354  \n",
    "genomic_offset = hg38_start - t2t_start #= +734704\n",
    "print(\"Genomic offset (hg38 - T2T v2.0): \", genomic_offset)\n",
    "\n",
    "# CpG_hg38 = CpG_T2T + 734,704\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1: Define offset\n",
    "hg38_offset = 734704  # Computed as hg38_start - t2t_start\n",
    "\n",
    "# Step 2: Add new column with hg38 positions\n",
    "df_dmc[\"Genomic_Position_hg38\"] = df_dmc[\"Genomic_Position\"] + hg38_offset\n",
    "\n",
    "# Step 3 (Optional): Reorder columns for clarity\n",
    "df_dmc = df_dmc[[\n",
    "    \"CpG Index\", \"Genomic_Position\", \"Genomic_Position_hg38\",\n",
    "    \"Î”mC\", \"Significant\", \"Region\"\n",
    "]]\n",
    "\n",
    "# Step 4: Save if needed\n",
    "df_dmc.to_csv(\"Annotated_CpGs_with_hg38.csv\", index=False)\n",
    "\n",
    "# Show preview\n",
    "print(df_dmc.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert to BED format (0-based start, 1-based end)\n",
    "df_bed = pd.DataFrame({\n",
    "    \"chrom\": \"chr1\",\n",
    "    \"start\": df_dmc[\"Genomic_Position_hg38\"],\n",
    "    \"end\": df_dmc[\"Genomic_Position_hg38\"] + 1,\n",
    "    \"name\": \"CpG_\" + df_dmc[\"CpG Index\"].astype(str),\n",
    "    \"score\": (df_dmc[\"Î”mC\"] * 1000).round(1),  # scale to 0-1000 range for UCSC\n",
    "    \"strand\": \"+\"\n",
    "})\n",
    "\n",
    "# Optional: sort by genomic position\n",
    "df_bed = df_bed.sort_values(by=\"start\")\n",
    "\n",
    "# Save as BED\n",
    "df_bed.to_csv(\"cd55_dmc_cpgs.hg38.bed\", sep=\"\\t\", index=False, header=False)\n",
    "\n",
    "\n",
    "df_bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67",
   "metadata": {},
   "source": [
    "You're seeing that error because UCSC Genome Browser doesn't accept .csv files directly â€” it expects standard genome formats like:\n",
    "\n",
    ".bed, .wig, .bigWig, .bigBed, .bam, etc.\n",
    "\n",
    "âœ… Solution\n",
    "\n",
    "You need to:\n",
    "\n",
    "1. âœ… Export the file in BED format\n",
    "\n",
    "Youâ€™ve already created a CSV (Annotated_CpGs_with_hg38.csv), but for genome browser upload, you must use BED format, which has:\n",
    "\n",
    "chrom    start    end    name    score    strand\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert annotated CpG data to BED format for UCSC Genome Browser\n",
    "\n",
    "# Step 1: Calculate BED coordinates (0-based start, 1-based end)\n",
    "df_bed = pd.DataFrame({\n",
    "    \"chrom\": \"chr1\",\n",
    "    \"start\": df_dmc[\"Genomic_Position_hg38\"],\n",
    "    \"end\": df_dmc[\"Genomic_Position_hg38\"] + 1,  # 1 bp span\n",
    "    \"name\": \"CpG_\" + df_dmc[\"CpG Index\"].astype(str),\n",
    "    \"score\": (df_dmc[\"Î”mC\"] * 1000).round(1),  # scaled to UCSC score range\n",
    "    \"strand\": \"+\"\n",
    "})\n",
    "\n",
    "# Step 2: Optional sorting\n",
    "df_bed = df_bed.sort_values(by=\"start\")\n",
    "\n",
    "# Step 3: Save as tab-delimited BED file (NO HEADER)\n",
    "df_bed.to_csv(\"cd55_dmc_cpgs.hg38.bed\", sep=\"\\t\", index=False, header=False)\n",
    "\n",
    "df_bed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create UCSC-compatible BED from annotated Î”mC table\n",
    "df_bed = pd.DataFrame({\n",
    "    \"chrom\": \"chr1\",\n",
    "    \"start\": df_dmc[\"Genomic_Position_hg38\"],\n",
    "    \"end\": df_dmc[\"Genomic_Position_hg38\"] + 1,\n",
    "    \"name\": \"CpG_\" + df_dmc[\"CpG Index\"].astype(str),\n",
    "    \"score\": (df_dmc[\"Î”mC\"] * 1000).round(0).astype(int),  # UCSC expects int scores from 0â€“1000\n",
    "    \"strand\": \"+\"\n",
    "})\n",
    "\n",
    "# Clip scores to UCSC-compliant range\n",
    "df_bed[\"score\"] = df_bed[\"score\"].clip(lower=0, upper=1000)\n",
    "\n",
    "# Save as plain BED file with no header\n",
    "df_bed.to_csv(\"cd55_dmc_cpgs_INTscores.hg38.bed\", sep=\"\\t\", index=False, header=False)\n",
    "df_bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70",
   "metadata": {},
   "source": [
    "âœ… UCSC Track Line (Required for Upload)\n",
    "\n",
    "When uploading this BED file to the UCSC Genome Browser, you'll need to add a \"track\" line manually in the custom tracks section:\n",
    "\n",
    "track name=\"CD55 Î”mC CpGs\" description=\"CRISPRoff vs Unedited T cells\" visibility=2 itemRgb=\"On\" useScore=1\n",
    "\n",
    "\n",
    "Then paste or upload your cd55_dmc_cpgs.hg38.bed file content below that line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dimelo_v2_modkit_parsing",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
