{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7418580e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current Date and Time: 2025-11-09 19:19:54\n"
     ]
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "\n",
    "def current_time():\n",
    "    \"\"\"Returns the current date and time as a formatted string.\"\"\"\n",
    "    return datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\") \n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    print(\"Current Date and Time:\", current_time())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1424e6d9",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (1430218669.py, line 1)",
     "output_type": "error",
     "traceback": [
      "  \u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[2]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[31m    \u001b[39m\u001b[31mInstall rust\u001b[39m\n            ^\n\u001b[31mSyntaxError\u001b[39m\u001b[31m:\u001b[39m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "Install rust\n",
    "\n",
    "curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh\n",
    "\n",
    "\n",
    "This is usually done by running one of the following (note the leading DOT):\n",
    ". \"$HOME/.cargo/env\"            # For sh/bash/zsh/ash/dash/pdksh\n",
    "# source \"$HOME/.cargo/env.fish\"  # For fish\n",
    "# source $\"($nu.home-path)/.cargo/env.nu\"  # For nushell\n",
    "\n",
    "rustup update"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2944ccdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "rustup show\n",
    "rustup toolchain list\n",
    "which rustc\n",
    "rustup which rustc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "580d8d78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Показать текущую версию\n",
    "# rustc --version\n",
    "\n",
    "# # Обновить до актуального stable\n",
    "# rustup update stable\n",
    "\n",
    "# # ИЛИ поставить конкретную версию, которая удовлетворяет (например 1.88.1)\n",
    "# rustup toolchain install 1.88.1\n",
    "# rustup default 1.88.1   # или: rustup override set 1.88.1 в каталоге проекта\n",
    "# rustc --version\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44472318",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd ~/software/modkit\n",
    "rustup override set stable     # или: rustup override set 1.91.0\n",
    "rustc --version                # должно стать 1.91.0\n",
    "cargo install --git https://github.com/nanoporetech/modkit.git --rev 8fa79e3e\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0475d6fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /home/michalula/software\n",
    "mkdir modkit\n",
    "ls -la modkit\n",
    "git clone https://github.com/nanoporetech/modkit.git\n",
    "cd modkit\n",
    "cargo install --path .\n",
    "# or\n",
    "cargo install --git https://github.com/nanoporetech/modkit.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f728c050",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: line 1: rustc: command not found\n"
     ]
    }
   ],
   "source": [
    "! rustc --version\n",
    "# rustc 1.91.0 (f8297e351 2025-10-28)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c746f461",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: line 1: modkit: command not found\n"
     ]
    }
   ],
   "source": [
    "! modkit --version "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a3f2d6e",
   "metadata": {},
   "source": [
    "## Old modkit --version : === mod_kit 0.2.4\n",
    "mod_kit 0.2.4\n",
    "\n",
    "## NEW modkit --version : 0.5.1\n",
    "modkit 0.5.1\n",
    "(modkit_new) michalula@Skynet:~/software/modkit$ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8e29222",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8d4884ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: line 1: modkit: command not found\n"
     ]
    }
   ],
   "source": [
    "! modkit --version "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20251105_modkit_updated_kernel.ipynb  dmr.log  modkit_dmr.ipynb  pileups\n",
      "dif_mC_day6\t\t\t      log.txt  modkit_output\n"
     ]
    }
   ],
   "source": [
    "! ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "39b9a895",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kernel PATH: /home/michalula/miniconda3/envs/modkit_new/bin:/home/michalula/.vscode-server/cli/servers/Stable-7d842fb85a0275a4a8e4d7e040d2625abbf7f084/server/bin/remote-cli:/home/michalula/.local/bin:/home/michalula/miniconda3/envs/modkit_new/bin:/home/michalula/miniconda3/condabin:/home/michalula/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin\n",
      "modkit in PATH? None\n",
      "Modkit is a bioinformatics tool for working with modified bases from Oxford\n",
      "Nanopore\n",
      "\n",
      "\u001b[1m\u001b[4mUsage:\u001b[0m \u001b[1mmodkit\u001b[0m <COMMAND>\n",
      "\n",
      "\u001b[1m\u001b[4mCommands:\u001b[0m\n",
      "  \u001b[1mpileup\u001b[0m          Tabulates base modification calls across genomic positions.\n",
      "                  This command produces a bedMethyl formatted file. Schema and\n",
      "                  description of fields can be found in the README\n",
      "  \u001b[1madjust-mods\u001b[0m     Performs various operations on BAM files containing base\n",
      "                  modification information, such as converting base modification\n",
      "                  codes and ignoring modification calls. Produces a BAM output\n",
      "                  file\n",
      "  \u001b[1mupdate-tags\u001b[0m     Renames Mm/Ml to tags to MM/ML. Also allows changing the mode\n",
      "                  flag from silent '.' to explicitly '?' or '.'\n",
      "  \u001b[1msample-probs\u001b[0m    Calculate an estimate of the base modification probability\n",
      "                  distribution\n",
      "  \u001b[1msummary\u001b[0m         Summarize the mod tags present in a BAM and get basic\n",
      "                  statistics. The default output is a totals table (designated\n",
      "                  by '#' lines) and a modification calls table. Descriptions of\n",
      "                  the columns can be found in the README\n",
      "  \u001b[1mcall-mods\u001b[0m       Call mods from a modbam, creates a new modbam with\n",
      "                  probabilities set to 100% if a base modification is called or\n",
      "                  0% if called canonical\n",
      "  \u001b[1mextract\u001b[0m         Extract read-level base modification information from a modBAM\n",
      "                  into a tab-separated values table\n",
      "  \u001b[1mrepair\u001b[0m          Repair MM and ML tags in one bam with the correct tags from\n",
      "                  another. To use this command, both modBAMs _must_ be sorted by\n",
      "                  read name. The \"donor\" modBAM's reads must be a superset of\n",
      "                  the acceptor's reads. Extra reads in the donor are allowed,\n",
      "                  and multiple reads with the same name (secondary, etc.) are\n",
      "                  allowed in the acceptor. Reads with an empty SEQ field cannot\n",
      "                  be repaired and will be rejected. Reads where there is an\n",
      "                  ambiguous alignment of the acceptor to the donor will be\n",
      "                  rejected (and logged). See the full documentation for details\n",
      "  \u001b[1mdmr\u001b[0m             Perform DMR test on a set of regions. Output a BED file of\n",
      "                  regions with the score column indicating the magnitude of the\n",
      "                  difference. Find the schema and description of fields can in\n",
      "                  the README as well as a description of the model and method.\n",
      "                  See subcommand help for additional details\n",
      "  \u001b[1mpileup-hemi\u001b[0m     Tabulates double-stranded base modification patters (such as\n",
      "                  hemi-methylation) across genomic motif positions. This command\n",
      "                  produces a bedMethyl file, the schema can be found in the\n",
      "                  online documentation\n",
      "  \u001b[1mvalidate\u001b[0m        Validate results from a set of mod-BAM files and associated\n",
      "                  BED files containing the ground truth modified base status at\n",
      "                  reference positions\n",
      "  \u001b[1mmotif\u001b[0m           Various commands to search for, evaluate, or further regine\n",
      "                  sequence motifs enriched for base modification. Also can\n",
      "                  generate BED files of motif locations\n",
      "  \u001b[1mentropy\u001b[0m         Use a mod-BAM to calculate methylation entropy over genomic\n",
      "                  windows\n",
      "  \u001b[1mlocalize\u001b[0m        Investigate patterns of base modifications, by aggregating\n",
      "                  pileup counts \"localized\" around genomic features of interest\n",
      "  \u001b[1mstats\u001b[0m           Calculate base modification levels over regions\n",
      "  \u001b[1mbedmethyl\u001b[0m       Utilities to work with bedMethyl files\n",
      "  \u001b[1mmodbam\u001b[0m          Utilities to work with modBAM files\n",
      "  \u001b[1mopen-chromatin\u001b[0m  Identify regions of open chromatin based on exogenous 6mA\n",
      "                  signal\n",
      "  \u001b[1mhelp\u001b[0m            Print this message or the help of the given subcommand(s)\n",
      "\n",
      "\u001b[1m\u001b[4mOptions:\u001b[0m\n",
      "  \u001b[1m-h\u001b[0m, \u001b[1m--help\u001b[0m     Print help\n",
      "  \u001b[1m-V\u001b[0m, \u001b[1m--version\u001b[0m  Print version\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['/home/michalula/.cargo/bin/modkit', '--help'], returncode=0)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os, shutil, subprocess, sys\n",
    "print(\"Kernel PATH:\", os.environ.get(\"PATH\"))\n",
    "print(\"modkit in PATH?\", shutil.which(\"modkit\"))\n",
    "# Try calling via absolute path if needed:\n",
    "subprocess.run([\"/home/michalula/.cargo/bin/modkit\", \"--help\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d66e9cd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.environ[\"PATH\"] = \"/home/michalula/.cargo/bin:\" + os.environ[\"PATH\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9ee9f909",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "modkit 0.5.1\n"
     ]
    }
   ],
   "source": [
    "!modkit --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "962b7722",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "modkit_new",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
