{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "63928453",
   "metadata": {},
   "source": [
    "Based on:\n",
    "https://nanoporetech.github.io/modkit/intro_dmr.html#perform-differential-methylation-scoring"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e42a1933",
   "metadata": {},
   "source": [
    "Select kernal: dimelo_v2_modkit_parsing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d52a56c",
   "metadata": {},
   "source": [
    "Preparing the input data\n",
    "The inputs to all modkit dmr commands are two or more bedMethyl files (created by modkit pileup) that have been compressed with bgzip and indexed with tabix. An example of how to generate the input data is shown below:\n",
    "\n",
    "\n",
    "ref=grch38.fasta\n",
    "threads=32\n",
    "\n",
    "norm=normal_sample.bam\n",
    "norm_pileup=normal_pileup.bed\n",
    "\n",
    "modkit pileup ${norm} ${norm_pileup} \\\n",
    "  --cpg \\\n",
    "  --ref ${ref} \\\n",
    "  --threads ${threads} \\\n",
    "  --log-filepath log.txt\n",
    "\n",
    "bgzip -k ${norm_pileup}\n",
    "tabix -p bed ${norm_pileup}.gz\n",
    "\n",
    "# pileup and compression can also be done in one step\n",
    "tumor=tumor_sample.bam\n",
    "tumor_pileup=tumor_pileup.bed.gz\n",
    "\n",
    "modkit pileup ${tumor} - \\\n",
    "  --cpg \\\n",
    "  --ref ${ref} \\\n",
    "  --threads ${threads} \\\n",
    "  --log-filepath log.txt | ${bgzip} -c > ${tumor_pileup}\n",
    "\n",
    "tabix -p bed ${tumor_pileup}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e80d3d6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "db586195",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hello\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "echo \"hello\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b72398da",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c716521e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "\n",
    "# ref=\"/home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta\"\n",
    "# threads=32\n",
    "\n",
    "# CROFF_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/sort_merge_sort_align_t2t_v1_1_trim_20250721_nCATs_Tcells_CROFF_Day28_minion_run2_day8.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "# pileup_CROFF_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/pileup_sort_merge_sort_align_t2t_v1_1_trim_20250721_nCATs_Tcells_CROFF_Day28_minion_run2_day8.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "\n",
    "# # norm=normal_sample.bam\n",
    "# # norm_pileup=normal_pileup.bed\n",
    "\n",
    "# modkit pileup ${CROFF_day28_bam} ${pileup_CROFF_day28_bam} \\\n",
    "#   --cpg \\\n",
    "#   --ref ${ref} \\\n",
    "#   --threads ${threads} \\\n",
    "#   --log-filepath log.txt\n",
    "\n",
    "# bgzip -k ${pileup_CROFF_day28_bam}\n",
    "# tabix -p bed ${pileup_CROFF_day28_bam}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b6312661",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chm13v2.0.fa\t  chm13v2.0.fa.fai\t\t   haplotype_vcf\n",
      "chm13v2.0.fa.amb  chm13v2.0.fa.pac\t\t   up_chm13v2.0.fasta\n",
      "chm13v2.0.fa.ann  chm13v2.0.fa.sa\t\t   up_chm13v2.0.fasta.fai\n",
      "chm13v2.0.fa.bwt  convert_to_uppercase_fasta.bash\n"
     ]
    }
   ],
   "source": [
    "! ls /home/michalula/data/ref_genomes/t2t_v2_0/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a539faad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "align_t2t_v2_0_trim_20250908_Day35_CROFF_Tcells_2Libraries_Minion_R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\n",
      "sort_align_t2t_v2_0_trim_20250908_Day35_CROFF_Tcells_2Libraries_Minion_R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\n",
      "sort_align_t2t_v2_0_trim_20250908_Day35_CROFF_Tcells_2Libraries_Minion_R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam.bai\n",
      "summary_sort_align_t2t_v2_0_trim_20250908_Day35_CROFF_Tcells_2Libraries_Minion_R9.dna_r9.4.1_e8_sup@v3.3.5mCG.tsv\n"
     ]
    }
   ],
   "source": [
    "! ls /home/michalula/data/cas9_nanopore/data/20250908_nCATs_T_CRoff_Day_35/5mCG/to_t2t_v2_0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b64671bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "> calculated chunk size: 48, interval size 100000, processing 4800000 positions concurrently\n",
      "> filtering to only CpG motifs\n",
      "> attempting to sample 10042 reads\n",
      "> Using filter threshold 0.7597656 for C.\n",
      "> Done, processed 11762972 rows. Processed ~129977 reads and skipped ~150 reads.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "ref_genome_fa=\"/home/michalula/data/ref_genomes/t2t_v2_0/chm13v2.0.fa\"\n",
    "# ref= \"/home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta\"\n",
    "\n",
    "threads=32\n",
    "\n",
    "date_today=\"20251109\"\n",
    "data_folder_path=\"/home/michalula/data/cas9_nanopore/data/20250908_nCATs_T_CRoff_Day_35/5mCG/to_t2t_v2_0/\"\n",
    "CROFF_day35_bam=${data_folder_path}\"sort_align_t2t_v2_0_trim_20250908_Day35_CROFF_Tcells_2Libraries_Minion_R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "pileup_CROFF_day35_bam=${data_folder_path}{date_today}\"_pileup_CROFF_Day35_Tcells.bam\"\n",
    "# \"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/pileup_sort_merge_sort_align_t2t_v1_1_trim_20250721_nCATs_Tcells_CROFF_Day28_minion_run2_day8.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "# norm=normal_sample.bam\n",
    "# norm_pileup=normal_pileup.bed\n",
    "\n",
    "modkit pileup ${CROFF_day35_bam} ${pileup_CROFF_day35_bam} \\\n",
    "  --cpg \\\n",
    "  --ref ${ref_genome_fa} \\\n",
    "  --threads ${threads} \\\n",
    "  --log-filepath log.txt\n",
    "\n",
    "bgzip -k ${pileup_CROFF_day35_bam}\n",
    "tabix -p bed ${pileup_CROFF_day35_bam}.gz\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2ec9525",
   "metadata": {},
   "source": [
    "<!>\n",
    "> Threshold of  0.7597656 for base C is low. Consider increasing the filter-percentile or specifying a higher threshold.\n",
    "> Done, processed 11762972 rows. Processed ~129977 reads and skipped ~150 reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da598f32",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/pileup’: File exists\n",
      "> calculated chunk size: 48, interval size 100000, processing 4800000 positions concurrently\n",
      "> filtering to only CpG motifs\n",
      "> attempting to sample 10042 reads\n",
      "> failed to get modbase info for record 02866349-9f01-4302-9a1c-708042ab1756, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 81342b67-3ffb-432f-967c-cbd4f5f6fd23, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record c6a4eb69-74f6-4cc7-a0b2-a818bef21c3a, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record d64f46b8-658e-44ca-a77d-f3ad558b0a1e, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record f5ec4f2e-355c-4e07-a439-71b74a60109f, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 04d90c73-5027-40e1-aa06-7884fdbe50d4, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 2b00e0c6-0f10-4529-92d2-8a78677e500b, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 18598aef-c5f2-4b4d-ad1b-9548fdece5e8, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 90143ed1-de6f-467e-bce7-806c4533892e, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 43cd2c92-5991-4818-88b8-d2e508b5fb73, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 85e799ec-6159-4bb9-bc9f-0b8122e6c3be, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 431b2940-8de7-4d77-b215-4138fb593f88, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 50ddaf48-6bd4-4329-a28b-cfd5ba1df7e4, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record d93875b8-870a-4a16-8f9f-4135776c0d68, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record c80fe139-6ae4-45b6-ba2b-2af6d61f71e3, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record bb595193-45fa-4662-b4ea-c0f63f1479ed, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 0adaa275-d4af-461b-acb7-83b925347930, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 38a9c91a-c998-4d97-9b54-cb4f5954ded1, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record fba67fbb-47dc-4a3d-a415-f10b69c698a6, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 03c6bad0-2a0b-4886-b9b8-5ad97bb6763e, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 4449cc2e-918c-45b0-ac12-b58880cf4ccd, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 94e72f1c-cdb1-4563-9f31-7860145fd0b5, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 954ce772-7e0b-4ec5-bf8f-92a7389ee498, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record e02e1bf1-a9bf-42c2-b0a7-f04b5b4d125f, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record ad068dad-ce6c-407f-8f85-5c201a7097c3, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 9450f590-1b57-4cf6-a116-de1cc55b778a, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 5806cff7-42fc-40bc-919e-79c864135f7d, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record ba15c667-48f1-4cef-b140-5cdbd8e62478, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 2a90b1e8-7482-428d-8f0a-c0d22f09fc54, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record c98c8415-93d7-484b-a1ed-4ad29dd81334, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 7228de08-8527-440c-b13f-063d71b3c798, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 1dc7d29b-0cf5-4fa9-99ef-9f0af7e984d6, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record f74388b3-c8da-43c3-a1af-22bba1f5a6f5, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 709fa0ac-1363-422a-9171-acca957141d7, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 6d969ab2-7bc0-4c04-bd4f-6d09e63887e4, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record f51911d7-140e-4513-ba8a-47c72c910f9c, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 08c2a7de-fe9d-44cc-949b-383bb240b5f6, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record ff5f5011-cf61-4ed7-a742-539d2ccbf9fb, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record bbbda2c1-575a-4d63-87e6-de0b2df8893a, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 429265e1-5f50-4181-b3af-375a13427540, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 0842f7b7-89da-419a-a54f-027c471feb65, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 893e2f64-6e05-44f2-b284-bfe54e6713ee, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record e9df7ca8-dab5-4819-ad3f-e475fffc079a, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 6a6a312f-cfd1-4288-b3a2-9df130b4f64f, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record b0891369-3621-4672-a9ba-1ce2aba286a0, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record efab7f56-eed5-4077-b04e-1039bce8726d, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 7a0aa124-88ea-4052-bb3a-06df7789bcdb, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 2efe9a2d-0f61-42d8-99b4-ccfe58abe2e1, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 1ac5e8ce-ce3a-4723-9ed2-1b26baf4f900, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 38115b92-f112-4b01-ba13-9e85703b4ddd, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 61c84fc7-8419-401e-ab86-a4f5b09c6484, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 26e1d06d-264b-43e3-a434-c3520551b6a0, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 759a3305-a0c8-4bac-8005-998fd47f306c, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 8b941085-db9f-4179-ae5c-421022d319fb, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 481c466a-4451-4cec-b645-74345ad12651, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 76f63c00-94c0-4bb5-9edf-c324cccdf9a2, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 52b10b19-6248-4540-90a5-210a56efa802, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record dbf3e640-f05c-44cd-9253-7765dbd8bf78, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 33f6e8ad-674b-41ad-a606-cc78554759d5, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 507f5d39-40d1-44fc-9dac-9b14bb816d85, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record e75c479d-9bd1-41b9-a791-847592ed67d4, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 2d9db948-9dc6-4ac1-8dde-d42a08357f43, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record d31a4938-c00e-47b9-8b7f-7b2ae397b472, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 7f4a4a08-4395-4834-bd07-fcf8e677d80e, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 9b1170b5-a4c1-4c53-b79e-f4536149ea58, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record ebe878d6-ad63-40c7-a5cd-1dce14b6e732, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 2898c84f-8f03-43cb-8a89-c2415c379120, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record 9246a514-aa4b-4497-a398-a54c6f738a67, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> failed to get modbase info for record e77655e5-3b7a-4d99-aafc-90ab15a6c08e, Bad Input: invalid MM delta list, Parsing Error: Error { input: \"\", code: Digit }\n",
      "> Threshold of 0.6621094 for base C is low. Consider increasing the filter-percentile or specifying a higher threshold.\n",
      "> Done, processed 867290 rows. Processed ~12804 reads and skipped ~688 reads.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir ${data_folder_path}\"pileup\"\n",
    "# pileup_dir_path=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/pileup\"\n",
    "# mkdir ${pileup_dir_path}\n",
    "\n",
    "Unedit_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/sort_align_t2t_v2_0_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged_flowcells_output.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "# pileup_Unedit_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "# pileup_Unedit_day28_bam=${pileup_dir_path}\"/pileup_Unedit_Day28_Tcells_20250721.bam\"\n",
    "pileup_Unedit_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/pileup/pileup_Unedit_Day28_Tcells_20250721.bam\"\n",
    "\n",
    "# \"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam\"\n",
    "  \n",
    "ref_genome_fa=\"/home/michalula/data/ref_genomes/t2t_v2_0/chm13v2.0.fa\"\n",
    "# ref=\"/home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta\"\n",
    "\n",
    "threads=32\n",
    " \n",
    "\n",
    "modkit pileup ${Unedit_day28_bam} ${pileup_Unedit_day28_bam} \\\n",
    "  --cpg \\\n",
    "  --ref ${ref_genome_fa} \\\n",
    "  --threads ${threads} \\\n",
    "  --log-filepath log.txt\n",
    "\n",
    "bgzip -k ${pileup_Unedit_day28_bam}\n",
    "tabix -p bed ${pileup_Unedit_day28_bam}.gz\n",
    "\n",
    "# modkit pileup ${Unedit_day28_bam} ${pileup_Unedit_day28_bam} \\\n",
    "#   --cpg \\\n",
    "#   --ref ${ref} \\\n",
    "#   --threads ${threads} \\\n",
    "#   --log-filepath log.txt\n",
    "\n",
    "# bgzip -k ${pileup_Unedit_day28_bam}\n",
    "# tabix -p bed ${pileup_Unedit_day28_bam}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54e88472",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "\n",
    "# Unedit_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "# # pileup_Unedit_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "# pileup_Unedit_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam\"\n",
    "  \n",
    "# ref=\"/home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta\"\n",
    "# threads=32\n",
    " \n",
    "# modkit pileup ${Unedit_day28_bam} ${pileup_Unedit_day28_bam} \\\n",
    "#   --cpg \\\n",
    "#   --ref ${ref} \\\n",
    "#   --threads ${threads} \\\n",
    "#   --log-filepath log.txt\n",
    "\n",
    "# bgzip -k ${pileup_Unedit_day28_bam}\n",
    "# tabix -p bed ${pileup_Unedit_day28_bam}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdbf4487",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "\n",
    "# Unedit_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "# # pileup_Unedit_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "# pileup_Unedit_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam\"\n",
    "  \n",
    "# ref=\"/home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta\"\n",
    "# threads=32\n",
    " \n",
    "# modkit pileup ${Unedit_day28_bam} ${pileup_Unedit_day28_bam} \\\n",
    "#   --cpg \\\n",
    "#   --ref ${ref} \\\n",
    "#   --threads ${threads} \\\n",
    "#   --log-filepath log.txt\n",
    "\n",
    "# bgzip -k ${pileup_Unedit_day28_bam}\n",
    "# tabix -p bed ${pileup_Unedit_day28_bam}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18f50234",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "# # pileup and compression can also be done in one step\n",
    "\n",
    "# ref=\"/home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta\"\n",
    "# threads=32\n",
    "\n",
    "# Unedit_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "# pileup_Unedit_day28_bam_gz=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam.gz\"\n",
    " \n",
    "# # tumor=tumor_sample.bam\n",
    "# # tumor_pileup=tumor_pileup.bed.gz\n",
    "\n",
    "# # modkit pileup ${Unedit_day28_bam} - \\\n",
    "# #   --cpg \\\n",
    "# #   --ref ${ref} \\\n",
    "# #   --threads ${threads} \\\n",
    "# #   --log-filepath log.txt | ${bgzip} -c > ${pileup_Unedit_day28_bam}\n",
    "\n",
    "# # tabix -p bed ${pileup_Unedit_day28_bam}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c04c9a8",
   "metadata": {},
   "source": [
    "3. Detecting differential modification at single base positions\n",
    "The modkit dmr pair command has the ability to score individual bases (e.g. differentially methylated CpGs). To run single-base analysis on one or more paired samples, simply omit the --regions (-r) option when running modkit dmr pair. When performing single-base analysis the likelihood ratio score and a MAP-based p-value are available. For details on the likelihood ratio score and the MAP-based p-value, see the scoring details section. For example the above command becomes:\n",
    "\n",
    "dmr_result=single_base_haplotype_dmr.bed\n",
    "\n",
    "modkit dmr pair \\\n",
    "  -a ${hp1_pileup}.gz \\\n",
    "  -b ${hp2_pileup}.gz \\\n",
    "  -o ${dmr_result} \\\n",
    "  --ref ${ref} \\\n",
    "  --base C \\\n",
    "  --threads ${threads} \\\n",
    "  --log-filepath dmr.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3686ce95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged_flowcells_output.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\n",
      "t2t_v1_1\n",
      "to_t2t_v2_0\n",
      "trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged_flowcells_output.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\n",
      "trim_5mCG_5hmCG_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged_flowcells_output.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\n"
     ]
    }
   ],
   "source": [
    "! ls \"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d9ff37e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ls: cannot access '/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam.gz': No such file or directory\n"
     ]
    }
   ],
   "source": [
    "! ls /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da56ca74",
   "metadata": {},
   "outputs": [],
   "source": [
    "/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5af9df28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "drwxrwxr-x 2 michalula michalula 4096 Jul 31 00:43 /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs\n",
      "-rw-rw-r-- 1 michalula michalula 5758746 Jul 31 00:43 /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam.gz\n"
     ]
    }
   ],
   "source": [
    "    %%bash\n",
    "\n",
    "   ls -ld /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs\n",
    "   ls -l /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90fd3c37",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "   # chmod u+rw /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam.gz\n",
    "   # chmod u+rwx /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ef2395b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "   chmod u+rwx \"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/pileup/\"\n",
    "   chmod u+rwx \"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/5mCG/to_t2t_v2_0/pileup_CROFF_Day28_Tcells_20250721.bam.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "275b31f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "drwxrwxr-x 4 michalula michalula 4096 Aug  1 01:35 /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs\n",
      "-rwxrw-r-- 1 michalula michalula 5354303 Sep 24 12:10 /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/5mCG/to_t2t_v2_0/pileup_CROFF_Day28_Tcells_20250721.bam.gz\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "ls -ld /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs\n",
    "ls -ld \"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/5mCG/to_t2t_v2_0/pileup_CROFF_Day28_Tcells_20250721.bam.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b855f1cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ls /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9f020c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/pileup/pileup_Unedit_Day28_Tcells_20250721.bam.gz\n",
      "-rw-rw-r-- 1 michalula michalula 7514010 Sep 24 12:17 /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/pileup/pileup_Unedit_Day28_Tcells_20250721.bam.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "> loading sites from input 'a' bedMethyl\n",
      "> loaded 867290 sites\n",
      "> loading 48 sites at a time\n",
      "> 111264 regions processed successfully and 756026 regions failed\n"
     ]
    }
   ],
   "source": [
    " %%bash\n",
    " \n",
    " # 3. Detecting differential modification at single base positions\n",
    "# The modkit dmr pair command has the ability to score individual bases (e.g. differentially methylated CpGs). To run single-base analysis on one or more paired samples, simply omit the --regions (-r) option when running modkit dmr pair. When performing single-base analysis the likelihood ratio score and a MAP-based p-value are available. For details on the likelihood ratio score and the MAP-based p-value, see the scoring details section. For example the above command becomes:\n",
    "\n",
    "dmr_result=\"/home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/single_base_day28_CRoff_vs_Unedit_dmr_20251109.bed\"\n",
    "\n",
    "# pileup_Unedit_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "# removed the @ from the file name\n",
    "# pileup_Unedit_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam\"\n",
    "# pileup_Unedit_day28_bam=\"/home/michalula/code/epiCausality/epiCode/differential_methyl/pileups/t2t_v1_1/pileup_Unedit_Day28_Tcells_20250721.bam\"\n",
    "pileup_Unedit_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/pileup/pileup_Unedit_Day28_Tcells_20250721.bam\"\n",
    "\n",
    "echo \"${pileup_Unedit_day28_bam}.gz\"\n",
    "ls -l \"${pileup_Unedit_day28_bam}.gz\"\n",
    "# \"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam\"\n",
    "# pileup_Unedit_day28_bam_gz=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam.gz\"\n",
    "  \n",
    "# pileup_CROFF_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/pileup_sort_merge_sort_align_t2t_v1_1_trim_20250721_nCATs_Tcells_CROFF_Day28_minion_run2_day8.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "# pileup_CROFF_day28_bam=\"/home/michalula/code/epiCausality/epiCode/differential_methyl/pileups/t2t_v1_1/pileup_CROFF_Day28_Tcells_20250721.bam\"\n",
    "pileup_CROFF_day28_bam=\"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/5mCG/to_t2t_v2_0/pileup_CROFF_Day28_Tcells_20250721.bam\"\n",
    "\n",
    "ref_genome_fa=\"/home/michalula/data/ref_genomes/t2t_v2_0/chm13v2.0.fa\"\n",
    "# ref=\"/home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta\"\n",
    "threads=32\n",
    " \n",
    "cd '/home/michalula/code/epiCausality/epiCode/differential_methyl'\n",
    "\n",
    "modkit dmr pair \\\n",
    "  -a ${pileup_Unedit_day28_bam}.gz \\\n",
    "  -b ${pileup_CROFF_day28_bam}.gz \\\n",
    "  -o ${dmr_result} \\\n",
    "  --ref ${ref_genome_fa} \\\n",
    "  --base C \\\n",
    "  --threads ${threads} \\\n",
    "  --log-filepath dmr.log\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d3ae4b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-rw-r-- 1 michalula michalula 7.8M Nov  9 12:34 /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/single_base_day28_CRoff_vs_Unedit_dmr_20251109.bed\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "dmr_result=\"/home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/single_base_day28_CRoff_vs_Unedit_dmr_20251109.bed\"\n",
    "\n",
    "ls -lah $dmr_result\n",
    "# cat $dmr_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "196255ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/michalula/code/epiCausality/epiCode/differential_methyl'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78ae53cd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fe1d352",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a2df92f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "befa7146",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e61a5823",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5deeb9f3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05ff4bcf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a65d83f0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f6296aa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a1f36a1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f2d68e9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "323ac53c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c488c6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DMR file: /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/single_base_day28_CRoff_vs_Unedit_dmr_20251109.bed\n",
      "Rows,cols: (111263, 11)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_105915/2984644142.py:32: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead\n",
      "  df[c] = pd.to_numeric(df[c], errors='ignore')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1</th>\n",
       "      <th>1839174</th>\n",
       "      <th>1839175</th>\n",
       "      <th>chr1:1839174-1839175</th>\n",
       "      <th>0</th>\n",
       "      <th>m:1</th>\n",
       "      <th>1</th>\n",
       "      <th>.</th>\n",
       "      <th>0.1</th>\n",
       "      <th>m:100.00</th>\n",
       "      <th>..1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1839205</td>\n",
       "      <td>1839206</td>\n",
       "      <td>chr1:1839205-1839206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2014307</td>\n",
       "      <td>2014308</td>\n",
       "      <td>chr1:2014307-2014308</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2014312</td>\n",
       "      <td>2014313</td>\n",
       "      <td>chr1:2014312-2014313</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2014495</td>\n",
       "      <td>2014496</td>\n",
       "      <td>chr1:2014495-2014496</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2014703</td>\n",
       "      <td>2014704</td>\n",
       "      <td>chr1:2014703-2014704</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr1  1839174  1839175  chr1:1839174-1839175    0  m:1  1  .  0.1  \\\n",
       "0  chr1  1839205  1839206  chr1:1839205-1839206  0.0  m:1  1  .    0   \n",
       "1  chr1  2014307  2014308  chr1:2014307-2014308  0.0  m:1  1  .    0   \n",
       "2  chr1  2014312  2014313  chr1:2014312-2014313  0.0  m:1  1  .    0   \n",
       "3  chr1  2014495  2014496  chr1:2014495-2014496  0.0  m:1  1  .    0   \n",
       "4  chr1  2014703  2014704  chr1:2014703-2014704  0.0  m:1  1  .    0   \n",
       "\n",
       "   m:100.00 ..1  \n",
       "0  m:100.00   .  \n",
       "1  m:100.00   .  \n",
       "2  m:100.00   .  \n",
       "3  m:100.00   .  \n",
       "4  m:100.00   .  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>unique</th>\n",
       "      <th>top</th>\n",
       "      <th>freq</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>111263</td>\n",
       "      <td>25</td>\n",
       "      <td>chr1</td>\n",
       "      <td>10199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1839174</th>\n",
       "      <td>111263.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>67038412.499924</td>\n",
       "      <td>58278252.06657</td>\n",
       "      <td>1965.0</td>\n",
       "      <td>17183386.5</td>\n",
       "      <td>50978348.0</td>\n",
       "      <td>102522914.5</td>\n",
       "      <td>247909190.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1839175</th>\n",
       "      <td>111263.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>67038413.499924</td>\n",
       "      <td>58278252.06657</td>\n",
       "      <td>1966.0</td>\n",
       "      <td>17183387.5</td>\n",
       "      <td>50978349.0</td>\n",
       "      <td>102522915.5</td>\n",
       "      <td>247909191.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1839174-1839175</th>\n",
       "      <td>111263</td>\n",
       "      <td>111263</td>\n",
       "      <td>chrY:60366108-60366109</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>111263.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.023171</td>\n",
       "      <td>0.396716</td>\n",
       "      <td>-0.340835</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.662811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>m:1</th>\n",
       "      <td>111263</td>\n",
       "      <td>30</td>\n",
       "      <td>m:1</td>\n",
       "      <td>84432</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>111263.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.074589</td>\n",
       "      <td>0.971748</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>.</th>\n",
       "      <td>111263</td>\n",
       "      <td>17</td>\n",
       "      <td>.</td>\n",
       "      <td>98843</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.1</th>\n",
       "      <td>111263.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.147866</td>\n",
       "      <td>0.666812</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>81.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>m:100.00</th>\n",
       "      <td>111263</td>\n",
       "      <td>97</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>84924</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>..1</th>\n",
       "      <td>111263</td>\n",
       "      <td>48</td>\n",
       "      <td>.</td>\n",
       "      <td>98843</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         count  unique                     top   freq  \\\n",
       "chr1                    111263      25                    chr1  10199   \n",
       "1839174               111263.0     NaN                     NaN    NaN   \n",
       "1839175               111263.0     NaN                     NaN    NaN   \n",
       "chr1:1839174-1839175    111263  111263  chrY:60366108-60366109      1   \n",
       "0                     111263.0     NaN                     NaN    NaN   \n",
       "m:1                     111263      30                     m:1  84432   \n",
       "1                     111263.0     NaN                     NaN    NaN   \n",
       ".                       111263      17                       .  98843   \n",
       "0.1                   111263.0     NaN                     NaN    NaN   \n",
       "m:100.00                111263      97                m:100.00  84924   \n",
       "..1                     111263      48                       .  98843   \n",
       "\n",
       "                                 mean             std       min         25%  \\\n",
       "chr1                              NaN             NaN       NaN         NaN   \n",
       "1839174               67038412.499924  58278252.06657    1965.0  17183386.5   \n",
       "1839175               67038413.499924  58278252.06657    1966.0  17183387.5   \n",
       "chr1:1839174-1839175              NaN             NaN       NaN         NaN   \n",
       "0                            0.023171        0.396716 -0.340835         0.0   \n",
       "m:1                               NaN             NaN       NaN         NaN   \n",
       "1                            1.074589        0.971748       1.0         1.0   \n",
       ".                                 NaN             NaN       NaN         NaN   \n",
       "0.1                          0.147866        0.666812       0.0         0.0   \n",
       "m:100.00                          NaN             NaN       NaN         NaN   \n",
       "..1                               NaN             NaN       NaN         NaN   \n",
       "\n",
       "                             50%          75%          max  \n",
       "chr1                         NaN          NaN          NaN  \n",
       "1839174               50978348.0  102522914.5  247909190.0  \n",
       "1839175               50978349.0  102522915.5  247909191.0  \n",
       "chr1:1839174-1839175         NaN          NaN          NaN  \n",
       "0                            0.0          0.0    20.662811  \n",
       "m:1                          NaN          NaN          NaN  \n",
       "1                            1.0          1.0        100.0  \n",
       ".                            NaN          NaN          NaN  \n",
       "0.1                          0.0          0.0         81.0  \n",
       "m:100.00                     NaN          NaN          NaN  \n",
       "..1                          NaN          NaN          NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/dmr_hist_1839174.html\n",
      "Saved /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/dmr_hist_1839175.html\n",
      "Saved /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/dmr_hist_0.html\n",
      "Saved /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/dmr_hist_1.html\n",
      "Saved /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/dmr_hist_0.1.html\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1</th>\n",
       "      <th>1839174</th>\n",
       "      <th>1839175</th>\n",
       "      <th>chr1:1839174-1839175</th>\n",
       "      <th>0</th>\n",
       "      <th>m:1</th>\n",
       "      <th>1</th>\n",
       "      <th>.</th>\n",
       "      <th>0.1</th>\n",
       "      <th>m:100.00</th>\n",
       "      <th>..1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10198</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247909190</td>\n",
       "      <td>247909191</td>\n",
       "      <td>chr1:247909190-247909191</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10197</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247908237</td>\n",
       "      <td>247908238</td>\n",
       "      <td>chr1:247908237-247908238</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10196</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247908067</td>\n",
       "      <td>247908068</td>\n",
       "      <td>chr1:247908067-247908068</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:0</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:0.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10195</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247908029</td>\n",
       "      <td>247908030</td>\n",
       "      <td>chr1:247908029-247908030</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10194</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247907834</td>\n",
       "      <td>247907835</td>\n",
       "      <td>chr1:247907834-247907835</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10193</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247907773</td>\n",
       "      <td>247907774</td>\n",
       "      <td>chr1:247907773-247907774</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10192</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247907753</td>\n",
       "      <td>247907754</td>\n",
       "      <td>chr1:247907753-247907754</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:0</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:0.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247906959</td>\n",
       "      <td>247906960</td>\n",
       "      <td>chr1:247906959-247906960</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10190</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247905815</td>\n",
       "      <td>247905816</td>\n",
       "      <td>chr1:247905815-247905816</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10189</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247905805</td>\n",
       "      <td>247905806</td>\n",
       "      <td>chr1:247905805-247905806</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10188</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247905731</td>\n",
       "      <td>247905732</td>\n",
       "      <td>chr1:247905731-247905732</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10187</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247905460</td>\n",
       "      <td>247905461</td>\n",
       "      <td>chr1:247905460-247905461</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10186</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247905020</td>\n",
       "      <td>247905021</td>\n",
       "      <td>chr1:247905020-247905021</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:0</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:0.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10185</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247904442</td>\n",
       "      <td>247904443</td>\n",
       "      <td>chr1:247904442-247904443</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10184</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247904196</td>\n",
       "      <td>247904197</td>\n",
       "      <td>chr1:247904196-247904197</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10183</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247904170</td>\n",
       "      <td>247904171</td>\n",
       "      <td>chr1:247904170-247904171</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:0</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:0.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10182</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247904138</td>\n",
       "      <td>247904139</td>\n",
       "      <td>chr1:247904138-247904139</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:0</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:0.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10181</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247904110</td>\n",
       "      <td>247904111</td>\n",
       "      <td>chr1:247904110-247904111</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10180</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247904049</td>\n",
       "      <td>247904050</td>\n",
       "      <td>chr1:247904049-247904050</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10179</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247903999</td>\n",
       "      <td>247904000</td>\n",
       "      <td>chr1:247903999-247904000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10178</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247903787</td>\n",
       "      <td>247903788</td>\n",
       "      <td>chr1:247903787-247903788</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10177</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247903711</td>\n",
       "      <td>247903712</td>\n",
       "      <td>chr1:247903711-247903712</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10176</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247903664</td>\n",
       "      <td>247903665</td>\n",
       "      <td>chr1:247903664-247903665</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10175</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247903650</td>\n",
       "      <td>247903651</td>\n",
       "      <td>chr1:247903650-247903651</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:0</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:0.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10174</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247903560</td>\n",
       "      <td>247903561</td>\n",
       "      <td>chr1:247903560-247903561</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10173</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247903543</td>\n",
       "      <td>247903544</td>\n",
       "      <td>chr1:247903543-247903544</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10172</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247902681</td>\n",
       "      <td>247902682</td>\n",
       "      <td>chr1:247902681-247902682</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10171</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247902678</td>\n",
       "      <td>247902679</td>\n",
       "      <td>chr1:247902678-247902679</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10170</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247902278</td>\n",
       "      <td>247902279</td>\n",
       "      <td>chr1:247902278-247902279</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:0</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:0.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10169</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247901928</td>\n",
       "      <td>247901929</td>\n",
       "      <td>chr1:247901928-247901929</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10168</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247901430</td>\n",
       "      <td>247901431</td>\n",
       "      <td>chr1:247901430-247901431</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10167</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247901377</td>\n",
       "      <td>247901378</td>\n",
       "      <td>chr1:247901377-247901378</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10166</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247901336</td>\n",
       "      <td>247901337</td>\n",
       "      <td>chr1:247901336-247901337</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10165</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247901281</td>\n",
       "      <td>247901282</td>\n",
       "      <td>chr1:247901281-247901282</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10164</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247901253</td>\n",
       "      <td>247901254</td>\n",
       "      <td>chr1:247901253-247901254</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10163</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247901161</td>\n",
       "      <td>247901162</td>\n",
       "      <td>chr1:247901161-247901162</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:0</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:0.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10162</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247900627</td>\n",
       "      <td>247900628</td>\n",
       "      <td>chr1:247900627-247900628</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10161</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247900509</td>\n",
       "      <td>247900510</td>\n",
       "      <td>chr1:247900509-247900510</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10160</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247336961</td>\n",
       "      <td>247336962</td>\n",
       "      <td>chr1:247336961-247336962</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10159</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247336959</td>\n",
       "      <td>247336960</td>\n",
       "      <td>chr1:247336959-247336960</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10158</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247336884</td>\n",
       "      <td>247336885</td>\n",
       "      <td>chr1:247336884-247336885</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10157</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247336842</td>\n",
       "      <td>247336843</td>\n",
       "      <td>chr1:247336842-247336843</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10156</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247336838</td>\n",
       "      <td>247336839</td>\n",
       "      <td>chr1:247336838-247336839</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10155</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247336613</td>\n",
       "      <td>247336614</td>\n",
       "      <td>chr1:247336613-247336614</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10154</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247336386</td>\n",
       "      <td>247336387</td>\n",
       "      <td>chr1:247336386-247336387</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10153</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247336362</td>\n",
       "      <td>247336363</td>\n",
       "      <td>chr1:247336362-247336363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10152</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247336146</td>\n",
       "      <td>247336147</td>\n",
       "      <td>chr1:247336146-247336147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:0</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:0.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10151</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247336138</td>\n",
       "      <td>247336139</td>\n",
       "      <td>chr1:247336138-247336139</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10150</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247336068</td>\n",
       "      <td>247336069</td>\n",
       "      <td>chr1:247336068-247336069</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10149</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247335904</td>\n",
       "      <td>247335905</td>\n",
       "      <td>chr1:247335904-247335905</td>\n",
       "      <td>0.0</td>\n",
       "      <td>m:1</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>m:100.00</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr1    1839174    1839175      chr1:1839174-1839175    0  m:1  1  .  \\\n",
       "10198  chr1  247909190  247909191  chr1:247909190-247909191  0.0  m:1  1  .   \n",
       "10197  chr1  247908237  247908238  chr1:247908237-247908238  0.0  m:1  1  .   \n",
       "10196  chr1  247908067  247908068  chr1:247908067-247908068  0.0  m:0  1  .   \n",
       "10195  chr1  247908029  247908030  chr1:247908029-247908030  0.0  m:1  1  .   \n",
       "10194  chr1  247907834  247907835  chr1:247907834-247907835  0.0  m:1  1  .   \n",
       "10193  chr1  247907773  247907774  chr1:247907773-247907774  0.0  m:1  1  .   \n",
       "10192  chr1  247907753  247907754  chr1:247907753-247907754  0.0  m:0  1  .   \n",
       "10191  chr1  247906959  247906960  chr1:247906959-247906960  0.0  m:1  1  .   \n",
       "10190  chr1  247905815  247905816  chr1:247905815-247905816  0.0  m:1  1  .   \n",
       "10189  chr1  247905805  247905806  chr1:247905805-247905806  0.0  m:1  1  .   \n",
       "10188  chr1  247905731  247905732  chr1:247905731-247905732  0.0  m:1  1  .   \n",
       "10187  chr1  247905460  247905461  chr1:247905460-247905461  0.0  m:1  1  .   \n",
       "10186  chr1  247905020  247905021  chr1:247905020-247905021  0.0  m:0  1  .   \n",
       "10185  chr1  247904442  247904443  chr1:247904442-247904443  0.0  m:1  1  .   \n",
       "10184  chr1  247904196  247904197  chr1:247904196-247904197  0.0  m:1  1  .   \n",
       "10183  chr1  247904170  247904171  chr1:247904170-247904171  0.0  m:0  1  .   \n",
       "10182  chr1  247904138  247904139  chr1:247904138-247904139  0.0  m:0  1  .   \n",
       "10181  chr1  247904110  247904111  chr1:247904110-247904111  0.0  m:1  1  .   \n",
       "10180  chr1  247904049  247904050  chr1:247904049-247904050  0.0  m:1  1  .   \n",
       "10179  chr1  247903999  247904000  chr1:247903999-247904000  0.0  m:1  1  .   \n",
       "10178  chr1  247903787  247903788  chr1:247903787-247903788  0.0  m:1  1  .   \n",
       "10177  chr1  247903711  247903712  chr1:247903711-247903712  0.0  m:1  1  .   \n",
       "10176  chr1  247903664  247903665  chr1:247903664-247903665  0.0  m:1  1  .   \n",
       "10175  chr1  247903650  247903651  chr1:247903650-247903651  0.0  m:0  1  .   \n",
       "10174  chr1  247903560  247903561  chr1:247903560-247903561  0.0  m:1  1  .   \n",
       "10173  chr1  247903543  247903544  chr1:247903543-247903544  0.0  m:1  1  .   \n",
       "10172  chr1  247902681  247902682  chr1:247902681-247902682  0.0  m:1  1  .   \n",
       "10171  chr1  247902678  247902679  chr1:247902678-247902679  0.0  m:1  1  .   \n",
       "10170  chr1  247902278  247902279  chr1:247902278-247902279  0.0  m:0  1  .   \n",
       "10169  chr1  247901928  247901929  chr1:247901928-247901929  0.0  m:1  1  .   \n",
       "10168  chr1  247901430  247901431  chr1:247901430-247901431  0.0  m:1  1  .   \n",
       "10167  chr1  247901377  247901378  chr1:247901377-247901378  0.0  m:1  1  .   \n",
       "10166  chr1  247901336  247901337  chr1:247901336-247901337  0.0  m:1  1  .   \n",
       "10165  chr1  247901281  247901282  chr1:247901281-247901282  0.0  m:1  1  .   \n",
       "10164  chr1  247901253  247901254  chr1:247901253-247901254  0.0  m:1  1  .   \n",
       "10163  chr1  247901161  247901162  chr1:247901161-247901162  0.0  m:0  1  .   \n",
       "10162  chr1  247900627  247900628  chr1:247900627-247900628  0.0  m:1  1  .   \n",
       "10161  chr1  247900509  247900510  chr1:247900509-247900510  0.0  m:1  1  .   \n",
       "10160  chr1  247336961  247336962  chr1:247336961-247336962  0.0  m:1  1  .   \n",
       "10159  chr1  247336959  247336960  chr1:247336959-247336960  0.0  m:1  1  .   \n",
       "10158  chr1  247336884  247336885  chr1:247336884-247336885  0.0  m:1  1  .   \n",
       "10157  chr1  247336842  247336843  chr1:247336842-247336843  0.0  m:1  1  .   \n",
       "10156  chr1  247336838  247336839  chr1:247336838-247336839  0.0  m:1  1  .   \n",
       "10155  chr1  247336613  247336614  chr1:247336613-247336614  0.0  m:1  1  .   \n",
       "10154  chr1  247336386  247336387  chr1:247336386-247336387  0.0  m:1  1  .   \n",
       "10153  chr1  247336362  247336363  chr1:247336362-247336363  0.0  m:1  1  .   \n",
       "10152  chr1  247336146  247336147  chr1:247336146-247336147  0.0  m:0  1  .   \n",
       "10151  chr1  247336138  247336139  chr1:247336138-247336139  0.0  m:1  1  .   \n",
       "10150  chr1  247336068  247336069  chr1:247336068-247336069  0.0  m:1  1  .   \n",
       "10149  chr1  247335904  247335905  chr1:247335904-247335905  0.0  m:1  1  .   \n",
       "\n",
       "       0.1  m:100.00 ..1  \n",
       "10198    0  m:100.00   .  \n",
       "10197    0  m:100.00   .  \n",
       "10196    0    m:0.00   .  \n",
       "10195    0  m:100.00   .  \n",
       "10194    0  m:100.00   .  \n",
       "10193    0  m:100.00   .  \n",
       "10192    0    m:0.00   .  \n",
       "10191    0  m:100.00   .  \n",
       "10190    0  m:100.00   .  \n",
       "10189    0  m:100.00   .  \n",
       "10188    0  m:100.00   .  \n",
       "10187    0  m:100.00   .  \n",
       "10186    0    m:0.00   .  \n",
       "10185    0  m:100.00   .  \n",
       "10184    0  m:100.00   .  \n",
       "10183    0    m:0.00   .  \n",
       "10182    0    m:0.00   .  \n",
       "10181    0  m:100.00   .  \n",
       "10180    0  m:100.00   .  \n",
       "10179    0  m:100.00   .  \n",
       "10178    0  m:100.00   .  \n",
       "10177    0  m:100.00   .  \n",
       "10176    0  m:100.00   .  \n",
       "10175    0    m:0.00   .  \n",
       "10174    0  m:100.00   .  \n",
       "10173    0  m:100.00   .  \n",
       "10172    0  m:100.00   .  \n",
       "10171    0  m:100.00   .  \n",
       "10170    0    m:0.00   .  \n",
       "10169    0  m:100.00   .  \n",
       "10168    0  m:100.00   .  \n",
       "10167    0  m:100.00   .  \n",
       "10166    0  m:100.00   .  \n",
       "10165    0  m:100.00   .  \n",
       "10164    0  m:100.00   .  \n",
       "10163    0    m:0.00   .  \n",
       "10162    0  m:100.00   .  \n",
       "10161    0  m:100.00   .  \n",
       "10160    0  m:100.00   .  \n",
       "10159    0  m:100.00   .  \n",
       "10158    0  m:100.00   .  \n",
       "10157    0  m:100.00   .  \n",
       "10156    0  m:100.00   .  \n",
       "10155    0  m:100.00   .  \n",
       "10154    0  m:100.00   .  \n",
       "10153    0  m:100.00   .  \n",
       "10152    0    m:0.00   .  \n",
       "10151    0  m:100.00   .  \n",
       "10150    0  m:100.00   .  \n",
       "10149    0  m:100.00   .  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done. Visualizations saved to: /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output\n"
     ]
    }
   ],
   "source": [
    "# Visualize modkit DMR result\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "# Path to the DMR BED output (adjust if you use a different file)\n",
    "dmr_path = \"/home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/single_base_day28_CRoff_vs_Unedit_dmr_20251109.bed\"\n",
    "out_dir = os.path.dirname(dmr_path) or \".\"\n",
    "os.makedirs(out_dir, exist_ok=True)\n",
    "\n",
    "# Try reading with a header first, fall back to headerless read\n",
    "try:\n",
    "    df = pd.read_csv(dmr_path, sep='\t', comment='#', header=0, engine='python')\n",
    "    if df.shape[1] < 3:\n",
    "        df = pd.read_csv(dmr_path, sep='\t', comment='#', header=None, engine='python')\n",
    "except Exception as e:\n",
    "    print(\"Failed to read DMR file:\", e)\n",
    "    raise\n",
    "\n",
    "# If the file had no header, pandas will give integer column names - assign common BED/DMR names\n",
    "if isinstance(df.columns[0], int):\n",
    "    ncols = df.shape[1]\n",
    "    default_names = [\"chrom\", \"start\", \"end\", \"name\", \"score\", \"strand\"]\n",
    "    names = default_names[:ncols] + [f\"col{i}\" for i in range(len(default_names), ncols)]\n",
    "    df.columns = names\n",
    "\n",
    "# Coerce numeric-looking columns where possible\n",
    "for c in df.columns:\n",
    "    try:\n",
    "        df[c] = pd.to_numeric(df[c], errors='ignore')\n",
    "    except Exception:\n",
    "        pass\n",
    "\n",
    "print(\"DMR file:\", dmr_path)\n",
    "print(\"Rows,cols:\", df.shape)\n",
    "display(df.head())\n",
    "display(df.describe(include=\"all\").T)\n",
    "\n",
    "# Make histograms of numeric columns and save as HTML\n",
    "numcols = df.select_dtypes(include=[np.number]).columns.tolist()\n",
    "if len(numcols):\n",
    "    for col in numcols:\n",
    "        fig = px.histogram(df, x=col, nbins=80, title=f\"Histogram of {col}\")\n",
    "        html = os.path.join(out_dir, f\"dmr_hist_{col}.html\")\n",
    "        fig.write_html(html)\n",
    "        print(\"Saved\", html)\n",
    "\n",
    "# Scatter: genomic position vs score (if present)\n",
    "score_col = None\n",
    "if 'start' in df.columns:\n",
    "    if 'score' in df.columns:\n",
    "        score_col = 'score'\n",
    "    else:\n",
    "        # try to find a numeric column to plot against position\n",
    "        for c in numcols:\n",
    "            if c != 'start' and c != 'end':\n",
    "                score_col = c\n",
    "                break\n",
    "    if score_col is not None:\n",
    "        df_plot = df.copy()\n",
    "        df_plot['pos'] = df_plot['start'].astype(int)\n",
    "        fig = px.scatter(df_plot, x='pos', y=score_col, color='chrom' if 'chrom' in df_plot.columns else None,\n",
    "                         title=f\"Position vs {score_col}\", hover_data=df_plot.columns.tolist(), height=600)\n",
    "        html = os.path.join(out_dir, 'dmr_position_score_scatter.html')\n",
    "        fig.write_html(html)\n",
    "        print(\"Saved\", html)\n",
    "\n",
    "# Show a table of top hits (by p-value or by score if p-value not present)\n",
    "rank_col = None\n",
    "for key in ['pvalue','p_val','p.value','p','padj','qval','score']:\n",
    "    if key in df.columns:\n",
    "        rank_col = key\n",
    "        break\n",
    "if rank_col is None and len(numcols):\n",
    "    rank_col = numcols[0]\n",
    "if rank_col is not None:\n",
    "    ascending = True if 'p' in str(rank_col).lower() or 'q' in str(rank_col).lower() else False\n",
    "    top = df.sort_values(by=rank_col, ascending=ascending).head(50)\n",
    "    display(top)\n",
    "\n",
    "print(\"Done. Visualizations saved to:\", out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "907d23d7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dimelo_v2_modkit_parsing",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
