{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "557d7279",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "32f9e5e0",
   "metadata": {},
   "source": [
    "# K562 ZFPoff Post SORTING HIGH\n",
    "# Pipeline to Extract and Analyze mC profiles from BAM files and look at Forward & Reverse strands\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00a2ccac",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "\n",
    "def current_time():\n",
    "    \"\"\"Returns the current date and time as a formatted string.\"\"\"\n",
    "    return datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\") \n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    print(\"Current Date and Time:\", current_time())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f29181f",
   "metadata": {},
   "source": [
    "Save subsets of reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b9542dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from pathlib import Path\n",
    "from datetime import datetime\n",
    "\n",
    "from matplotlib import pyplot as plt \n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "# import time\n",
    "# import shap\n",
    "\n",
    "\n",
    "# Add the folder containing 'my_functions.py' to sys.path\n",
    "sys.path.append(\"/home/michalula/code/epiCausality/epiCode/utils/\") # str(Path(__file__).parent / 'utils'))\n",
    "\n",
    "# Import the module or specific functions\n",
    "# from /home/michalula/code/epiCausality/epiCode/utils/funcs_extract_mC_profiles_from_BAMs.py\n",
    "# from funcs_extract_mC_profiles_from_BAMs import system_info, extract_from_bam\n",
    "from funcs_extract_mC_profiles_from_BAMs import (\n",
    "    system_info,\n",
    "    get_reference_sequence,\n",
    "    create_output_directory,\n",
    "    extract_from_bam,\n",
    "    process_extracted_reads,\n",
    "    visualize_data,\n",
    "    create_padded_reads,\n",
    "    plot_padded_reads,\n",
    "    save_padded_reads,\n",
    "    remove_low_methylated_reads,\n",
    "    process_extracted_reads_no_fully_unmethylated,\n",
    "    create_padded_reads_no_fully_unmethylated,\n",
    "    subset_BAM_by_read_IDs,\n",
    "    bam_to_sam,\n",
    "    plot_bam_quality_metrics\n",
    "    # main,\n",
    ")\n",
    "from funcs_analize_forward_reverse_extracted_mC_reads import analize_forward_reverse_CGs_pipeline\n",
    "\n",
    "from funcs_check_quality_bams import get_read_info_by_index, get_read_info_by_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c6eed75",
   "metadata": {},
   "outputs": [],
   "source": [
    "system_info()\n",
    "date_today = datetime.today().strftime('%Y-%m-%d')\n",
    "ref_genome_path = Path('/home/michalula/data/ref_genomes/t2t_v2_0/up_chm13v2.0.fasta')\n",
    "reg_genome_version = \"T2Tv2_0\"\n",
    "region_chr = 'chr1'\n",
    "\n",
    "# 6500bps: EXACT cutting region  \n",
    "# chr1:206,583,334-206,589,873\n",
    "region_start = 206583334    + 20\n",
    "region_end = 206589874      - 20 \n",
    "# CD55 TSS start:Â chr1:206586828-206606065 (+)\n",
    "# (before TSS: 3474, after TSS: 3026 bps)\n",
    "\n",
    "region_str = region_chr + \":\" + str(region_start) + \"-\" + str(region_end) #'chr1:206586162-206586192'\n",
    "region_length = region_end - region_start\n",
    "print(\"region_length\", region_length)\n",
    "\n",
    "\n",
    "motifs=['CG,0']\n",
    "ref_seq_list = get_reference_sequence(ref_genome_path, region_chr, region_start, region_end)\n",
    "\n",
    "print('region_chr', region_chr)\n",
    "print('region_start', region_start)\n",
    "print('region_end', region_end)\n",
    "print(f'{region_chr}:{region_start}-{region_end}') "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36306bc4",
   "metadata": {},
   "source": [
    "# K562 ZFPoff Post Sort LOW\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aedf9733",
   "metadata": {},
   "outputs": [],
   "source": [
    "day_num = 16 #\n",
    "day_num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ee8372d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# output_dir_path = '/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/K562/unedited/analyze_single_reads/output'\n",
    "# # '/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_35/croff/analyze_single_reads/output'\n",
    "# # '/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/croff/analyze_single_reads/output/' # \"/home/michalula/code/epiCausality/epiCode/utils/development_notebooks/\"\n",
    "# output_dir_path\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3debfba9",
   "metadata": {},
   "outputs": [],
   "source": [
    "editor_type = \"ZFPoff_PostSort_HIGH\"  # \"CRISPRoff\"  #\n",
    "cell_type = \"K562\"\n",
    "\n",
    "experiment_condition = f\"{cell_type}_{editor_type}_Day{day_num}_postEP_R9minion\"  # {editor_type}\n",
    "print(\"experiment_condition:\", experiment_condition)\n",
    "# original_bam_path = \"/home/michalula/data/cas9_nanopore/data/20250908_nCATs_T_CRoff_Day_35/5mCG/to_t2t_v2_0/sort_align_t2t_v2_0_trim_20250908_Day35_CROFF_Tcells_2Libraries_Minion_R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "# \"/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/5mCG/to_t2t_v2_0/sort_align_t2t_v2_0_trim_20250721_Day28_CROFF_T_cells_minion_merged_flowcells_output.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "# folder_path = \"/home/michalula/data/cas9_nanopore/data/20241226_nCATs_K562_ZFPOFFpostSort_LOW70/pod5_converted_basecall/5mCG/to_t2t_v2_0/\"\n",
    "# bam_name = \"sort_align_t2t_v2_0_trim_20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "original_bam_path = \"/home/michalula/data/cas9_nanopore/data/20241226_nCATS_K562_ZFPOFFpostSort_HIGH/pod5_converted_basecall/5mCG/to_t2t_v2_0/sort_align_t2t_v2_0_trim_20241226_nCATS_K562_ZFPOFFpostSort_HIGH_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam\"\n",
    "\n",
    "# original_bam_path = folder_path + bam_name\n",
    "original_bam_path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc547399",
   "metadata": {},
   "source": [
    "# Full run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "341d5d75",
   "metadata": {},
   "outputs": [],
   "source": [
    "threshold_mC = 0.7 # 0.995 #0.7 #  0.9 #0.99\n",
    "\n",
    "\n",
    "threshold_mode_base_qualities=10\n",
    "threshold_mapping_qualities=60 \n",
    "# threshold_avg_base_qualities=20\n",
    "\n",
    "threshold_fraction_overlap_aligned = 0.9 # 0.1\n",
    "threshold_fraction_mismatches = 0.7 # 0.5\n",
    "# indel_fraction_threshold = 1  # 0.8\n",
    "# non_fraction_threshold   = 1\n",
    "max_reads_plot = 3000\n",
    "remove_reads_that_dont_fully_cover_region=True\n",
    "\n",
    "# experiment_name = \"day\"+str(day_num)+\"_unedited_Tcells_Thr\"+str(threshold_mC)+\"_NoFullyUnmethylated_T2Tv2\"\n",
    "experiment_name = f'{experiment_condition}_threshold_mC{threshold_mC}_{reg_genome_version}' + \\\n",
    "    \"_filterMode\" + str(threshold_mode_base_qualities) + \\\n",
    "    \"_NoFullyUnmeth_ovrlap\"+str(threshold_fraction_overlap_aligned)+\"_mismat\"+str(threshold_fraction_mismatches)+\"_mapQ\"+str(threshold_mapping_qualities)\n",
    "    # +\"_avgBaseQ\"+str(threshold_avg_base_qualities)\n",
    "\n",
    "bam_path = original_bam_path #  unedit_bam_path\n",
    "\n",
    "print('experiment_name = ', experiment_name)\n",
    "\n",
    "save_padded_reads_name_np = f\"padded_reads_{experiment_name}_mCthresh{threshold_mC}_{reg_genome_version}_{region_str}_{date_today}.npy\"\n",
    "output_dir = create_output_directory(\"./dimelo_v2_output\")\n",
    "\n",
    "extract_file, extract_regions = extract_from_bam(\n",
    "    experiment_name = experiment_name,\n",
    "    bam_path=bam_path,\n",
    "    ref_genome_file=ref_genome_path,\n",
    "    output_dir=output_dir,\n",
    "    region=region_str,\n",
    "    motifs=motifs,\n",
    "    output_name='extracted_reads',\n",
    "    threshold_mC=threshold_mC,\n",
    ")\n",
    "keep_full_coverage_reads_only=True\n",
    "\n",
    "if extract_file:\n",
    "    # extract_file, original_bam_path, region, motifs, \n",
    "    #                         ref_seq_list, ref_genome_path,\n",
    "    #                         experiment_name, output_dir, \n",
    "    #                         keep_full_coverage_reads_only=True,\n",
    "    #                         save_indels_mismatches_count_csv_path=\"indels_mismatches_count.csv\",\n",
    "    #                         threshold_fraction_overlap_aligned=0.5, threshold_fraction_mismatches=0.5, \n",
    "    #                         threshold_mapping_qualities=60, threshold_avg_base_qualities=np.nan,\n",
    "    #                         threshold_mode_base_qualities=10,\n",
    "    #                         max_reads_plot=3000,\n",
    "    #                         remove_reads_that_dont_fully_cover_region=True\n",
    "    reads_df, regions_dict = process_extracted_reads(extract_file, bam_path, region_str, \n",
    "        motifs, ref_seq_list, ref_genome_path=ref_genome_path, \n",
    "        experiment_name=experiment_name, output_dir=output_dir, \n",
    "        keep_full_coverage_reads_only=keep_full_coverage_reads_only,\n",
    "        threshold_fraction_overlap_aligned=threshold_fraction_overlap_aligned,\n",
    "        threshold_fraction_mismatches=threshold_fraction_mismatches,\n",
    "        threshold_mapping_qualities=threshold_mapping_qualities, \n",
    "        threshold_avg_base_qualities=threshold_avg_base_qualities,\n",
    "        threshold_mode_base_qualities=threshold_mode_base_qualities,\n",
    "        max_reads_plot=max_reads_plot,\n",
    "        remove_reads_that_dont_fully_cover_region=remove_reads_that_dont_fully_cover_region)\n",
    "        # indel_fraction_threshold=indel_fraction_threshold, \n",
    "        # non_fraction_threshold=non_fraction_threshold)\n",
    "        \n",
    "    visualize_data(reads_df)\n",
    "\n",
    "    padded_reads = create_padded_reads(reads_df, regions_dict, region_length)\n",
    "    if padded_reads is not None:\n",
    "        plot_padded_reads(padded_reads, ref_seq_list)\n",
    "        save_padded_reads(padded_reads, output_dir, save_padded_reads_name_np)\n",
    "\n",
    "# Processing and visualizing CpG units on forward and reverse strands \n",
    "try: \n",
    "\n",
    "    # Process pipeline\n",
    "    CGs_all, C_fwd_df, G_revs_df, padded_reads_df, CG_pair_idx, CG_coordinates, CG_info_df = analize_forward_reverse_CGs_pipeline(\n",
    "        experiment_name=experiment_name, save_folder_path=output_dir, \n",
    "        save_padded_reads_name_np=save_padded_reads_name_np, \n",
    "        ref_genome_file=ref_genome_path, region_chr=region_chr, region_start=region_start, region_end=region_end\n",
    "    )\n",
    "\n",
    "    print(\"Pipeline executed successfully (analize_forward_reverse_CGs_pipeline function)\")\n",
    "    # return CGs_all, C_fwd_df, G_revs_df, padded_reads_df\n",
    "\n",
    "except Exception as e:\n",
    "    print(f\"Error in main pipeline (analize_forward_reverse_CGs_pipeline function): {e}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc9cc33f",
   "metadata": {},
   "source": [
    "# threshold_mC  0.995 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da4c7575",
   "metadata": {},
   "outputs": [],
   "source": [
    "threshold_mC = 0.995 #0.7 #  0.9 #0.99\n",
    "\n",
    "\n",
    "threshold_mode_base_qualities=10\n",
    "threshold_mapping_qualities=60 \n",
    "# threshold_avg_base_qualities=20\n",
    "\n",
    "threshold_fraction_overlap_aligned = 0.9 # 0.1\n",
    "threshold_fraction_mismatches = 0.7 # 0.5\n",
    "# indel_fraction_threshold = 1  # 0.8\n",
    "# non_fraction_threshold   = 1\n",
    "max_reads_plot = 3000\n",
    "remove_reads_that_dont_fully_cover_region=True\n",
    "\n",
    "# experiment_name = \"day\"+str(day_num)+\"_unedited_Tcells_Thr\"+str(threshold_mC)+\"_NoFullyUnmethylated_T2Tv2\"\n",
    "experiment_name = f'{experiment_condition}_threshold_mC{threshold_mC}_{reg_genome_version}' + \\\n",
    "    \"_filterMode\" + str(threshold_mode_base_qualities) + \\\n",
    "    \"_NoFullyUnmeth_ovrlap\"+str(threshold_fraction_overlap_aligned)+\"_mismat\"+str(threshold_fraction_mismatches)+\"_mapQ\"+str(threshold_mapping_qualities)\n",
    "    # +\"_avgBaseQ\"+str(threshold_avg_base_qualities)\n",
    "\n",
    "bam_path = original_bam_path #  unedit_bam_path\n",
    "\n",
    "print('experiment_name = ', experiment_name)\n",
    "\n",
    "save_padded_reads_name_np = f\"padded_reads_{experiment_name}_mCthresh{threshold_mC}_{reg_genome_version}_{region_str}_{date_today}.npy\"\n",
    "output_dir = create_output_directory(\"./dimelo_v2_output\")\n",
    "\n",
    "extract_file, extract_regions = extract_from_bam(\n",
    "    experiment_name = experiment_name,\n",
    "    bam_path=bam_path,\n",
    "    ref_genome_file=ref_genome_path,\n",
    "    output_dir=output_dir,\n",
    "    region=region_str,\n",
    "    motifs=motifs,\n",
    "    output_name='extracted_reads',\n",
    "    threshold_mC=threshold_mC,\n",
    ")\n",
    "keep_full_coverage_reads_only=True\n",
    "\n",
    "if extract_file:\n",
    "    reads_df, regions_dict = process_extracted_reads(extract_file, bam_path, region_str, \n",
    "        motifs, ref_seq_list, ref_genome_path=ref_genome_path, \n",
    "        experiment_name=experiment_name, output_dir=output_dir, \n",
    "        keep_full_coverage_reads_only=keep_full_coverage_reads_only,\n",
    "        threshold_fraction_overlap_aligned=threshold_fraction_overlap_aligned,\n",
    "        threshold_fraction_mismatches=threshold_fraction_mismatches,\n",
    "        threshold_mapping_qualities=threshold_mapping_qualities, \n",
    "        threshold_avg_base_qualities=threshold_avg_base_qualities,\n",
    "        threshold_mode_base_qualities=threshold_mode_base_qualities,\n",
    "        max_reads_plot=max_reads_plot,\n",
    "        remove_reads_that_dont_fully_cover_region=remove_reads_that_dont_fully_cover_region)\n",
    "        # indel_fraction_threshold=indel_fraction_threshold, \n",
    "        # non_fraction_threshold=non_fraction_threshold)\n",
    "        \n",
    "    visualize_data(reads_df)\n",
    "\n",
    "    padded_reads = create_padded_reads(reads_df, regions_dict, region_length)\n",
    "    if padded_reads is not None:\n",
    "        plot_padded_reads(padded_reads, ref_seq_list)\n",
    "        save_padded_reads(padded_reads, output_dir, save_padded_reads_name_np)\n",
    "\n",
    "# Processing and visualizing CpG units on forward and reverse strands \n",
    "try: \n",
    "\n",
    "    # Process pipeline\n",
    "    CGs_all, C_fwd_df, G_revs_df, padded_reads_df, CG_pair_idx, CG_coordinates, CG_info_df = analize_forward_reverse_CGs_pipeline(\n",
    "        experiment_name=experiment_name, save_folder_path=output_dir, \n",
    "        save_padded_reads_name_np=save_padded_reads_name_np, \n",
    "        ref_genome_file=ref_genome_path, region_chr=region_chr, region_start=region_start, region_end=region_end\n",
    "    )\n",
    "\n",
    "    print(\"Pipeline executed successfully (analize_forward_reverse_CGs_pipeline function)\")\n",
    "    # return CGs_all, C_fwd_df, G_revs_df, padded_reads_df\n",
    "\n",
    "except Exception as e:\n",
    "    print(f\"Error in main pipeline (analize_forward_reverse_CGs_pipeline function): {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85a0b1d8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python dimelo_v2_modkit_parsing",
   "language": "python",
   "name": "dimelo_v2_modkit_parsing"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
