{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Database creation of Telomere-to-telomere consortium CHM13 Annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Before you begin, please make sure you have the proper files downloaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download link for CHM13 Annotations: https://github.com/marbl/CHM13?tab=readme-ov-file\n",
    "##Scroll Down to find the UCSC GENCODEv35 CAT/Liftoff v2 Annotation file, which will be used for this database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting gffutils\n",
      "  Downloading gffutils-0.13-py3-none-any.whl.metadata (1.5 kB)\n",
      "Collecting pyfaidx>=0.5.5.2 (from gffutils)\n",
      "  Downloading pyfaidx-0.8.1.3-py3-none-any.whl.metadata (25 kB)\n",
      "Collecting argh>=0.26.2 (from gffutils)\n",
      "  Downloading argh-0.31.3-py3-none-any.whl.metadata (7.4 kB)\n",
      "Collecting argcomplete>=1.9.4 (from gffutils)\n",
      "  Downloading argcomplete-3.6.1-py3-none-any.whl.metadata (16 kB)\n",
      "Collecting simplejson (from gffutils)\n",
      "  Downloading simplejson-3.20.1-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.2 kB)\n",
      "Collecting importlib-metadata (from pyfaidx>=0.5.5.2->gffutils)\n",
      "  Downloading importlib_metadata-8.6.1-py3-none-any.whl.metadata (4.7 kB)\n",
      "Requirement already satisfied: packaging in /home/michalula/miniconda3/envs/dimelo_v2_modkit_parsing/lib/python3.11/site-packages (from pyfaidx>=0.5.5.2->gffutils) (24.2)\n",
      "Collecting zipp>=3.20 (from importlib-metadata->pyfaidx>=0.5.5.2->gffutils)\n",
      "  Downloading zipp-3.21.0-py3-none-any.whl.metadata (3.7 kB)\n",
      "Downloading gffutils-0.13-py3-none-any.whl (1.6 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m15.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading argcomplete-3.6.1-py3-none-any.whl (43 kB)\n",
      "Downloading argh-0.31.3-py3-none-any.whl (44 kB)\n",
      "Downloading pyfaidx-0.8.1.3-py3-none-any.whl (28 kB)\n",
      "Downloading simplejson-3.20.1-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (144 kB)\n",
      "Downloading importlib_metadata-8.6.1-py3-none-any.whl (26 kB)\n",
      "Downloading zipp-3.21.0-py3-none-any.whl (9.6 kB)\n",
      "Installing collected packages: zipp, simplejson, argh, argcomplete, importlib-metadata, pyfaidx, gffutils\n",
      "Successfully installed argcomplete-3.6.1 argh-0.31.3 gffutils-0.13 importlib-metadata-8.6.1 pyfaidx-0.8.1.3 simplejson-3.20.1 zipp-3.21.0\n"
     ]
    }
   ],
   "source": [
    "! python3 -m pip install gffutils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gffutils\n",
    "import pysam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### If this is your first time importing the gff3 database, run this command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This can take a long time, so only run this once (remove the #)\n",
    "# gene_annotation_gff3_chm13_v2_path ='/home/michalula/code/Ella/data/chm13.draft_v2.0.gene_annotation.gff3'\n",
    "\n",
    "# db = gffutils.create_db(gene_annotation_gff3_chm13_v2_path,\n",
    "#                          dbfn='chm13db.db', force=True, keep_order=True, \n",
    "# merge_strategy='create_unique', sort_attribute_values=True) \n",
    "\n",
    "# replace gff3 with your gff3 file, dbfn can be whichever name you want to store the transposed file as\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Feature the selected database "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<gffutils.interface.FeatureDB at 0x78cf0bd1d910>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation_DB_chm13_v2_path = '/home/michalula/code/Ella/data/chm13db.db'\n",
    "db = gffutils.FeatureDB(gene_annotation_DB_chm13_v2_path, keep_order=True)\n",
    "db"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Select the desired range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<generator object FeatureDB.region at 0x78cf14dee110>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slct = db.region(region=('chr1', 206560169, 206614236), completely_within=False)\n",
    "slct"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Iterate through the region and transpose it to a dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "records = []\n",
    "for gene in slct:\n",
    "    records.append({\n",
    "        \"seqid\": gene.seqid,\n",
    "        \"source\": gene.source,\n",
    "        \"featuretype\": gene.featuretype,\n",
    "        \"start\": gene.start,\n",
    "        \"end\": gene.end,\n",
    "        \"strand\": gene.strand,\n",
    "        \"score\": gene.score,\n",
    "        \"attributes\": gene.attributes  # Attributes as a dictionary (will be expanded later)\n",
    "    })"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Expand out the attributes and format the dictionary as a database, dropping unnecessary columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(records)\n",
    "attributes_df = df[\"attributes\"].apply(pd.Series)\n",
    "df = pd.concat([df.drop(columns=[\"attributes\"]), attributes_df], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>featuretype</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>source_gene_common_name</th>\n",
       "      <th>source_gene</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>...</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Parent</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>rna_support</th>\n",
       "      <th>reference_support</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>gene</td>\n",
       "      <td>206505417</td>\n",
       "      <td>206574588</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AL596218.1</td>\n",
       "      <td>ENSG00000237074.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>transcript</td>\n",
       "      <td>206505417</td>\n",
       "      <td>206574588</td>\n",
       "      <td>+</td>\n",
       "      <td>9480</td>\n",
       "      <td>AL596218.1</td>\n",
       "      <td>ENSG00000237074.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>OTTHUMG00000036254.2</td>\n",
       "      <td>OTTHUMT00000488535.1</td>\n",
       "      <td>CHM13_T0017321</td>\n",
       "      <td>CHM13_G0004546</td>\n",
       "      <td>AL596218.1-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>intron</td>\n",
       "      <td>206509833</td>\n",
       "      <td>206567065</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AL596218.1</td>\n",
       "      <td>ENSG00000237074.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>OTTHUMG00000036254.2</td>\n",
       "      <td>OTTHUMT00000488535.1</td>\n",
       "      <td>CHM13_T0017321</td>\n",
       "      <td>CHM13_T0017321</td>\n",
       "      <td>AL596218.1-201</td>\n",
       "      <td>N/A</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>transcript</td>\n",
       "      <td>206514363</td>\n",
       "      <td>206574417</td>\n",
       "      <td>+</td>\n",
       "      <td>9350</td>\n",
       "      <td>AL596218.1</td>\n",
       "      <td>ENSG00000237074.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>OTTHUMG00000036254.2</td>\n",
       "      <td>OTTHUMT00000088207.2</td>\n",
       "      <td>CHM13_T0017322</td>\n",
       "      <td>CHM13_G0004546</td>\n",
       "      <td>AL596218.1-202</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>intron</td>\n",
       "      <td>206530333</td>\n",
       "      <td>206567065</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AL596218.1</td>\n",
       "      <td>ENSG00000237074.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>OTTHUMG00000036254.2</td>\n",
       "      <td>OTTHUMT00000088207.2</td>\n",
       "      <td>CHM13_T0017322</td>\n",
       "      <td>CHM13_T0017322</td>\n",
       "      <td>AL596218.1-202</td>\n",
       "      <td>N/A</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>intron</td>\n",
       "      <td>206605808</td>\n",
       "      <td>206624848</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CD55</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>OTTHUMG00000036255.7</td>\n",
       "      <td>OTTHUMT00000382038.2</td>\n",
       "      <td>CHM13_T0017327</td>\n",
       "      <td>CHM13_T0017327</td>\n",
       "      <td>CD55-201</td>\n",
       "      <td>N/A</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSP00000316333.8</td>\n",
       "      <td>HGNC:2665</td>\n",
       "      <td>CCDS44307.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>206605941</td>\n",
       "      <td>206605943</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CD55</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>OTTHUMG00000036255.7</td>\n",
       "      <td>OTTHUMT00000088210.2</td>\n",
       "      <td>CHM13_T0017323</td>\n",
       "      <td>CHM13_T0017323</td>\n",
       "      <td>CD55-203</td>\n",
       "      <td>N/A</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSP00000356030.2</td>\n",
       "      <td>HGNC:2665</td>\n",
       "      <td>CCDS73022.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>exon</td>\n",
       "      <td>206612746</td>\n",
       "      <td>206612860</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CD55</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>OTTHUMG00000036255.7</td>\n",
       "      <td>OTTHUMT00000494554.1</td>\n",
       "      <td>CHM13_T0017325</td>\n",
       "      <td>CHM13_T0017325</td>\n",
       "      <td>CD55-214</td>\n",
       "      <td>N/A</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSP00000496251.1</td>\n",
       "      <td>HGNC:2665</td>\n",
       "      <td>CCDS86046.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>CDS</td>\n",
       "      <td>206612746</td>\n",
       "      <td>206612860</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CD55</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>OTTHUMG00000036255.7</td>\n",
       "      <td>OTTHUMT00000494554.1</td>\n",
       "      <td>CHM13_T0017325</td>\n",
       "      <td>CHM13_T0017325</td>\n",
       "      <td>CD55-214</td>\n",
       "      <td>N/A</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSP00000496251.1</td>\n",
       "      <td>HGNC:2665</td>\n",
       "      <td>CCDS86046.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>intron</td>\n",
       "      <td>206612861</td>\n",
       "      <td>206624848</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CD55</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>OTTHUMG00000036255.7</td>\n",
       "      <td>OTTHUMT00000494554.1</td>\n",
       "      <td>CHM13_T0017325</td>\n",
       "      <td>CHM13_T0017325</td>\n",
       "      <td>CD55-214</td>\n",
       "      <td>N/A</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSP00000496251.1</td>\n",
       "      <td>HGNC:2665</td>\n",
       "      <td>CCDS86046.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>325 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqid source featuretype      start        end strand score  \\\n",
       "0    chr1    CAT        gene  206505417  206574588      +     .   \n",
       "1    chr1    CAT  transcript  206505417  206574588      +  9480   \n",
       "2    chr1    CAT      intron  206509833  206567065      +     .   \n",
       "3    chr1    CAT  transcript  206514363  206574417      +  9350   \n",
       "4    chr1    CAT      intron  206530333  206567065      +     .   \n",
       "..    ...    ...         ...        ...        ...    ...   ...   \n",
       "320  chr1    CAT      intron  206605808  206624848      +     .   \n",
       "321  chr1    CAT  stop_codon  206605941  206605943      +     .   \n",
       "322  chr1    CAT        exon  206612746  206612860      +     .   \n",
       "323  chr1    CAT         CDS  206612746  206612860      +     .   \n",
       "324  chr1    CAT      intron  206612861  206624848      +     .   \n",
       "\n",
       "    source_gene_common_name         source_gene    gene_biotype  ...  \\\n",
       "0                AL596218.1   ENSG00000237074.2          lncRNA  ...   \n",
       "1                AL596218.1   ENSG00000237074.2          lncRNA  ...   \n",
       "2                AL596218.1   ENSG00000237074.2          lncRNA  ...   \n",
       "3                AL596218.1   ENSG00000237074.2          lncRNA  ...   \n",
       "4                AL596218.1   ENSG00000237074.2          lncRNA  ...   \n",
       "..                      ...                 ...             ...  ...   \n",
       "320                    CD55  ENSG00000196352.16  protein_coding  ...   \n",
       "321                    CD55  ENSG00000196352.16  protein_coding  ...   \n",
       "322                    CD55  ENSG00000196352.16  protein_coding  ...   \n",
       "323                    CD55  ENSG00000196352.16  protein_coding  ...   \n",
       "324                    CD55  ENSG00000196352.16  protein_coding  ...   \n",
       "\n",
       "              havana_gene     havana_transcript   transcript_id  \\\n",
       "0                     NaN                   NaN             NaN   \n",
       "1    OTTHUMG00000036254.2  OTTHUMT00000488535.1  CHM13_T0017321   \n",
       "2    OTTHUMG00000036254.2  OTTHUMT00000488535.1  CHM13_T0017321   \n",
       "3    OTTHUMG00000036254.2  OTTHUMT00000088207.2  CHM13_T0017322   \n",
       "4    OTTHUMG00000036254.2  OTTHUMT00000088207.2  CHM13_T0017322   \n",
       "..                    ...                   ...             ...   \n",
       "320  OTTHUMG00000036255.7  OTTHUMT00000382038.2  CHM13_T0017327   \n",
       "321  OTTHUMG00000036255.7  OTTHUMT00000088210.2  CHM13_T0017323   \n",
       "322  OTTHUMG00000036255.7  OTTHUMT00000494554.1  CHM13_T0017325   \n",
       "323  OTTHUMG00000036255.7  OTTHUMT00000494554.1  CHM13_T0017325   \n",
       "324  OTTHUMG00000036255.7  OTTHUMT00000494554.1  CHM13_T0017325   \n",
       "\n",
       "             Parent transcript_name rna_support reference_support  \\\n",
       "0               NaN             NaN         NaN               NaN   \n",
       "1    CHM13_G0004546  AL596218.1-201         NaN               NaN   \n",
       "2    CHM13_T0017321  AL596218.1-201         N/A              True   \n",
       "3    CHM13_G0004546  AL596218.1-202         NaN               NaN   \n",
       "4    CHM13_T0017322  AL596218.1-202         N/A              True   \n",
       "..              ...             ...         ...               ...   \n",
       "320  CHM13_T0017327        CD55-201         N/A              True   \n",
       "321  CHM13_T0017323        CD55-203         N/A              True   \n",
       "322  CHM13_T0017325        CD55-214         N/A              True   \n",
       "323  CHM13_T0017325        CD55-214         N/A              True   \n",
       "324  CHM13_T0017325        CD55-214         N/A              True   \n",
       "\n",
       "            protein_id    hgnc_id       ccdsid  \n",
       "0                  NaN        NaN          NaN  \n",
       "1                  NaN        NaN          NaN  \n",
       "2                  NaN        NaN          NaN  \n",
       "3                  NaN        NaN          NaN  \n",
       "4                  NaN        NaN          NaN  \n",
       "..                 ...        ...          ...  \n",
       "320  ENSP00000316333.8  HGNC:2665  CCDS44307.1  \n",
       "321  ENSP00000356030.2  HGNC:2665  CCDS73022.1  \n",
       "322  ENSP00000496251.1  HGNC:2665  CCDS86046.1  \n",
       "323  ENSP00000496251.1  HGNC:2665  CCDS86046.1  \n",
       "324  ENSP00000496251.1  HGNC:2665  CCDS86046.1  \n",
       "\n",
       "[325 rows x 48 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for col in df.columns:\n",
    "    df[col] = df[col].apply(lambda x: x[0] if isinstance(x, list) and len(x) == 1 else x)\n",
    "df.dropna(how='all', axis=1, inplace=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dimelo_v2_modkit_parsing",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
