{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Database creation of Telomere-to-telomere consortium CHM13 Annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Before you begin, please make sure you have the proper files downloaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download link for CHM13 Annotations: https://github.com/marbl/CHM13?tab=readme-ov-file\n",
    "##Scroll Down to find the UCSC GENCODEv35 CAT/Liftoff v2 Annotation file, which will be used for this database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! python3 -m pip install gffutils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gffutils\n",
    "import pysam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### If this is your first time importing the gff3 database, run this command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This can take a long time, so only run this once (remove the #)\n",
    "# gene_annotation_gff3_chm13_v2_path ='/home/michalula/code/Ella/data/chm13.draft_v2.0.gene_annotation.gff3'\n",
    "\n",
    "# db = gffutils.create_db(gene_annotation_gff3_chm13_v2_path,\n",
    "#                          dbfn='chm13db.db', force=True, keep_order=True, \n",
    "# merge_strategy='create_unique', sort_attribute_values=True) \n",
    "\n",
    "# replace gff3 with your gff3 file, dbfn can be whichever name you want to store the transposed file as\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Feature the selected database "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<gffutils.interface.FeatureDB at 0x7453feb8e850>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation_DB_chm13_v2_path = '/home/michalula/code/Ella/data/t2tv2_0_gene_annotation_chm13db.db'  # '/home/michalula/code/Ella/data/chm13db.db'\n",
    "db = gffutils.FeatureDB(gene_annotation_DB_chm13_v2_path, keep_order=True)\n",
    "db"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Select the desired range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<generator object FeatureDB.region at 0x7453e85c0cc0>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slct = db.region(region=('chr1', 206586910, 206587029), completely_within=False)\n",
    "# t2tv2.0: chr: 206586910 -\t206587029\n",
    "# slct = db.region(region=('chr1', 206560169, 206614236), completely_within=False)\n",
    "slct"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Iterate through the region and transpose it to a dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "records = []\n",
    "for gene in slct:\n",
    "    records.append({\n",
    "        \"seqid\": gene.seqid,\n",
    "        \"source\": gene.source,\n",
    "        \"featuretype\": gene.featuretype,\n",
    "        \"start\": gene.start,\n",
    "        \"end\": gene.end,\n",
    "        \"strand\": gene.strand,\n",
    "        \"score\": gene.score,\n",
    "        \"attributes\": gene.attributes  # Attributes as a dictionary (will be expanded later)\n",
    "    })"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Expand out the attributes and format the dictionary as a database, dropping unnecessary columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(records)\n",
    "attributes_df = df[\"attributes\"].apply(pd.Series)\n",
    "df = pd.concat([df.drop(columns=[\"attributes\"]), attributes_df], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>featuretype</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>source_transcript</th>\n",
       "      <th>source_transcript_name</th>\n",
       "      <th>source_gene</th>\n",
       "      <th>...</th>\n",
       "      <th>Name</th>\n",
       "      <th>rna_support</th>\n",
       "      <th>reference_support</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>alternative_source_transcripts</th>\n",
       "      <th>collapsed_gene_ids</th>\n",
       "      <th>collapsed_gene_names</th>\n",
       "      <th>frameshift</th>\n",
       "      <th>extra_paralog</th>\n",
       "      <th>exon_anotation_support</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>exon</td>\n",
       "      <td>206586828</td>\n",
       "      <td>206587161</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000367063.6</td>\n",
       "      <td>CD55-203</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>True</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>transcript</td>\n",
       "      <td>206586828</td>\n",
       "      <td>206606065</td>\n",
       "      <td>+</td>\n",
       "      <td>10000</td>\n",
       "      <td>ENST00000367063.6</td>\n",
       "      <td>CD55-203</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>CD55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>nan</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>gene</td>\n",
       "      <td>206586828</td>\n",
       "      <td>206652117</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>N/A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>CD55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>exon</td>\n",
       "      <td>206586939</td>\n",
       "      <td>206587161</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000391921.9</td>\n",
       "      <td>CD55-206</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>True</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>exon</td>\n",
       "      <td>206586939</td>\n",
       "      <td>206587161</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000645323.1</td>\n",
       "      <td>CD55-214</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>True</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>transcript</td>\n",
       "      <td>206586939</td>\n",
       "      <td>206625016</td>\n",
       "      <td>+</td>\n",
       "      <td>10000</td>\n",
       "      <td>ENST00000391921.9</td>\n",
       "      <td>CD55-206</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>CD55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>nan</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>transcript</td>\n",
       "      <td>206586939</td>\n",
       "      <td>206625639</td>\n",
       "      <td>+</td>\n",
       "      <td>10000</td>\n",
       "      <td>ENST00000645323.1</td>\n",
       "      <td>CD55-214</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>CD55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>nan</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>exon</td>\n",
       "      <td>206586974</td>\n",
       "      <td>206587161</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000367064.9</td>\n",
       "      <td>CD55-204</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>True</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>transcript</td>\n",
       "      <td>206586974</td>\n",
       "      <td>206626269</td>\n",
       "      <td>+</td>\n",
       "      <td>10000</td>\n",
       "      <td>ENST00000367064.9</td>\n",
       "      <td>CD55-204</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>CD55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>nan</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>exon</td>\n",
       "      <td>206586997</td>\n",
       "      <td>206587161</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000314754.12</td>\n",
       "      <td>CD55-201</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>True</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>transcript</td>\n",
       "      <td>206586997</td>\n",
       "      <td>206625804</td>\n",
       "      <td>+</td>\n",
       "      <td>10000</td>\n",
       "      <td>ENST00000314754.12</td>\n",
       "      <td>CD55-201</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>CD55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>nan</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>exon</td>\n",
       "      <td>206587029</td>\n",
       "      <td>206587161</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000367067.8</td>\n",
       "      <td>CD55-205</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>True</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>transcript</td>\n",
       "      <td>206587029</td>\n",
       "      <td>206625070</td>\n",
       "      <td>+</td>\n",
       "      <td>9970</td>\n",
       "      <td>ENST00000367067.8</td>\n",
       "      <td>CD55-205</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>CD55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CD55</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid source featuretype      start        end strand  score  \\\n",
       "0   chr1    CAT        exon  206586828  206587161      +      .   \n",
       "1   chr1    CAT  transcript  206586828  206606065      +  10000   \n",
       "2   chr1    CAT        gene  206586828  206652117      +      .   \n",
       "3   chr1    CAT        exon  206586939  206587161      +      .   \n",
       "4   chr1    CAT        exon  206586939  206587161      +      .   \n",
       "5   chr1    CAT  transcript  206586939  206625016      +  10000   \n",
       "6   chr1    CAT  transcript  206586939  206625639      +  10000   \n",
       "7   chr1    CAT        exon  206586974  206587161      +      .   \n",
       "8   chr1    CAT  transcript  206586974  206626269      +  10000   \n",
       "9   chr1    CAT        exon  206586997  206587161      +      .   \n",
       "10  chr1    CAT  transcript  206586997  206625804      +  10000   \n",
       "11  chr1    CAT        exon  206587029  206587161      +      .   \n",
       "12  chr1    CAT  transcript  206587029  206625070      +   9970   \n",
       "\n",
       "     source_transcript source_transcript_name         source_gene  ...  Name  \\\n",
       "0    ENST00000367063.6               CD55-203  ENSG00000196352.16  ...  CD55   \n",
       "1    ENST00000367063.6               CD55-203  ENSG00000196352.16  ...  CD55   \n",
       "2                  N/A                    NaN  ENSG00000196352.16  ...  CD55   \n",
       "3    ENST00000391921.9               CD55-206  ENSG00000196352.16  ...  CD55   \n",
       "4    ENST00000645323.1               CD55-214  ENSG00000196352.16  ...  CD55   \n",
       "5    ENST00000391921.9               CD55-206  ENSG00000196352.16  ...  CD55   \n",
       "6    ENST00000645323.1               CD55-214  ENSG00000196352.16  ...  CD55   \n",
       "7    ENST00000367064.9               CD55-204  ENSG00000196352.16  ...  CD55   \n",
       "8    ENST00000367064.9               CD55-204  ENSG00000196352.16  ...  CD55   \n",
       "9   ENST00000314754.12               CD55-201  ENSG00000196352.16  ...  CD55   \n",
       "10  ENST00000314754.12               CD55-201  ENSG00000196352.16  ...  CD55   \n",
       "11   ENST00000367067.8               CD55-205  ENSG00000196352.16  ...  CD55   \n",
       "12   ENST00000367067.8               CD55-205  ENSG00000196352.16  ...  CD55   \n",
       "\n",
       "   rna_support reference_support gene_name alternative_source_transcripts  \\\n",
       "0          N/A              True      CD55                            N/A   \n",
       "1          NaN               NaN      CD55                            N/A   \n",
       "2          NaN               NaN      CD55                            N/A   \n",
       "3          N/A              True      CD55                            N/A   \n",
       "4          N/A              True      CD55                            N/A   \n",
       "5          NaN               NaN      CD55                            N/A   \n",
       "6          NaN               NaN      CD55                            N/A   \n",
       "7          N/A              True      CD55                            N/A   \n",
       "8          NaN               NaN      CD55                            N/A   \n",
       "9          N/A              True      CD55                            N/A   \n",
       "10         NaN               NaN      CD55                            N/A   \n",
       "11         N/A              True      CD55                            N/A   \n",
       "12         NaN               NaN      CD55                            N/A   \n",
       "\n",
       "   collapsed_gene_ids collapsed_gene_names frameshift extra_paralog  \\\n",
       "0                 N/A                  N/A        N/A         False   \n",
       "1                 N/A                  N/A        nan         False   \n",
       "2                 N/A                  N/A        N/A         False   \n",
       "3                 N/A                  N/A        N/A         False   \n",
       "4                 N/A                  N/A        N/A         False   \n",
       "5                 N/A                  N/A        nan         False   \n",
       "6                 N/A                  N/A        nan         False   \n",
       "7                 N/A                  N/A        N/A         False   \n",
       "8                 N/A                  N/A        nan         False   \n",
       "9                 N/A                  N/A        N/A         False   \n",
       "10                N/A                  N/A        nan         False   \n",
       "11                N/A                  N/A        N/A         False   \n",
       "12                N/A                  N/A      False         False   \n",
       "\n",
       "   exon_anotation_support  \n",
       "0                     NaN  \n",
       "1                     NaN  \n",
       "2                     N/A  \n",
       "3                     NaN  \n",
       "4                     NaN  \n",
       "5                     NaN  \n",
       "6                     NaN  \n",
       "7                     NaN  \n",
       "8                     NaN  \n",
       "9                     NaN  \n",
       "10                    NaN  \n",
       "11                    NaN  \n",
       "12                    NaN  \n",
       "\n",
       "[13 rows x 48 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for col in df.columns:\n",
    "    df[col] = df[col].apply(lambda x: x[0] if isinstance(x, list) and len(x) == 1 else x)\n",
    "df.dropna(how='all', axis=1, inplace=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dimelo_v2_modkit_parsing",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
