sudo 
sudo nano hi
ls /mnt/faststorage/
ll /mnt/faststorage/
exit
ls
ls /
ls /tmp/
ls /dev/
ls /
ls /mnt/
ls /mnt/faststorage/
ls /mnt/faststorage/20240802_JL_MR_dimelo_cas9/
ls /mnt/faststorage/20240802_JL_MR_dimelo_cas9/no_sample/
ls /mnt/faststorage/20240802_JL_MR_dimelo_cas9/no_sample/20240802_1737_MN31715_FAZ02208_6cad213e/
ls /mnt/faststorage/20240802_JL_MR_dimelo_cas9/no_sample/20240802_1737_MN31715_FAZ02208_6cad213e/report_FAZ02208_20240802_1742_6cad213e.html 
ls /mnt/faststorage/20240802_JL_MR_dimelo_cas9/no_sample
scp michalula@128.32.141.251:/mnt/faststorage/20240802_JL_MR_dimelo_cas9/no_sample/20240802_1737_MN31715_FAZ02208_6cad213e/report_FAZ02208_20240802_1742_6cad213e.html   
ls
ls /mnt/faststorage/20240802_JL_MR_dimelo_cas9/no_sample
ls /mnt/faststorage/20240802_JL_MR_dimelo_cas9/no_sample/20240802_1737_MN31715_FAZ02208_6cad213e/
ls /mnt/faststorage/20240802_JL_MR_dimelo_cas9/no_sample/20240802_1737_MN31715_FAZ02208_6cad213e/bam_pass/
ls
dh
df
ls
df .
pwd
ls /home
ls /home/jmarcus/
sudo ls /home/julianalee96/
sudo ls /home/jmarcus/
ls
ls 
ls /mnt/
ls /mnt/faststorage/
df
df /bin
df /usr/
df /usr/bin/
ls /usr/
pwd
df .
~/dorado-0.7.3-linux-x64/bin/dorado demux --kit-name SQK-NBD114.24 --output-dir /mnt/faststorage/20240802_JL_MR_dimelo_cas9/dorado_demux_bam_pass /mnt/faststorage/20240802_JL_MR_dimelo_cas9/bam_pass
sudo ~/dorado-0.7.3-linux-x64/bin/dorado demux --kit-name SQK-NBD114.24 --output-dir /mnt/faststorage/20240802_JL_MR_dimelo_cas9/dorado_demux_bam_pass /mnt/faststorage/20240802_JL_MR_dimelo_cas9/bam_pass
sudo ~/dorado-0.7.3-linux-x64/bin/dorado demux --kit-name SQK-NBD114.24 --output-dir /mnt/faststorage/20240802_JL_MR_dimelo_cas9/no_sample/20240802_1737_MN31715_FAZ02208_6cad213e/dorado_demux_bam_pass /mnt/faststorage/20240802_JL_MR_dimelo_cas9/no_sample/20240802_1737_MN31715_FAZ02208_6cad213e/bam_pass
ls
sudo ~/dorado-0.7.3-linux-x64/bin/dorado demux --verbose --kit-name SQK-NBD114.24 --output-dir /mnt/faststorage/20240802_JL_MR_dimelo_cas9/no_sample/20240802_1737_MN31715_FAZ02208_6cad213e/dorado_demux_bam_pass /mnt/faststorage/20240802_JL_MR_dimelo_cas9/no_sample/20240802_1737_MN31715_FAZ02208_6cad213e/bam_pass
sudo ~/dorado-0.7.3-linux-x64/bin/dorado demux --verbose --kit-name SQK-NBD114-24 --output-dir /mnt/faststorage/20240802_JL_MR_dimelo_cas9/no_sample/20240802_1737_MN31715_FAZ02208_6cad213e/dorado_demux_bam_pass /mnt/faststorage/20240802_JL_MR_dimelo_cas9/no_sample/20240802_1737_MN31715_FAZ02208_6cad213e/bam_pass
exit
ls
exit
ls
exit
ls
wget https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
ls
rm -r hg38.fa.gz 
ls
cd ref_genomes/
ls
wget https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
ls
gunzip hg38.fa.gz 
ls
screen -ls
screen -r dorado_demux_mir
screen -r 3540079
screen -d -r dorado_demux_mir 
ls
screen -r dorado_demux_mir
pwd
ls
ls -lha
exit
pwd
sudo  chown root /mnt/faststorage/
sudo chown -R root /mnt/faststorage/20240802_JL_MR_dimelo_cas9/
ll /mnt/faststorage/20240802_JL_MR_dimelo_cas9/
~/dorado-0.7.3-linux-x64/bin/dorado 
~/dorado-0.7.3-linux-x64/bin/dorado demux
~/dorado-0.7.3-linux-x64/bin/dorado demux --kit-name SQK-NBD114.24 --output-dir /mnt/faststorage/20240802_JL_MR_dimelo_cas9/dorado_demux_bam_pass /mnt/faststorage/20240802_JL_MR_dimelo_cas9/bam_pass
tmux -a 
screen -S dorado_demux_mir
screen -ls
screen -S dorado_demux_mir
screen -ls
screen -r dorado_demux_mir 
screen -ls
cd /mnt/faststorage/20240802_JL_MR_dimelo_cas9/
ls
ll dorado_demux_bam_pass/
ls
rm -rf dorado_demux_bam_pass/
sudo rm -rf dorado_demux_bam_pass/
cd no_sample/20240802_1737_MN31715_FAZ02208_6cad213e/
l
ls
du -h -d 1 .
ll dorado_demux_bam_pass/
ls -lh dorado_demux_bam_pass/
ls
cd dorado_demux_bam_pass/
ls
mkdir mishas_reads_barcodes_20to24_unclass
sudo mkdir mishas_reads_barcodes_20to24_unclass
ls
cp -r SQK-NBD114-24_barcode2* ./mishas_reads_barcodes_20to24_unclass/
sudo cp -r SQK-NBD114-24_barcode2* ./mishas_reads_barcodes_20to24_unclass/
sudo cp  unclassified.bam ./mishas_reads_barcodes_20to24_unclass/unclassified.bam
ls
cd mishas_reads_barcodes_20to24_unclass/
ls
sudo dorado trim . > trimmed_ALL.bam
pwd
sudo chown michalula .
~ls -l .
ls -l .
ls -l ..
sudo chown michalula ..
ls -l ..
sudo chown -r michalula ..
sudo chown -R michalula ..
ls -l ..
ls ~
ls ~/dorado-0.7.3-linux-x64/
ls ~/dorado-0.7.3-linux-x64/bin/
ls ~/dorado-0.7.3-linux-x64/bin/dorado 
~/dorado-0.7.3-linux-x64/bin/dorado trim . > trimmed_ALL.bam
~/dorado-0.7.3-linux-x64/bin/dorado   trimmed_barcode24.bam
ls
screen -S dorado_demux_mir
htop
nvidia-smi 
ll /mnt/
ll /mnt/faststorage/
df -h
ll /mnt/faststorage/
exit
ls
ls /
ls /mnt
ls /mnt/faststorage/
ls /mnt/faststorage/cas9_R9_promethion_20240927/
ls /mnt/faststorage/cas9_R9_promethion_20240927/no_sample_id/
ls /mnt/faststorage/cas9_R9_promethion_20240927/no_sample_id/20240927_1853_P2S_00066-B_PAW41556_4d3d5fd5/
cd  /mnt/faststorage/cas9_R9_promethion_20240927/no_sample_id/20240927_1853_P2S_00066-B_PAW41556_4d3d5fd5/
dorado
ls
cd ..
ls
cd 20240927_1853_P2S_00066-B_PAW41556_4d3d5fd5/
ls
ls -lah
screen -d -r dorado_align
tmux
screen -d -r dorado_align
screen -S dorado_align
ls
screen -d -r dorado_align 
ls
cd sofware/
git clone https://github.com/vahidAK/NanoMethPhase.git
ls
cd NanoMethPhase
conda env create -f envs/environment.yaml
mkdir -p ~/miniconda3
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O ~/miniconda3/miniconda.sh
bash ~/miniconda3/miniconda.sh -b -u -p ~/miniconda3
rm ~/miniconda3/miniconda.sh
source ~/miniconda3/bin/activate
conda init --all
ls
mkdir data
cd data
ls
ls
cd sofware/
ls
conda env create -f envs/environment.yaml
cd NanoMethPhase/
ls
conda env create -f envs/environment.yaml
conda activate nanomethphase
python nanomethphase.py
ls
screen -S align_hyplotypes
conda activate nanomethphase
ls
cd ref_genomes/
mkdir ./hg19
wget --timestamping 'ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/*'
fg
g
fg
la
cd hg19/
ls
wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_19/GRCh37.p13.genome.fa.gz
ls
unzip GRCh37.p13.genome.fa.gz
gzip -d GRCh37.p13.genome.fa.gz 
ls
sls
ks
ls .
ls 
ls /
ls /mnt/
ls /mnt/faststorage/
ls /mnt/faststorage/cas9_R9_promethion_20240927/
ls /mnt/faststorage/cas9_R9_promethion_20240927/no_sample_id/
ls /mnt/faststorage/cas9_R9_promethion_20240927/no_sample_id/20240927_1853_P2S_00066-B_PAW41556_4d3d5fd5/
ls
ls /mnt/faststorage/cas9_R9_promethion_20240927/no_sample_id/20240927_1853_P2S_00066-B_PAW41556_4d3d5fd5/fast5_pass/
pip install pod5
pod5 convert fast5 --help
;s
ls
cd  /mnt/faststorage/cas9_R9_promethion_20240927/no_sample_id/20240927_1853_P2S_00066-B_PAW41556_4d3d5fd5/fast5_pass/
ls
cd ..
ls
mkdir pod_convert
cd /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/pod5_converted
pod5 convert fast5   /mnt/faststorage/cas9_R9_promethion_20240927/no_sample_id/20240927_1853_P2S_00066-B_PAW41556_4d3d5fd5/fast5_pass/* --output 20240927_cas9_R9_promethion_converted_fast5.pod5
ls
screen -d -r dorado_align
cd /home/michalula/ref_genomes/hg19
# Use rsync to download the hg19 chromosome files
rsync -avzP rsync://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/ ~/hg19/
ps aux 
screen -d -r align_hyplotypes 
cd /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/pod5_converted
ls
dorado basecaller dna_r9.4.1_e8_sup 20240927_cas9_R9_promethion_converted_fast5.pod5 > 2024092
7_cas9_R9_promethion_converted_fast_calls.bam
dorado download --model dna_r9.4.1_e8_sup@v3.6
dorado basecaller dna_r9.4.1_e8_sup@v3.6 20240927_cas9_R9_promethion_converted_fast5.pod5 > 2024092
7_cas9_R9_promethion_converted_fast_calls.bam
dorado basecaller dna_r9.4.1_e8_sup@v3.6 20240927_cas9_R9_promethion_converted_fast5.pod5 --kit-name SQK-CS9109 > 2024092
dorado basecaller dna_r9.4.1_e8_sup@v3.6 20240927_cas9_R9_promethion_converted_fast5.pod5 --kit-name SQK-LSK109  > 20240927_cas9_R9_promethion_converted_fast_calls.bam 
dorado basecaller dna_r9.4.1_e8_sup@v3.6 20240927_cas9_R9_promethion_converted_fast5.pod5   > 20240927_cas9_R9_promethion_converted_fast_calls.bam 
ls
cd ..
ls
cd to_hg38/
ls
rm -rf samtools.2361842.1280.*
ls
cd dorado_aligner_noTrim/
ls
rm -r\
rm -rf PAW41556_pass_4d3d5fd5_e703d20e_*
ls
pwd
cd /
ls
cd /mnt/
ls
cd /mnt/faststorage/cas9_R9_promethion_20240927/no_sample_id/20240927_1853_P2S_00066-B_PAW41556_4d3d5fd5/
ls
cd /mnt/faststorage/cas9_R9_promethion_20240927/no_sample_id/20240927_1853_P2S_00066-B_PAW41556_4d3d5fd5/report_PAW41556_20240927_1856_4d3d5fd5.html
screen -d -r dorado_align
screen -d -r align_hyplotypes 
conda create -n whatshap-env whatshap
conda activate whatshap-env
conda create -n whatshap-env -c conda-forge whatshap
conda activate whatshap-env
conda create -n whatshap-env 
conda activate whatshap-env
pip install whatshap
whatshap --version
pip install git+https://github.com/whatshap/whatshap
whatshap --version
whatshap polyphase /home/michalula/ref_genomes/K562/ENCFF960SSF.vcf   /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_hg19/sort_aligned_hg19_trim_bam_pass_merged.bam --ploidy p --reference /home/michalula/ref_genomes/hg19/GRCh37.p13.genome.fa  -o /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_hg19/hyplotyped/whatshap_output.vcf
whatshap polyphase /home/michalula/ref_genomes/K562/ENCFF960SSF.vcf   /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_hg19/sort_aligned_hg19_trim_bam_pass_merged.bam --ploidy 4 --reference /home/michalula/ref_genomes/hg19/GRCh37.p13.genome.fa  -o /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_hg19/hyplotyped/whatshap_output.vcf
samtools faidx /home/michalula/ref_genomes/hg19/GRCh37.p13.genome.fa
whatshap polyphase /home/michalula/ref_genomes/K562/ENCFF960SSF.vcf   /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_hg19/sort_aligned_hg19_trim_bam_pass_merged.bam --ploidy 4 --reference /home/michalula/ref_genomes/hg19/GRCh37.p13.genome.fa  -o /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_hg19/hyplotyped/whatshap_output.vcf
dorado summary /home/michalula/data/cas9_nanopore/data/20221114_cas9_T_primes/exp3_mod_mappings.20221114_0044_MN36507_FAS72302_148f7cf6.sorted.bam > summary_t2tv10_20221114_cas9_T_primes/exp3_mod_mappings.20221114_0044_MN36507_FAS72302_148f7cf6.sorted.tsv
dorado summary /home/michalula/data/cas9_nanopore/data/20221114_cas9_T_primes/exp3_mod_mappings.20221114_0044_MN36507_FAS72302_148f7cf6.sorted.bam > summary_t2tv10_20221114_cas9_T_primes.exp3_mod_mappings.20221114_0044_MN36507_FAS72302_148f7cf6.sorted.tsv
pwd
ls ./summary_t2tv10_20221114_cas9_T_primes.exp3_mod_mappings.20221114_0044_MN36507_FAS72302_148f7cf6.sorted.tsv 
pwd
ls /home/michalula/summary_t2tv10_20221114_cas9_T_primes.exp3_mod_mappings.20221114_0044_MN36507_FAS72302_148f7cf6.sorted.tsv 
cd /home/michalula/data/cas9_nanopore/data/20221114_cas9_T_primes
mv  /home/michalula/summary_t2tv10_20221114_cas9
_T_primes.exp3_mod_mappings.20221114_0044_MN36507_FAS72302_148f7cf6.sorted.tsv 
/home/michalula/summary_t2tv10_20221114_cas9_T_primes.exp3_mod_mappings.20221114_0044_MN36507_FAS72302_148f7cf6.sorted.tsv ./summary_t2tv10_20221114_cas9_T_primes.exp3_mod_mappings.20221114_0044_MN36507_FAS72302_148f7cf6.sorted.tsv
mv /home/michalula/summary_t2tv10_20221114_cas9_T_primes.exp3_mod_mappings.20221114_0044_MN36507_FAS72302_148f7cf6.sorted.tsv ./summary_t2tv10_20221114_cas9_T_primes.exp3_mod_mappings.20221114_0044_MN36507_FAS72302_148f7cf6.sorted.tsv
conda activate whatshap-env
whatshap --version
conda config --add channels bioconda
conda config --add channels conda-forge
conda config --set channel_priority strict
conda config --add channels bioconda
conda create -n whatshap-env whatshap
conda activate whatshap-env               
whatshap --version
conda deactivate  
conda create -n whatshap-env whatshap
conda activate whatshap-env
whatshap --version
pip install dataclasses
whatshap --version
whatshap polyphase /home/michalula/ref_genomes/K562/ENCFF960SSF.vcf   /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_hg19/sort_aligned_hg19_trim_bam_pass_merged.bam --ploidy p --reference /home/michalula/ref_genomes/hg19/GRCh37.p13.genome.fa  -o /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_hg19/hyplotyped/whatshap_output.vcf
whatshap phase /home/michalula/ref_genomes/K562/ENCFF960SSF.vcf   /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_hg19/sort_aligned_hg19_trim_bam_pass_merged.bam --ploidy p --reference /home/michalula/ref_genomes/hg19/GRCh37.p13.genome.fa  -o /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_hg19/hyplotyped/whatshap_output.vcf
whatshap stats /home/michalula/ref_genomes/K562/ENCFF960SSF.vcf
whatshap polyphase /home/michalula/ref_genomes/K562/ENCFF960SSF.vcf   /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_hg19/sort_aligned_hg19_trim_bam_pass_merged.bam --ploidy p --reference /home/michalula/ref_genomes/hg19/GRCh37.p13.genome.fa  -o /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_hg19/hyplotyped/whatshap_output.vcf
whatshap --version
pip install --upgrade whatshap
ls
aws s3 sync . s3://streets-lab-data/michalula/nanoporedata/20240927_K562_cas9_CD55_R9_promethiom
aws config
aws configure
aws s3 sync . s3://streets-lab-data/michalula/nanoporedata/20240927_K562_cas9_CD55_R9_promethiom
sudo pip install megalodon==2.3.4
megalodon
cd /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_t2t_v1_1
gunzip ./chm13.draft_v1.1.fasta.gz 
screen -S align_hyplotypes
screen -ls
screen -x 2509253
screen -ls
screen -x 2485222
screen -ls
screen -x 2325723
lx
ls
ls ./data/
ls ./data/ref_genomes/
ls ./data/ref_genomes/K562/
ls ./data/ref_genomes/K562/WGS/
ls ./data/ref_genomes/K562/
ls ./data/ref_genomes/
ls ./data/ref_genomes/K562/WGS/ -lsh
ls ./data/ref_genomes/K562/WGS/SRR30591547/ -lsh
ls ./data/ref_genomes/K562/WGS/ -lsh
ls ./data/ref_genomes/K562/WGS/SRR30591547/ -lsh
ls ./data/ref_genomes/K562/WGS/ -lsh
ls ./data/ref_genomes/K562/WGS/SRR30591547/ -lsh
ls
cd /home/michalula/data/ref_genomes/K562/WGS
screen -S download_K562_separate
dh
df -h --total
ls
cd data/
ls
cd ref_genomes/
wget --output-document sratoolkit.tar.gz https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-ubuntu64.tar.gz
tar -vxzf sratoolkit.tar.gz
ls
export PATH=$PATH:$PWD/sratoolkit.3.1.1-ubuntu64
which fastq-dump
export PATH=$PATH:$PWD/sratoolkit.3.1.1-ubuntu64 
which fastq-dump
ls -lah
export PATH=$PATH:$PWD/sratoolkit.3.1.1-ubuntu64 
echo $PATH
export PATH=$PATH:$PWD/sratoolkit.3.1.1-ubuntu64 
echo $PATH
/home/michalula/.vscode-server/bin/8b3775030ed1a69b13e4f4c628c612102e30a681/bin/remote-cli:/home/michalula/.local/bin:/home/michalula/miniconda3/bin:/home/michalula/miniconda3/condabin:/home/michalula/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/michalula/data/ref_genomes/sratoolkit.3.1.1-ubuntu64:/home/michalula/data/ref_genomes/sratoolkit.3.1.1-ubuntu64
export PATH=/home/michalula/.vscode-server/bin/8b3775030ed1a69b13e4f4c628c612102e30a681/bin/remote-cli:/home/michalula/.local/bin:/home/michalula/miniconda3/bin:/home/michalula/miniconda3/condabin:/home/michalula/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/michalula/data/ref_genomes/sratoolkit.3.1.1-ubuntu64:/home/michalula/data/ref_genomes/sratoolkit.3.1.1-ubuntu64
echo $PATH
which fastq-dump
export PATH=/home/michalula/.vscode-server/bin/8b3775030ed1a69b13e4f4c628c612102e30a681/bin/remote-cli:/home/michalula/.local/bin:/home/michalula/miniconda3/bin:/home/michalula/miniconda3/condabin:/home/michalula/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/michalula/data/ref_genomes/sratoolkit.3.1.1-ubuntu64:/home/michalula/data/ref_genomes/sratoolkit.3.1.1-ubuntu64/bin
which fastq-dump
screen -ls
screen -S k562_refs
screen -ls
screen -x 2325723
screen -ls
screen -x 2535953
screen -ls
screen -x 2536246
screen -ls
screen -x 2535953.k562_refs
screen -x 2325723.dorado_align
df -h --total
screen -ls
screen -x 2509253.align_hyplotypes
screen -x 2535953.k562_refs
screen -ls
screen -x 2509253.align_hyplotypes
screen -ls
screen -x 2325723.dorado_align
screen -ls
screen -x 2509253.align_hyplotypes
screen -ls
screen -x 2535953.k562_refs
screen -ls
screen -x 2509253.align_hyplotypes
screen -x 2325723.dorado_align
screen -x 2535953.k562_refs
screen -ls
screen -x  2535953.k562_refs
screen -ls
screen -ls
screen -x 2325723.dorado_align
screen -ls
screen -x 2325723.dorado_align
screen -x  2535953.k562_refs
screen -ls
screen -x 2509253.align_hyplotypes
screen -ls
screen -x  2535953.k562_refs
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
screen -ls
screen -ls 2325723
screen -ls 2325723.dorado_align
screen -x 2325723.dorado_align
screen -ls
screen -x  2536246.download_K562_separate
screen -x 2325723.dorado_align
echo $PATH
conda
ls
cd sofware/
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
bash Miniconda3-latest-Linux-x86_64.sh
conda
conda init
source ~/.bashrc
conda
df -T 
df 
df .
screen -x 2325723.dorado_align
cd /home/michalula/data/cas9_nanopore/data/20231019_cas9_12ef_Jurkat/no_mC_to_t2t_v1_1
screen -ls
screen -x 2485222.align_hyplotypes
conda env list
nanoph
cd ~/sofware/
ls
cd NanoMethPhase/
ls
conda activate nanomethphase
ls
conda env create -f envs/environment.yaml
conda activate whatshap-env
whatshap --version
conda config --add channels bioconda
conda config --add channels conda-forge
conda config --set channel_priority strict
conda config --add channels bioconda
conda create -n whatshap-env whatshap
conda activate whatshap-env  
whatshap stats 
pip install medaka
medaka_consensus
cd /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_hg19/hyplotyped
cd medaka/
conda create -n medaka -c conda-forge -c nanoporetech -c bioconda medaka
medaka_consensus
source ~/.bashrc
conda
cd sofware/
bash Miniconda3-latest-Linux-x86_64.sh
source ~/.bashrc
grep -o 'chr'  ./summary_aligned_dorado_k30_t2t_v1_1_trim_bam_pass_merged.tsv | wc -l
grep -o 'chr'   /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_t2t_v1_1/more_mismatches/aligned_k30_t2t_v1_1_trim_bam_pass_merged/summary_aligned_dorado_k30_t2t_v1_1_trim_bam_pass_merged.tsv | wc -l
samtools view -c -F 4  /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_t2t_v1_1/more_mismatches/aligned_k5_t2t_v1_1_trim_bam_pass_merged/summary_aligned_dorado_k5_t2t_v1_1_trim_bam_pass_merged.tsv  chr1:206,569,827-206,612,634
samtools view -c -F 4  /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_t2t_v1_1/more_mismatches/aligned_k5_t2t_v1_1_trim_bam_pass_merged/summary_aligned_dorado_k5_t2t_v1_1_trim_bam_pass_merged.tsv  1:206,569,827-206,612,634
cd /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_t2t_v1_1/more_mismatches/aligned_k5_t2t_v1_1_trim_bam_pass_merged/
samtools view -c -F 4     ./sort_aligned_dorado_k5_t2t_v1_1_trim_bam_pass_merged.bam  1:206,569,827-206,612,634
samtools view -c -F 4     ./sort_aligned_dorado_k5_t2t_v1_1_trim_bam_pass_merged.bam  chr1:206,569,827-206,612,634
samtools view -c -F 4     /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_t2t_v1_1/more_mismatches/aligned_k10_t2t_v1_1_trim_bam_pass_merged/sort_aligned_dorado_k10_w10_t2t_v1_1_trim_bam_pass_merged.bam   chr1:206,569,827-206,612,634
cd /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_t2t_v1_1/more_mismatches/aligned_k30_t2t_v1_1_trim_bam_pass_merged
samtools view -c -F 4     ./aligned_dorado_k30_t2t_v1_1_trim_bam_pass_merged  chr1:206,569,827-206,612,634
samtools view -c -F 4     ./aligned_dorado_k30_t2t_v1_1_trim_bam_pass_merged.bam  chr1:206,569,827-206,612,634
cd /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_t2t_v1_1/more_mismatches/aligned_k30_t2t_v1_1_trim_bam_pass_merged
samtools view -c -F 4     ./aligned_dorado_k30_t2t_v1_1_trim_bam_pass_merged.bam  chr1:206569827-206612634
samtools view -c -F 4     ./sort_aligned_dorado_k30_t2t_v1_1_trim_bam_pass_merged.bam  chr1:206569827-206612634
samtools view -c -F 4     ./sort_aligned_dorado_k16_t2t_v1_1_trim_bam_pass_merged.bam   chr1:206,569,827-206,612,634
cd /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_t2t_v1_1/more_mismatches/aligned_k16_t2t_v1_1_trim_bam_pass_merged
samtools view -c -F 4     ./sort_aligned_dorado_k16_t2t_v1_1_trim_bam_pass_merged.bam   chr1:206,569,827-206,612,634
conda
source miniconda3/bin/activate
source miniconda3_backup/bin/activate
cd sofware/
conda info
conda env list
conda init
conda activate whatshap-env
conda activate dimelo_v2dimelo_v2_modkit_parsing
conda activate dimelo_v2_modkit_parsing
python
python -m ipykernel install --user --name=dimelo_v2_modkit_parsing --display-name "Python dimelo_v2_modkit_parsing"
conda activate nanomethphase
cd NanoMethPhase/
conda env create -f envs/environment.yaml
ls
ls /
ls /mnt/
ls /mnt/faststorage/
ls -lha  /mnt/faststorage/
ls -lha  /mnt/faststorage/20241127_MR_Cas9_K562unedited_day2_R9min/
ls -lha  /mnt/faststorage/20241127_MR_Cas9_K562unedited_day2_R9min/no_sample_id/
ls -lha  /mnt/faststorage/20241127_MR_Cas9_K562unedited_day2_R9min/no_sample_id/20241127_1920_MN31715_FAY94709_404bfff1/
ls -lha  /mnt/faststorage/20241127_MR_Cas9_K562unedited_day2_R9min/no_sample_id/20241127_1920_MN31715_FAY94709_404bfff1/bam_pass/
dorado
ls -lha  /mnt/faststorage/20241127_MR_Cas9_K562unedited_day2_R9min/no_sample_id/20241127_1920_MN31715_FAY94709_404bfff1/bam_pass/
cd /home/michalula/data/cas9_nanopore/data/20241127_Cas9_K562unedited_day2_R9_minion
dorado trim  /mnt/faststorage/20241127_MR_Cas9_K562unedited_day2_R9min/no_sample_id/20241127_1920_MN31715_FAY94709_404bfff1/bam_pass/*.bam     >  ./trim_20241127_Cas9_K562unedited_day2_R9_minion.fastbasecalling_while_seq.bam     
dorado trim  /mnt/faststorage/20241127_MR_Cas9_K562unedited_day2_R9min/no_sample_id/20241127_1920_MN31715_FAY94709_404bfff1/bam_pass/     >  ./trim_20241127_Cas9_K562unedited_day2_R9_minion.fastbasecalling_while_seq.bam     
samtools merge  -o ./merged_20241127_Cas9_K562unedited_day2_R9_minion.fastbasecalling_while_seq.bam /mnt/faststorage/20241127_MR_Cas9_K562unedited_day2_R9min/no_sample_id/20241127_1920_MN31715_FAY94709_404bfff1/bam_pass/*.bam
dorado trim   ./merged_20241127_Cas9_K562unedited_day2_R9_minion.fastbasecalling_while_seq.bam    >  ./trim_20241127_Cas9_K562unedited_day2_R9_minion.fastbasecalling_while_seq.bam     
dorado aligner  /home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta  ./trim_20241127_Cas9_K562unedited_day2_R9_minion.fastbasecalling_while_seq.bam    > ./to_t2t_v1_1/align_trim_20241127_Cas9_K562unedited_day2_R9_minion.fastbasecalling_while_seq.bam 
dorado summary   ./to_t2t_v1_1/align_t2t_v1_1_trim_20241127_Cas9_K562unedited_day2_R9_minion.fastbasecalling_while_seq.bam   > ./to_t2t_v1_1/summary_align_t2t_v1_1_trim_20241127_Cas9_K562unedited_day2_R9_minion.fastbasecalling_while_seq.bam   
samtools sort   -o ./to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241127_Cas9_K562unedited_day2_R9_minion.fastbasecalling_while_seq.bam    ./to_t2t_v1_1/align_t2t_v1_1_trim_20241127_Cas9_K562unedited_day2_R9_minion.fastbasecalling_while_seq.bam  
samtools index   ./to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241127_Cas9_K562unedited_day2_R9_minion.fastbasecalling_while_seq.bam   
ls -lha  /mnt/faststorage/20241127_MR_Cas9_K562unedited_day2_R9min/no_sample_id/20241127_1920_MN31715_FAY94709_404bfff1/
ls -lha  /mnt/faststorage/20241127_MR_Cas9_K562unedited_day2_R9min/no_sample_id/20241127_1920_MN31715_FAY94709_404bfff1/fast5
ls -lha  /mnt/faststorage/20241127_MR_Cas9_K562unedited_day2_R9min/no_sample_id/20241127_1920_MN31715_FAY94709_404bfff1/fast5_pass/
ls -lha  /mnt/faststorage/20241127_MR_Cas9_K562unedited_day2_R9min/no_sample_id/20241127_1920_MN31715_FAY94709_404bfff1/fast5_skip/
cd  /home/michalula/data/cas9_nanopore/data/20241127_Cas9_K562unedited_day2_R9_minion/post_seq_basecall
dorado basecaller    /home/michalula/sofware/dna_r9.4.1_e8_sup@v3.3	     ./20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.pod5     > ./20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mC.bam   
dorado trim   ./20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mC.bam       >  ./trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mC.bam       
dorado aligner  /home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta  ./trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mC.bam       > ./to_t2t_v1_1/align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mC.bam   
samtools sort   -o ./to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mC.bam       ./to_t2t_v1_1/align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mC.bam   
samtools index   ./to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mC.bam
dorado summary   ./to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mC.bam   > ./to_t2t_v1_1/summary_sort_align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mC.tsv
ls\
samtools sort   -o ./to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mC.bam       ./to_t2t_v1_1/align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mC.bam   
ls -lah
cd to_t2t_v1_1/
ls -lah
cd /home/michalula/data/cas9_nanopore/data/20241127_Cas9_K562unedited_day2_R9_minion/post_seq_basecall/5mCG_5hmCG
dorado basecaller    /home/michalula/sofware/dna_r9.4.1_e8_sup@v3.3	     ./20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.pod5     --modified-bases 5mCG_5hmCG     > ./20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG_5hmCG.bam   
dorado basecaller    /home/michalula/sofware/dna_r9.4.1_e8_sup@v3.3	     ../20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.pod5     --modified-bases 5mCG_5hmCG     > ./20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG_5hmCG.bam   
cd ..
ls
dorado basecaller    /home/michalula/sofware/dna_r9.4.1_e8_sup@v3.3	     ./20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.pod5     --modified-bases 5mCG     > ./5mCG/20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
/home/michalula/data/cas9_nanopore/data/20241127_Cas9_K562unedited_day2_R9_minion/post_seq_basecall/5mCG
cd /home/michalula/data/cas9_nanopore/data/20241127_Cas9_K562unedited_day2_R9_minion/post_seq_basecall/5mCG
ls
dorado trim   ./20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       >  ./trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       
dorado aligner  /home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta  ./trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       > ./to_t2t_v1_1/align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
ls
dorado aligner  /home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta  ./trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       > ./to_t2t_v1_1/align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
samtools sort   -o ./to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       ./to_t2t_v1_1/align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
samtools index   ./to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam
dorado summary   ./to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   > ./to_t2t_v1_1/summary_sort_align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.tsv
cd /home/michalula/data/cas9_nanopore/data/20241127_Cas9_K562unedited_day2_R9_minion/post_seq_basecall/5mCG_5hmCG
dorado trim   ./20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG_5hmCG.bam       >  ./trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG_5hmCG.bam       
dorado aligner  /home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta  ./trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG_5hmCG.bam       > ./to_t2t_v1_1/align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG_5hmCG.bam   
dorado aligner  /home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta  ./trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG_5hmCG.bam       > ./to_t2t_v1_1/align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG_5hmCG.bam   
samtools sort   -o ./to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG_5hmCG.bam       ./to_t2t_v1_1/align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG_5hmCG.bam   
samtools index   ./to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG_5hmCG.bam
dorado summary   ./to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG_5hmCG.bam   > ./to_t2t_v1_1/summary_sort_align_t2t_v1_1_trim_20241127_MR_Cas9_K562unedited_day2_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG_5hmCG.tsv
cd   /mnt/faststorage/20241127_MR_Cas9_K562unedited_day2_R9min/no_sample_id/20241127_1920_MN31715_FAY94709_404bfff1/
ls
ls -lah
aws 
aws config
aws configure
screen  -ls
screen  -S load_data
htop
ls -lha  
cd ..
aws s3 sync . s3://streets-lab-data/michalula/nanoporedata/20241127_MR_Cas9_K562unedited_day2_R9min
htop
cd /mnt/faststorage/
ls
cd 20241226_MR_nCATs_TcellsPrES_unedit_P2R9/
ls
cd no_sample_id/
ls
ls 20241227_0021_P2S_00066-A_PAW35660_fc74d3ec/
aws s3 sync . s3://streets-lab-data/michalula/nanoporedata/20241226_MR_nCATs_TcellsPrES_unedit_P2R9/
screen -S basecall_Tcells_unedit
pod5 convert fast5  /home/michalula/data/cas9_nanopore/data/20241226_nCATs_K562_ZFPOFFpostSort_LOW70/* --output /home/michalula/data/cas9_nanopore/data/20241226_nCATs_K562_ZFPOFFpostSort_LOW70/pod5_converted_basecall/20241226_nCATs_K562_ZFPOFFpostSort_LOW70.pod5
cd /home/michalula/data/cas9_nanopore/data/20241226_nCATs_K562_ZFPOFFpostSort_LOW70/pod5_converted_basecall/
dorado basecaller    /home/michalula/sofware/dna_r9.4.1_e8_sup@v3.3	     ./20241226_nCATs_K562_ZFPOFFpostSort_LOW70.pod5     --modified-bases 5mCG     > ./5mCG/20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
screen ls
screen -S basecall
screen -ls
screen -x 575863.basecall
screen ls
screen -ls
screen -x 829007.basecall_Tcells_unedit
screen -ls
screen -x 575863.basecall
screen -ls
screen -x 829007.basecall_Tcells_unedit
screen -x 575863.basecall
ls /mnt/faststorage/20241226*
ls /mnt/faststorage/20241226_MR_nCATs_TcellsPrES_unedit_P2R9
ls /mnt/faststorage/20241226_MR_nCATs_TcellsPrES_unedit_P2R9/no_sample_id/
ls /mnt/faststorage/20241226_MR_nCATs_TcellsPrES_unedit_P2R9/no_sample_id/20241227_0021_P2S_00066-A_PAW35660_fc74d3ec/
ls /mnt/faststorage/20241226_MR_nCATs_TcellsPrES_unedit_P2R9/no_sample_id/20241227_0021_P2S_00066-A_PAW35660_fc74d3ec/fast5_fail/
ls /mnt/faststorage/20241226_MR_nCATs_TcellsPrES_unedit_P2R9/no_sample_id/20241227_0021_P2S_00066-A_PAW35660_fc74d3ec/fast5_pass/
pod5 convert fast5 /mnt/faststorage/20241226_MR_nCATs_TcellsPrES_unedit_P2R9/no_sample_id/20241227_0021_P2S_00066-A_PAW35660_fc74d3ec/fast5_pass/* --output /home/michalula/data/cas9_nanopore/data/20241226_nCATS_K562_ZFPOFFpostSort_HIGH/pod5_converted_basecall/20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.pod5
ls
dorado trim   ./5mCG/20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       >  ./5mCG/trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       
mkdir /home/michalula/data/ref_genomes/to_t2t_v1_1
dorado aligner  /home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta  ./5mCG/trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       > ./5mCG/to_t2t_v1_1/align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
[alignment took 1 mins]
samtools sort   -o ./5mCG/to_t2t_v1_1/sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       ./5mCG/to_t2t_v1_1/align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
samtools index   ./5mCG/to_t2t_v1_1/sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam
dorado summary   ./5mCG/to_t2t_v1_1/sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   > ./5mCG/to_t2t_v1_1/summary_sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.tsv
ls
ls 5mCG/
mkdir ./5mCG/to_t2t_v1_1
dorado aligner  /home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta  ./5mCG/trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       > ./5mCG/to_t2t_v1_1/align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
samtools sort   -o ./5mCG/to_t2t_v1_1/sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       ./5mCG/to_t2t_v1_1/align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
samtools index   ./5mCG/to_t2t_v1_1/sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam
dorado summary   ./5mCG/to_t2t_v1_1/sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   > ./5mCG/to_t2t_v1_1/summary_sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_passed.dna_r9.4.1_e8_sup@v3.3.5mCG.tsv
ls ./5mCG/to_t2t_v1_1/
ls -lah  ./5mCG/to_t2t_v1_1/ 
ls /mnt/faststorage/20241226_MR_nCATs_TcellsPrES_unedit_P2R9/no_sample_id/20241227_0021_P2S_00066-A_PAW35660_fc74d3ec
ls /mnt/faststorage/20241226_MR_nCATs_TcellsPrES_unedit_P2R9/no_sample_id/20241227_0021_P2S_00066-A_PAW35660_fc74d3ec/fast5_fail/
cd /home/michalula/data/cas9_nanopore/data/20241226_MR_nCATs_TcellsPrES_unedit_P2R9/failed_fast5
pod5 convert fast5 /mnt/faststorage/20241226_MR_nCATs_TcellsPrES_unedit_P2R9/no_sample_id/20241227_0021_P2S_00066-A_PAW35660_fc74d3ec/fast5_fail/* --output ./20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.pod5
#  [basecall took 7 mins]
dorado trim   ./5mCG/20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       >  ./5mCG/trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       
mkdir ./5mCG/to_t2t_v1_1
dorado aligner  /home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta  ./5mCG/trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       > ./5mCG/to_t2t_v1_1/align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
# [alignment took 1 mins]
samtools sort   -o ./5mCG/to_t2t_v1_1/sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       ./5mCG/to_t2t_v1_1/align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
samtools index   ./5mCG/to_t2t_v1_1/sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam
dorado summary   ./5mCG/to_t2t_v1_1/sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   > ./5mCG/to_t2t_v1_1/summary_sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.tsv
# # Failed T cells unedited:
# >> 
# # Passed T cells unedited:
# >> 226441
# # ZFP OFF:
# >> 850 454 
# # >> 55376
# No modification calls:
# >> 55376 
# When basecall on the fly:
# >> 23981
ls
dorado summary   ./5mCG/to_t2t_v1_1/sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   > ./5mCG/to_t2t_v1_1/summary_sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.tsv
pwd
ls
dorado trim   ./5mCG/20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       >  ./5mCG/trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       
mkdir ./5mCG/to_t2t_v1_1
dorado aligner  /home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta  ./5mCG/trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       > ./5mCG/to_t2t_v1_1/align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
# [alignment took 1 mins]
samtools sort   -o ./5mCG/to_t2t_v1_1/sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       ./5mCG/to_t2t_v1_1/align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
samtools index   ./5mCG/to_t2t_v1_1/sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam
dorado summary   ./5mCG/to_t2t_v1_1/sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   > ./5mCG/to_t2t_v1_1/summary_sort_align_trim_20241226_MR_nCATs_TcellsPrES_unedit_P2R9_failed.dna_r9.4.1_e8_sup@v3.3.5mCG.tsv
ls -lah /home/michalula/data/cas9_nanopore/data/20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9/pod5_converted_basecall/5mCG/to_t2t_v1_1
dorado basecaller    /home/michalula/sofware/dna_r9.4.1_e8_sup@v3.3     ./20241226_nCATs_K562_ZFPOFFpostSort_LOW70.pod5     --modified-bases 5mCG     > ./5mCG/20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
dorado trim   ./20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       >  ./trim_20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       
pwd
ls
dorado trim   ./20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       >  ./trim_20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       
dorado trim   ./5mCG/20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       >  ./5mCG/trim_20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       
dorado aligner  /home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta  ./5mCG/trim_20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       > ./5mCG/to_t2t_v1_1/align_t2t_v1_1_trim_20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
samtools sort   -o ./5mCG/to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       .5mCG/to_t2t_v1_1/align_t2t_v1_1_trim_20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
samtools sort   -o ./5mCG/to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       ./5mCG/to_t2t_v1_1/align_t2t_v1_1_trim_20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
samtools index   ./5mCG/to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam
dorado summary   ./5mCG/to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   > ./5mCG/to_t2t_v1_1/summary_sort_align_t2t_v1_1_trim_20241226_nCATs_K562_ZFPOFFpostSort_LOW70_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.tsv
cd 5mCG/
ls -lah
cd to_t2t_v1_1/
ls
ls -lah
\
cd /home/michalula/data/cas9_nanopore/data/20241226_nCATS_K562_ZFPOFFpostSort_HIGH
pod5 convert fast5  /home/michalula/data/cas9_nanopore/data/20241226_nCATS_K562_ZFPOFFpostSort_HIGH/fast5/* --output /home/michalula/data/cas9_nanopore/data/20241226_nCATS_K562_ZFPOFFpostSort_HIGH/pod5_converted_basecall/20241226_nCATS_K562_ZFPOFFpostSort_HIGH.pod5
ls
dorado basecaller    /home/michalula/sofware/dna_r9.4.1_e8_sup@v3.3     ./pod5_converted_basecall/20241226_nCATS_K562_ZFPOFFpostSort_HIGH.pod5     --modified-bases 5mCG     > ./5mCG/20241226_nCATS_K562_ZFPOFFpostSort_HIGH_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
pwd
ls
dorado basecaller    /home/michalula/sofware/dna_r9.4.1_e8_sup@v3.3     ./20241226_nCATS_K562_ZFPOFFpostSort_HIGH.pod5     --modified-bases 5mCG     > ./5mCG/2024122pws6_nCATS_K562_ZFPOFFpostSort_HIGH_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
dorado trim   ./5mCG/2024122pws6_nCATS_K562_ZFPOFFpostSort_HIGH_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       >  ./5mCG/trim_2024122pws6_nCATS_K562_ZFPOFFpostSort_HIGH_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       
dorado aligner  /home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta  ./5mCG/trim_2024122pws6_nCATS_K562_ZFPOFFpostSort_HIGH_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       > ./5mCG/to_t2t_v1_1/align_t2t_v1_1_trim_2024122pws6_nCATS_K562_ZFPOFFpostSort_HIGH_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
dorado aligner  /home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta  ./5mCG/trim_2024122pws6_nCATS_K562_ZFPOFFpostSort_HIGH_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       > ./5mCG/to_t2t_v1_1/align_t2t_v1_1_trim_2024122pws6_nCATS_K562_ZFPOFFpostSort_HIGH_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
samtools sort   -o ./5mCG/to_t2t_v1_1/sort_align_t2t_v1_1_trim_2024122pws6_nCATS_K562_ZFPOFFpostSort_HIGH_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       ./5mCG/to_t2t_v1_1/align_t2t_v1_1_trim_2024122pws6_nCATS_K562_ZFPOFFpostSort_HIGH_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
samtools index   ./5mCG/to_t2t_v1_1/sort_align_t2t_v1_1_trim_2024122pws6_nCATS_K562_ZFPOFFpostSort_HIGH_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam
dorado summary   ./5mCG/to_t2t_v1_1/sort_align_t2t_v1_1_trim_2024122pws6_nCATS_K562_ZFPOFFpostSort_HIGH_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   > ./5mCG/to_t2t_v1_1/summary_sort_align_t2t_v1_1_trim_2024122pws6_nCATS_K562_ZFPOFFpostSort_HIGH_R9min_converted_fast5.dna_r9.4.1_e8_sup@v3.3.5mCG.tsv
ls -lah ./5mCG/to_t2t_v1_1/
samtools view -c -F 4     /home/michalula/data/cas9_nanopore/data/2024927_Cas9_R9_promethion_K562/to_t2t_v1_1/more_mismatches/aligned_k10_t2t_v1_1_trim_bam_pass_merged/sort_aligned_dorado_k10_w10_t2t_v1_1_trim_bam_pass_merged.bam   chr1:206,569,827-206,612,634
ls /mnt/faststorage/
ls /mnt/faststorage/20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9
ls /mnt/faststorage/20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9/no_sample_id/
ls /mnt/faststorage/20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9/no_sample_id/20241227_0046_P2S_00066-B_PAW39661_4e67d296/
ls /mnt/faststorage/20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9/no_sample_id/20241227_0046_P2S_00066-B_PAW39661_4e67d296/fast5_pass
ls /mnt/faststorage/20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9/no_sample_id/20241227_0046_P2S_00066-B_PAW39661_4e67d296/
pod5 convert fast5 /mnt/faststorage/20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9/no_sample_id/20241227_0046_P2S_00066-B_PAW39661_4e67d296/fast5_pass/* --output /home/michalula/data/cas9_nanopore/data/20241226_nCATS_K562_ZFPOFFpostSort_HIGH/pod5_converted_basecall/20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9.pod5
ls
cd /home/michalula/data/cas9_nanopore/data/20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9/pod5_converted_basecall
ls
pwd
ls
ls -lah
pod5 convert fast5 /mnt/faststorage/20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9/no_sample_id/20241227_0046_P2S_00066-B_PAW39661_4e67d296/fast5_pass/* --output /home/michalula/data/cas9_nanopore/data/20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9/pod5_converted_basecall/20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9.pod5
ls
pwd
dorado trim   ./5mCG/20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       >  ./5mCG/trim_20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       
mkdir ./5mCG/to_t2t_v1_1
dorado aligner  /home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta  ./5mCG/trim_20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       > ./5mCG/to_t2t_v1_1/align_t2t_v1_1_trim_20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
# [alignment took 1 mins]
samtools sort   -o ./5mCG/to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam       ./5mCG/to_t2t_v1_1/align_t2t_v1_1_trim_20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   
samtools index   ./5mCG/to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam
dorado summary   ./5mCG/to_t2t_v1_1/sort_align_t2t_v1_1_trim_20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam   > ./5mCG/to_t2t_v1_1/summary_sort_align_t2t_v1_1_trim_20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9.dna_r9.4.1_e8_sup@v3.3.5mCG.tsv
ls -lah ./5mCG/to_t2t_v1_1/
cd /home/michalula/data/cas9_nanopore/data/20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9/pod5_converted_basecall/5mCG/to_t2t_v1_1
pwd
ls -lah
samtools index chr1_sort_align_t2t_v1_1_trim_20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam
ls -lah
ls
cd data/
ls
cd cas9_nanopore/
ls
cd data/
ls
aws s3 cp s3://streets-lab-data/michalula/nanoporedata/20241226_nCATS_K562_ZFPOFFpostSort_HIGH/no_sample_id/20241227_0140_MN36507_FAW26101_7cfd685d/ ./20241226_nCATS_K562_ZFPOFFpostSort_HIGH
aws s3 cp s3://streets-lab-data/michalula/nanoporedata/20241226_nCATS_K562_ZFPOFFpostSort_HIGH/no_sample_id/20241227_0140_MN36507_FAW26101_7cfd685d/ ./20241226_nCATS_K562_ZFPOFFpostSort_HIGH --recursive
aws s3 cp s3://streets-lab-data/michalula/nanoporedata/20241226_nCATs_K562_ZFPOFFpostSort_LOW70/20241227_0115_MN31715_FAY94990_ef7625a1/  ./20241226_nCATs_K562_ZFPOFFpostSort_LOW70  --recursive
ls
cd /mnt/faststorage/
ls
cd 20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9/
ls
cd no_sample_id/
ls
aws s3 sync . s3://streets-lab-data/michalula/nanoporedata/20241226_MR_nCATs_TPrES_CROFF90prD6_P2R9/
\
ls -lah /home/michalula/code/epiCausality
ls -lah /home/michalula/code/epiCausality/single_reads_analysis/
ls -lah /home/michalula/code/epiCausality/single_reads_analysis/K562/
ls -lah /home/michalula/code/epiCausality/single_reads_analysis/K562/unedited_r9P2_20240917/
ls -lah /home/michalula/code/epiCausality/single_reads_analysis/K562/unedited_r9min_20241127/
conda env list
pip install ipykernel jupyter
conda env list
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.2.1/pythonFiles/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.2.1/pythonFiles/deactivate/bash/envVars.txt
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
cd /home/michalula/code/epiCausality
ls
git add .
git commit -m "extract read plots formy  individ cell lines"
git push origin master
git push 
git commit -m "extract read plots formy  individ cell lines"
git push
git remote -v
git push
git pull
apt install libcurl4-openssl-dev
sudo apt install libcurl4-openssl-dev
git pull
git push
ls
git push
git --help
git config --global user.email "michal.rozenwald@gmail.com"
git push origin main
yum install openssl-devel.
git remote remove origin
git commit -m "extract read plots formy  individ cell lines"
git push origin main
git pull
git config --global user.email "michal.rozenwald@gmail.com"
git push origin main
git merge origin main_2
git merge
git push origin main
git merge origin master
git pull
git pull origin main
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
cd /home/michalula/code/epiCausality
ls -lah
pwd
git config pull.ff only
git pull --tags origin main
git config pull.rebase false
git pull --tags origin main
git config pull.rebase false
git pull --rebase --tags origin main
git add .
cd /home/michalula/code/epiCausality/single_reads_analysis/T_primers_ES/Extract_single_reads_2025Jan22.ipynb
cd /home/michalula/code/epiCausality
git add .
git commit -m "build mC  vector from bam functions"
git push
git add .
cd /home/michalula/code/epiCausality
git add .
git commit -m "build mC  vector from bam"
git push
git add .
git commit -m "build mC  vector from bam - working!"
git push
git add .
git commit -m "build mC  vector from bam - working!"
git push
git add .
git commit -m "epiCode with funcs: extract  mC  vector from bam"
git push
git add .
git commit -m "epiCode with funcs: extract  mC  vector from bam"
git push
git add .
git commit -m "epiCode with funcs: extract  mC  vector from bam"
git add .
git commit -m "epiCode with funcs: extract  mC  vector from bam"
git push
git add .
git commit -m "analyze frw vs rvs strands"
git push
git add .
git commit -m "analyze frw vs rvs strands"
git add .
git push
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
git add .
cd /home/michalula/code/epiCausality
git add .
git commit -m "analyze frw vs rvs strands"
git push
git add .
git commit -m "analyze frw vs rvs strands in one func"
git push
git add .
git commit -m "analyze frw vs rvs strands in one func"
git push
git add .
git commit -m "analyze frw vs rvs strands in one func"
git push
cd /home/michalula/code/epiCausality
git add .
git commit -m "analyze frw vs rvs strands in one func"
git push
git add .
git commit -m "analyze frw vs rvs strands in utils  func"
git push
git add .
git commit -m "analyze frw vs rvs strands in utils  func"
git add .
git commit -m "analyze frw vs rvs strands in utils  func"
git push
git add .
git commit -m "silenced vs unedited T primes ROI ES data analyze frw vs rvs strands in utils  func"
git push
git add .
git commit -m "70% vs 90% mC confidance for  silenced vs unedited T primes ROI ES data analyze frw vs rvs strands in utils  func"
git push
/home/michalula/miniconda3_backup/bin/python
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
/home/michalula/miniconda3_backup/bin/python
cd /home/michalula/code/epiCausality
git add .
git commit -m "window size increase attomatically"
git push
git add .
git commit -m "window size increase attomatically"
git push
git add .
git commit -m "input any window size and adjust plotting  attomatically"
git push
git commit -m "input any window size and adjust plotting  attomatically"
git add .
git commit -m "dubUg the deBug for  input any window size and adjust plotting  attomatically"
git push
git commit -m "dubUg the deBug for  input any window size and adjust plotting  attomatically + todo ++ todo"
git add .
git commit -m "dubUg the deBug for  input any window size and adjust plotting  attomatically + todo ++ todo"
git push
git add .
git commit -m "Build ML models silenced vs unedited"
git push
git add .
git commit -m "Build ML models silenced vs unedited"
git push
git add .
git commit -m "Build ML models silenced vs unedited"
git add .
git push
cd /home/michalula/code/epiCausality
git add .
git commit -m "Build ML models silenced vs unedited"
git push
git add .
git commit -m "Extand input window size to Build better  ML models silenced vs unedited"
git push
git add .
git commit -m "Extand input 16 CpGs with 31  window size to Build better  ML models silenced vs unedited"
git push
git add .
git commit -m "Extand input 16 CpGs with 31  window size to Build better  ML models silenced vs unedited"
git commit -m "Extand input 16 CpGs with 331  window size to Build better  ML models silenced vs unedited"
git add .
git commit -m "Extand input 16 CpGs with 331  window size to Build better  ML models silenced vs unedited"
git add .
git commit -m "Extand input 16 CpGs with 331  window size to Build better  ML models silenced vs unedited"
git push
git add .
git commit -m "Extand input 106 CpGs with 3031  window size to Build better  ML models silenced vs unedited"
git add .
git commit -m "93% SUCCESS! Extand input 106 CpGs with 3031  window size to Build better  ML models silenced vs unedited"
git push
git add .
git push
x
git commit -m "93% SUCCESS! Extand input 106 CpGs with 3031  window size to Build better  ML models silenced vs unedited"
git push
git add .
git commit -m "93% SUCCESS! Extand input 106 CpGs with 3031  window size to Build better  ML models silenced vs unedited"
git push
git reset --soft origin/main  # or replace 'main' with your branch name
git push
git add .
git commit -m "93% SUCCESS! Extand input 106 CpGs with 3031  window size to Build better  ML models silenced vs unedited"
git push
git add .
git commit -m "93% SUCCESS! Extand input 142 CpGs with 7KBps  window size to Build better  ML models silenced vs unedited"
git push
git add .
git commit -m "93% SUCCESS! Extand input 142 CpGs with 7KBps  window size to Build better  ML models silenced vs unedited"
git add .
git push
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
cd /home/michalula/code/epiCausality
git add .
git commit -m "K562: extract reads from LOW, High silenced post sort vs unedited"
git reset --soft origin/main  # or replace 'main' with your branch name
git add .
git commit -m "K562: extract reads from LOW, High silenced post sort vs unedited"
git push
git add .
git commit -m "K562: extract reads from ZFPOff LOW silenced post sort vs unedited - ML "
git push
git add .
git commit -m "K562: extract reads from ZFPOff LOW silenced post sort vs unedited - ML "
git add .
cd /home/michalula/code/epiCausality
git add .
git commit -m "K562: extract reads from ZFPOff LOW silenced post sort vs unedited - ML reorg "
git push
git push origin main
git remote -v
git credential reject https://github.com
git push origin main
git credential reject https://github.com
git credential reject
git remote set-url origin git@github.com:MichalRozenwald/epiCausality.git
git push origin main
git fetch
ls ~/.ssh/
ls ~/.ssh/known_hosts 
ssh -T git@github.com
ssh-keygen -t ed25519 -C  michal.rozenwald@gmail.com
eval "$(ssh-agent -s)"
ssh-add ~/.ssh/id_ed25519
cat ~/.ssh/id_ed25519.pub
cat /home/michalula/.ssh/id_ed25519.pub 
git remote set-url origin git@github.com:MichalRozenwald/epiCausality.git
git remote -v
ssh -T git@github.com
git remote set-url origin git@github.com:MichalRozenwald/epiCausality.git
ssh -T git@github.com
git push origin main
git add .
git commit -m "Use the SUM of the mCpG values instead of individual CpG unit profiles"
git push origin main
git add .
git commit -m "Use the SUM of the mCpG values instead of individual CpG unit profiles"
git push 
git add .
git commit -m "Use the SUM of the mCpG values instead of individual CpG unit profiles"
git push 
cd /home/michalula/code/epiCausality
git add .
git commit -m "ML modeling to functions"
git add .
git commit -m "ML modeling to functions"
git push 
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
cd /home/michalula/code/epiCausality/epiCode
git add .
git commit -m "add plots of fractions mCGs"
git push 
git remote -v
git add .
git commit -m "add plots of fractions mCGs"
git push 
git reset --soft HEAD~1
git push 
git reset --soft HEAD~1
git push 
git add .
git commit -m "add plots of fractions mCGs"
git push 
git add .
git commit -m "add plots of fractions mCGs"
git push 
python
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
/home/michalula/miniconda3_backup/bin/python
cd /home/michalula/code/epiCausality/epiCode
git add .
git commit -m "add prettier PLOTLY  plots"
git push 
git add .
git commit -m "discreptancy between matplotlib and  PLOTLY  bar plots"
git add .
git commit -m "FIX  matplotlib bar plots using  snap=False"
git push 
kkk
git add .
git commit -m "keep fully unmethylated reads"
git reset --soft HEAD~1
git add .
git commit -m "keep fully unmethylated reads"
git push 
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
htop
cd /home/michalula/code/epiCausality/epiCode
git add .
git commit -m "discreptancy between matplotlib and  PLOTLY  bar plots"
git push 
htop
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
cd /home/michalula/data
sudo chmod -R a+rwx .
ls -lah
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
ls
ls -ld .
sudo setfacl -m u:oberon:rwx . 
ls -ld .
getfacl .\
ls
pwd
cd /home/michalula/code/epiCausality/epiCode
git add .
git commit -m "mCG plots using  matplotlib and  PLOTLY  bar plots"
git reset --soft HEAD~1
git add .
git commit -m "mCG plots using  matplotlib and  PLOTLY  bar plots"
git push 
git add .
git commit -m "delete extracted CG data - need to generate new when running new clean extractions"
git push 
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2024.3.10742127/python_files/deactivate/bash/envVars.txt
cd /home/michalula/code/epiCausality
git add .
git commit -m "mCG plots using  matplotlib and  PLOTLY  bar plots - Unedited vs. Edited cells freqs WORKING"
git push 
git add .
git commit -m "mCG plots using  matplotlib and  PLOTLY  bar plots - Unedited vs. Edited cells freqs WORKING - add plot titles with EXPERIMENT NAME"
git push 
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2025.3.2025031001-linux-x64/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2025.3.2025031001-linux-x64/python_files/deactivate/bash/envVars.txt
cd /home/michalula/code/epiCausality
git add .
git commit -m "K562 plots of ROI"
git push 
git add .
git commit -m ".gitignore added to skip aploading the padded_reads*.npy"
git push 
git add .
git commit -m "Remove reads that don't have ANY mC present in the ROI"
git push 
git add .
git commit -m "mCG plots x axis size fix"
git push 
/bin/python3 /home/michalula/.vscode-server/extensions/ms-python.python-2025.3.2025031001-linux-x64/python_files/printEnvVariablesToFile.py /home/michalula/.vscode-server/extensions/ms-python.python-2025.3.2025031001-linux-x64/python_files/deactivate/bash/envVars.txt
picard
