{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0",
   "metadata": {},
   "source": [
    "# Plot Filtered Data!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "\n",
    "def current_time():\n",
    "    \"\"\"Returns the current date and time as a formatted string.\"\"\"\n",
    "    return datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\") \n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    print(\"Current Date and Time:\", current_time())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare mC fractures across conditions\n",
    "import sys\n",
    "from pathlib import Path\n",
    "from datetime import datetime\n",
    "from matplotlib import pyplot as plt \n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "# import time\n",
    "# import shap\n",
    "\n",
    "\n",
    "# Add the folder containing 'my_functions.py' to sys.path\n",
    "sys.path.append(\"/home/michalula/code/epiCausality/epiCode/utils/\") # str(Path(__file__).parent / 'utils'))\n",
    "\n",
    "# Import the module or specific functions\n",
    "# from /home/michalula/code/epiCausality/epiCode/utils/funcs_extract_mC_profiles_from_BAMs.py\n",
    "# from funcs_extract_mC_profiles_from_BAMs import system_info, extract_from_bam\n",
    "from funcs_extract_mC_profiles_from_BAMs import (\n",
    "    system_info,\n",
    "    get_reference_sequence,\n",
    "    create_output_directory,\n",
    "    extract_from_bam,\n",
    "    process_extracted_reads,\n",
    "    visualize_data,\n",
    "    create_padded_reads,\n",
    "    plot_padded_reads,\n",
    "    save_padded_reads,\n",
    "    remove_low_methylated_reads,\n",
    "    process_extracted_reads_no_fully_unmethylated,\n",
    "    create_padded_reads_no_fully_unmethylated\n",
    "    # main,\n",
    ")\n",
    "from funcs_analize_forward_reverse_extracted_mC_reads import analize_forward_reverse_CGs_pipeline, load_padded_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fractions(CGs_pd):\n",
    "    CGs_sums = np.nansum(CGs_pd, axis=0) \n",
    "    CGs_fracs = CGs_sums  / len(CGs_pd) \n",
    "    return CGs_fracs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5",
   "metadata": {},
   "source": [
    "# Day 6 post ElectroPoration (EP)\n",
    "## Unedited D6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_folder_path = \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/unedited/analyze_single_reads/dimelo_v2_output/\"\n",
    "save_padded_reads_name_np = save_folder_path + \"CG_137_padded_reads_day6_unedited_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD490_numRVS644.npy\"\n",
    "\n",
    "CGs_D6_unedited_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "\n",
    "CGs_D6_unedited_mc07_fracs = get_fractions(CGs_D6_unedited_mc07)\n",
    "print(\"CGs_D6_unedited_mc07 shape: \", CGs_D6_unedited_mc07.shape)\n",
    "print(\"CGs_D6_unedited_mc07_fracs shape: \", CGs_D6_unedited_mc07_fracs.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_folder_path = \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/unedited/analyze_single_reads/dimelo_v2_output/\"\n",
    "save_padded_reads_name_np = save_folder_path + \"CG_137_padded_reads_day6_unedited_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD489_numRVS638.npy\"\n",
    "CGs_D6_unedited_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "\n",
    "CGs_D6_unedited_mc0995_fracs = get_fractions(CGs_D6_unedited_mc0995)\n",
    "print(\"CGs_D6_unedited_mc0995 shape: \", CGs_D6_unedited_mc0995.shape)\n",
    "print(\"CGs_D6_unedited_mc0995_fracs shape: \", CGs_D6_unedited_mc0995_fracs.shape)   \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8",
   "metadata": {},
   "source": [
    "\n",
    "## CRISPRoff Day 6 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_padded_reads_name_np =  \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/croff/analyze_single_reads/dimelo_v2_output/CG_137_padded_reads_day6_CRoff_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD802_numRVS1480.npy\"\n",
    "CGs_D6_CRoff_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "\n",
    "CGs_D6_CRoff_mc07_fracs = get_fractions(CGs_D6_CRoff_mc07)\n",
    "print(\"CGs_D6_CRoff_mc07 shape: \", CGs_D6_CRoff_mc07.shape)\n",
    "print(\"CGs_D6_CRoff_mc07_fracs shape: \", CGs_D6_CRoff_mc07_fracs.shape)\n",
    "\n",
    "\n",
    "save_padded_reads_name_np = \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/croff/analyze_single_reads/dimelo_v2_output/\" +\\\n",
    "     'CG_137_padded_reads_day6_CRoff_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD793_numRVS1449.npy'\n",
    "# \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/croff/analyze_single_reads/dimelo_v2_output/CG_137_padded_reads_day6_CRoff_Tcells_Thr0.995_NoFullyUnmethylated_T2Tv2_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-09-13_units_combined_numFWD652_numRVS1125.npy\"\n",
    "CGs_D6_CRoff_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "\n",
    "CGs_D6_CRoff_mc0995_fracs = get_fractions(CGs_D6_CRoff_mc0995)\n",
    "print(\"CGs_D6_CRoff_mc0995 shape: \", CGs_D6_CRoff_mc0995.shape)\n",
    "print(\"CGs_D6_CRoff_mc0995_fracs shape: \", CGs_D6_CRoff_mc0995_fracs.shape)     "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10",
   "metadata": {},
   "source": [
    "# Day 28: \n",
    "Unedited Cells \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Day 28 post EP\n",
    "## Unedited Cells \n",
    "folder_path = \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_28/unedited/analyze_single_reads/dimelo_v2_output/\"\n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day28_unedited_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD6_numRVS17.npy\"\n",
    "CGs_D28_unedited_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "\n",
    "CGs_D28_unedited_mc07_fracs = get_fractions(CGs_D28_unedited_mc07)\n",
    "print(\"CGs_D28_unedited_mc07 shape: \", CGs_D28_unedited_mc07.shape)\n",
    "print(\"CGs_D28_unedited_mc07_fracs shape: \", CGs_D28_unedited_mc07_fracs.shape)  \n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day28_unedited_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD6_numRVS18.npy\"\n",
    "CGs_D28_unedited_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "\n",
    "CGs_D28_unedited_mc0995_fracs = get_fractions(CGs_D28_unedited_mc0995)\n",
    "print(\"CGs_D28_unedited_mc0995 shape: \", CGs_D28_unedited_mc0995.shape)\n",
    "print(\"CGs_D28_unedited_mc0995_fracs shape: \", CGs_D28_unedited_mc0995_fracs.shape) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12",
   "metadata": {},
   "source": [
    "CRoff day 28"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_path = \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_28/croff/analyze_single_reads/dimelo_v2_output/\"\n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day28_CRoff_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD6_numRVS6.npy\"\n",
    "CGs_D28_CRoff_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "\n",
    "CGs_D28_CRoff_mc07_fracs = get_fractions(CGs_D28_CRoff_mc07)\n",
    "print('CGs_D28_CRoff_mc07 shape: ', CGs_D28_CRoff_mc07.shape) \n",
    "print(\"CGs_D28_CRoff_mc07_fracs shape: \", CGs_D28_CRoff_mc07_fracs.shape)\n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day28_CRoff_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD6_numRVS6.npy\"\n",
    "CGs_D28_CRoff_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "CGs_D28_CRoff_mc0995_fracs = get_fractions(CGs_D28_CRoff_mc0995)\n",
    "print(\"CGs_D28_CRoff_mc0995 shape: \", CGs_D28_CRoff_mc0995.shape)\n",
    "print(\"CGs_D28_CRoff_mc0995_fracs shape: \", CGs_D28_CRoff_mc0995_fracs.shape)   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14",
   "metadata": {},
   "source": [
    "# Day 35: CROff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Day 35\n",
    "# folder_path = \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_35/croff/analyze_single_reads/dimelo_v2_output/\"\n",
    "folder_path =  \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_35/croff/replica_1/analyze_single_reads/dimelo_v2_output/\"\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_Tcells_CRISPRoff_Day35_postEP_R9minion_threshold_mC0.7_T2Tv2_0_filterMode10_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_mCthresh0.7_T2Tv2_0_chr1:206583354-206589854_2025-11-09_units_combined_numFWD104_numRVS222.npy\"\n",
    "# \"CG_137_padded_reads_day35_CRoff_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD104_numRVS222.npy\"\n",
    "CGs_D35_CRoff_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "CGs_D35_CRoff_mc07_fracs = get_fractions(CGs_D35_CRoff_mc07)\n",
    "print(\"CGs_D35_CRoff_mc07 shape\", CGs_D35_CRoff_mc07.shape)\n",
    "print(\"CGs_D35_CRoff_mc07_fracs shape: \", CGs_D35_CRoff_mc07_fracs.shape)   \n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_Tcells_CRISPRoff_Day35_postEP_R9minion_threshold_mC0.995_T2Tv2_0_filterMode10_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_mCthresh0.995_T2Tv2_0_chr1:206583354-206589854_2025-11-09_units_combined_numFWD104_numRVS222.npy\"\n",
    "CGs_D35_CRoff_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "CGs_D35_CRoff_mc0995_fracs = get_fractions(CGs_D35_CRoff_mc0995)\n",
    "print(\"CGs_D35_CRoff_mc0995 shape:\", CGs_D35_CRoff_mc0995.shape)\n",
    "print(\"CGs_D35_CRoff_mc0995_fracs shape: \", CGs_D35_CRoff_mc0995_fracs.shape)   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16",
   "metadata": {},
   "source": [
    "Day 35: unedited; 37hours sequenced"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_path= \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_35/unedited/part1_37h_sequenced/dimelo_v2_output/\"\n",
    "\n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day35_CRoff_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-10-05_units_combined_numFWD86_numRVS115.npy\"\n",
    "CGs_D35_unedited_mc07_lib1 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "CGs_D35_unedited_mc07_frac_lib1 = get_fractions(CGs_D35_unedited_mc07_lib1)\n",
    "print(\"CGs_D35_unedited_mc07_lib1 shape:\", CGs_D35_unedited_mc07_lib1.shape)\n",
    "print(\"CGs_D35_unedited_mc07_fracs_lib1 shape: \", CGs_D35_unedited_mc07_frac_lib1.shape)   \n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day35_CRoff_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-10-05_units_combined_numFWD86_numRVS114.npy\"\n",
    "CGs_D35_unedited_mc0995_lib1 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "CGs_D35_unedited_mc0995_fracs_lib1 = get_fractions(CGs_D35_unedited_mc0995_lib1)\n",
    "print(\"CGs_D35_unedited_mc0995_lib1 shape:\", CGs_D35_unedited_mc0995_lib1.shape)\n",
    "print(\"CGs_D35_unedited_mc0995_fracs_lib1 shape: \", CGs_D35_unedited_mc0995_fracs_lib1.shape)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18",
   "metadata": {},
   "outputs": [],
   "source": [
    "# folder_path= \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_35/unedited/reloaded_library/dimelo_v2_output/\"\n",
    "\n",
    "\n",
    "# save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day35_REloadedLibrary_MT_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-10-13_units_combined_numFWD0_numRVS1.npy\"\n",
    "# CGs_D35_unedited_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "# CGs_D35_unedited_mc07_fracs = get_fractions(CGs_D35_unedited_mc07)\n",
    "# print(\"CGs_D35_unedited_mc07 shape:\", CGs_D35_unedited_mc07.shape)\n",
    "# print(\"CGs_D35_unedited_mc07_fracs shape: \", CGs_D35_unedited_mc07_fracs.shape)   \n",
    "\n",
    "# save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_day35_REloadedLibrary_MT_Tcells_mC0.9_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.9_t2t_v2_0_chr1:206583354-206589854_2025-10-13_units_combined_numFWD0_numRVS1.npy\"\n",
    "# CGs_D35_unedited_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "# CGs_D35_unedited_mc0995_fracs = get_fractions(CGs_D35_unedited_mc0995)\n",
    "# print(\"CGs_D35_unedited_mc0995 shape:\", CGs_D35_unedited_mc0995.shape)\n",
    "# print(\"CGs_D35_unedited_mc0995_fracs shape: \", CGs_D35_unedited_mc0995_fracs.shape)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "folder_path= \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_35/unedited/merged_2libraries/dimelo_v2_output/\"\n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_Tcells_NT_Day35_postEP_R9minion_threshold_mC0.7_T2Tv2_0_filterMode10_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_mCthresh0.7_T2Tv2_0_chr1:206583354-206589854_2025-11-09_units_combined_numFWD87_numRVS115.npy\"\n",
    "CGs_D35_unedited_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "CGs_D35_unedited_mc07_fracs = get_fractions(CGs_D35_unedited_mc07)\n",
    "print(\"CGs_D35_unedited_mc07 shape:\", CGs_D35_unedited_mc07.shape)\n",
    "print(\"CGs_D35_unedited_mc07_fracs shape: \", CGs_D35_unedited_mc07_fracs.shape)   \n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CG_137_padded_reads_Tcells_NT_Day35_postEP_R9minion_threshold_mC0.995_T2Tv2_0_filterMode10_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_mCthresh0.995_T2Tv2_0_chr1:206583354-206589854_2025-11-09_units_combined_numFWD87_numRVS114.npy\"\n",
    "CGs_D35_unedited_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "CGs_D35_unedited_mc0995_fracs = get_fractions(CGs_D35_unedited_mc0995)\n",
    "print(\"CGs_D35_unedited_mc0995 shape:\", CGs_D35_unedited_mc0995.shape)\n",
    "print(\"CGs_D35_unedited_mc0995_fracs shape: \", CGs_D35_unedited_mc0995_fracs.shape)   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20",
   "metadata": {},
   "source": [
    "# Filtered out reads"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21",
   "metadata": {},
   "source": [
    "D6 filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_path = \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_6/model_data/\"\n",
    "\n",
    "save_padded_reads_name_np = folder_path + 'CGs_CRoff_D6_filtered_884reads_0.3mCfrac_mC07_2025-10-22.npy'\n",
    "# \"CG_137_padded_reads_day35_2Library_NT_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-10-13_units_combined_numFWD87_numRVS114.npy\"\n",
    "filtered_CGs_D6_CRoff_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "filtered_CGs_D6_CRoff_mc07_fracs = get_fractions(filtered_CGs_D6_CRoff_mc07)\n",
    "print(\"filtered_CGs_D6_CRoff_mc07 shape:\", filtered_CGs_D6_CRoff_mc07.shape)\n",
    "print(\"filtered_CGs_D6_CRoff_mc07_fracs shape: \", filtered_CGs_D6_CRoff_mc07_fracs.shape)   \n",
    "\n",
    "\n",
    "save_padded_reads_name_np = folder_path +'CGs_NT_D6_filtered_884reads_0.16mCfrac_mC07_2025-10-22.npy'\n",
    "# \"CG_137_padded_reads_day6_unedited_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD489_numRVS638.npy\"\n",
    "filtered_CGs_D6_unedited_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "filtered_CGs_D6_unedited_mc07_fracs = get_fractions(filtered_CGs_D6_unedited_mc07)\n",
    "print(\"CGs_D6_unedited_mc07 shape: \", filtered_CGs_D6_unedited_mc07.shape)\n",
    "print(\"CGs_D6_unedited_mc07_fracs shape: \", filtered_CGs_D6_unedited_mc07_fracs.shape)   \n",
    "\n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CGs_CRoff_D6_filtered_539reads_0.26mCfrac_mC0995_2025-10-22.npy\"\n",
    "# \"CG_137_padded_reads_day35_2Library_NT_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-10-13_units_combined_numFWD87_numRVS114.npy\"\n",
    "filtered_CGs_D6_CRoff_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "filtered_CGs_D6_CRoff_mc0995_fracs = get_fractions(filtered_CGs_D6_CRoff_mc0995)\n",
    "print(\"filtered_CGs_D6_CRoff_mc0995 shape:\", filtered_CGs_D6_CRoff_mc0995.shape)\n",
    "print(\"filtered_CGs_D6_CRoff_mc0995_fracs shape: \", filtered_CGs_D6_CRoff_mc0995_fracs.shape)   \n",
    "\n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CGs_NT_D6_filtered_539reads_0.1mCfrac_mC0995_2025-10-22.npy\"\n",
    "# \"CG_137_padded_reads_day6_unedited_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD489_numRVS638.npy\"\n",
    "filtered_CGs_D6_unedited_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "filtered_CGs_D6_unedited_mc0995_fracs = get_fractions(filtered_CGs_D6_unedited_mc0995)\n",
    "print(\"CGs_D6_unedited_mc0995 shape: \", filtered_CGs_D6_unedited_mc0995.shape)\n",
    "print(\"CGs_D6_unedited_mc0995_fracs shape: \", filtered_CGs_D6_unedited_mc0995_fracs.shape)   \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23",
   "metadata": {},
   "source": [
    "D35 filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "folder_path= \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/T_cells/day_35/model_data/\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "save_padded_reads_name_np = folder_path + 'CGs_NT_D35_filtered_167reads_0.16mCfrac_mC07_2025-10-22.npy'\n",
    "# \"CG_137_padded_reads_day35_2Library_NT_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-10-13_units_combined_numFWD87_numRVS115.npy\"\n",
    "filtered_CGs_D35_unedited_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "filtered_CGs_D35_unedited_mc07_fracs = get_fractions(filtered_CGs_D35_unedited_mc07)\n",
    "print(\"filtered_CGs_D35_unedited_mc07 shape:\", filtered_CGs_D35_unedited_mc07.shape)\n",
    "print(\"filtered_CGs_D35_unedited_mc07_fracs shape: \", filtered_CGs_D35_unedited_mc07_fracs.shape)   \n",
    "\n",
    "save_padded_reads_name_np = folder_path + 'CGs_CRoff_D35_filtered_167reads_0.3mCfrac_mC07_2025-10-22.npy'\n",
    "# \"CG_137_padded_reads_day35_CRoff_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD104_numRVS222.npy\"\n",
    "filtered_CGs_D35_CRoff_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "filtered_CGs_D35_CRoff_mc07_fracs = get_fractions(filtered_CGs_D35_CRoff_mc07)\n",
    "print(\"CGs_D35_CRoff_mc07 shape\", filtered_CGs_D35_CRoff_mc07.shape)\n",
    "print(\"CGs_D35_CRoff_mc07_fracs shape: \", filtered_CGs_D35_CRoff_mc07_fracs.shape)   \n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CGs_CRoff_D35_filtered_141reads_0.26mCfrac_mC0995_2025-10-22.npy\"\n",
    "# \"CG_137_padded_reads_day35_2Library_NT_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-10-13_units_combined_numFWD87_numRVS114.npy\"\n",
    "filtered_CGs_D35_CRoff_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "filtered_CGs_D35_CRoff_mc0995_fracs = get_fractions(filtered_CGs_D35_CRoff_mc0995)\n",
    "print(\"filtered_CGs_D35_CRoff_mc0995 shape:\", filtered_CGs_D35_CRoff_mc0995.shape)\n",
    "print(\"filtered_CGs_D35_CRoff_mc0995_fracs shape: \", filtered_CGs_D35_CRoff_mc0995_fracs.shape)   \n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"CGs_NT_D35_filtered_141reads_0.1mCfrac_mC0995_2025-10-22.npy\"\n",
    "filtered_CGs_D35_unedited_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "filtered_CGs_D35_unedited_mc0995_fracs = get_fractions(filtered_CGs_D35_unedited_mc0995)\n",
    "print(\"filtered_CGs_D35_unedited_mc0995 shape:\", filtered_CGs_D35_unedited_mc0995.shape)\n",
    "print(\"filtered_CGs_D35_unedited_mc0995_fracs shape: \", filtered_CGs_D35_unedited_mc0995_fracs.shape)   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25",
   "metadata": {},
   "source": [
    "# K562"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "folder_path= \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/T2T_v2.0_mapped/K562/model_data/\"\n",
    "\n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"mC_thresh_07/CGs_NT_filtered_197reads_0.1mCfrac_mC07_2025-10-28.npy\"\n",
    "\n",
    "# 'CGs_NT_K562_filtered_167reads_0.16mCfrac_mC07_2025-10-22.npy'\n",
    "# \"CG_137_padded_reads_day35_2Library_NT_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-10-13_units_combined_numFWD87_numRVS115.npy\"\n",
    "filtered_CGs_K562_unedited_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "filtered_CGs_K562_unedited_mc07_fracs = get_fractions(filtered_CGs_K562_unedited_mc07)\n",
    "print(\"filtered_CGs_K562_unedited_mc07 shape:\", filtered_CGs_K562_unedited_mc07.shape)\n",
    "print(\"filtered_CGs_K562_unedited_mc07_fracs shape: \", filtered_CGs_K562_unedited_mc07_fracs.shape)   \n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"mC_thresh_07/CGs_Zoff_High_filtered_197reads_0.3mCfrac_mC07_2025-10-28.npy\"\n",
    "# 'CGs_ZoffHigh_K562_filtered_167reads_0.3mCfrac_mC07_2025-10-22.npy'\n",
    "# \"CG_137_padded_reads_day35_ZoffHigh_Tcells_mC0.7_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.7_t2t_v2_0_chr1:206583354-206589854_2025-09-29_units_combined_numFWD104_numRVS222.npy\"\n",
    "filtered_CGs_K562_ZoffHigh_mc07 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "filtered_CGs_K562_ZoffHigh_mc07_fracs = get_fractions(filtered_CGs_K562_ZoffHigh_mc07)\n",
    "print(\"CGs_K562_ZoffHigh_mc07 shape\", filtered_CGs_K562_ZoffHigh_mc07.shape)\n",
    "print(\"CGs_K562_ZoffHigh_mc07_fracs shape: \", filtered_CGs_K562_ZoffHigh_mc07_fracs.shape)   \n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"mC_thresh_0995/CGs_Zoff_High_filtered_192reads_0.3mCfrac_mC0995_2025-10-28.npy\"\n",
    "# \"CG_137_padded_reads_day35_2Library_NT_Tcells_mC0.995_T2Tv2_NoFullyUnmeth_ovrlap0.9_mismat0.7_mapQ60_modeBaseQ10_mCthresh0.995_t2t_v2_0_chr1:206583354-206589854_2025-10-13_units_combined_numFWD87_numRVS114.npy\"\n",
    "filtered_CGs_K562_ZoffHigh_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "filtered_CGs_K562_ZoffHigh_mc0995_fracs = get_fractions(filtered_CGs_K562_ZoffHigh_mc0995)\n",
    "print(\"filtered_CGs_K562_ZoffHigh_mc0995 shape:\", filtered_CGs_K562_ZoffHigh_mc0995.shape)\n",
    "print(\"filtered_CGs_K562_ZoffHigh_mc0995_fracs shape: \", filtered_CGs_K562_ZoffHigh_mc0995_fracs.shape)   \n",
    "\n",
    "save_padded_reads_name_np = folder_path + \"mC_thresh_0995/CGs_NT_filtered_192reads_0.04mCfrac_mC0995_2025-10-28.npy\"\n",
    "filtered_CGs_K562_unedited_mc0995 = load_padded_reads(save_folder_path, save_padded_reads_name_np)\n",
    "filtered_CGs_K562_unedited_mc0995_fracs = get_fractions(filtered_CGs_K562_unedited_mc0995)\n",
    "print(\"filtered_CGs_K562_unedited_mc0995 shape:\", filtered_CGs_K562_unedited_mc0995.shape)\n",
    "print(\"filtered_CGs_K562_unedited_mc0995_fracs shape: \", filtered_CGs_K562_unedited_mc0995_fracs.shape)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "28",
   "metadata": {},
   "source": [
    "# 99.5% mC confs - FILTERED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "plt.figure(figsize=(16, 5))\n",
    "plt.bar(np.arange(len(filtered_CGs_D6_CRoff_mc0995_fracs)), filtered_CGs_D6_CRoff_mc0995_fracs, snap=False, alpha=0.5, label='mC > 0.995 CRISPRoff', color='deepskyblue')\n",
    "plt.bar(np.arange(len(filtered_CGs_D6_unedited_mc0995_fracs)), filtered_CGs_D6_unedited_mc0995_fracs, snap=False, alpha=0.5, label='mC > 0.995 Unedited', color='navy')\n",
    "# mediumblue.'turquoise', 'aquamarine', 'deepskyblue', 'dodgerblue', 'cornflowerblue', 'royalblue', 'blue', 'mediumblue', 'darkblue', 'navy', 'midnightblue'\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('CpG units')\n",
    "plt.ylabel('Fraction of mC')\n",
    "plt.title('Filtered Fraction of mC for CRISPROff and Unedited T cells CGs Day 6 \\n mC threshold 99.5%')\n",
    "plt.legend()\n",
    "# Show the plot\n",
    "plt.show()\n",
    "\n",
    "\n",
    "# DIFFERENCES\n",
    "\n",
    "dif_filtered_D6_mCthresh_995 =   filtered_CGs_D6_CRoff_mc0995_fracs  - filtered_CGs_D6_unedited_mc0995_fracs\n",
    "\n",
    "plt.figure(figsize=(16, 5))\n",
    "plt.bar(np.arange(len(dif_filtered_D6_mCthresh_995)), dif_filtered_D6_mCthresh_995, snap=False, alpha=0.5, label='CRoff - Unedited, D6 with 99.5% mC', color='turquoise'  )\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('CpG units')\n",
    "plt.yticks(list(plt.yticks()[0]) + [-0.1])\n",
    "\n",
    "plt.ylabel('Fraction of mC difference')\n",
    "plt.title('Filtered Fraction Differences between mC Unedited T cells Day 6 \\n confidence threshold 99.5%')\n",
    "plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_fracs(fracs_dict, title=\"Methylation Fractions\",\n",
    "               figsize=(16, 5), alpha=0.6, colors=None):\n",
    "    \"\"\"Bar plot comparing methylation fractions across conditions.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    fracs_dict : dict[str -> np.ndarray]\n",
    "        {condition_label: fracs_array}  — plotted in insertion order.\n",
    "    title : str\n",
    "    figsize : tuple\n",
    "    alpha : float\n",
    "    colors : list of str, optional\n",
    "    \"\"\"\n",
    "    default_colors = [\"deepskyblue\", \"navy\", \"tomato\", \"darkred\",\n",
    "                      \"mediumseagreen\", \"darkgreen\"]\n",
    "    colors = colors or default_colors\n",
    "    n_cpgs = next(iter(fracs_dict.values())).shape[0]\n",
    "    x = np.arange(n_cpgs)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "    for i, (label, fracs) in enumerate(fracs_dict.items()):\n",
    "        ax.bar(x, fracs, snap=False, alpha=alpha,\n",
    "               label=label, color=colors[i % len(colors)])\n",
    "    ax.set_xlabel(\"CpG Index\")\n",
    "    ax.set_ylabel(\"Methylation Fraction\")\n",
    "    ax.set_title(title)\n",
    "\n",
    "\n",
    "    # Set x-axis ticks to show all CpG indices\n",
    "    ax.set_xticks(np.arange(n_cpgs))\n",
    "    ax.set_xticklabels(np.arange(n_cpgs), rotation=90, fontsize=8)\n",
    "\n",
    "    ax.legend()\n",
    "    plt.tight_layout()\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "# Day 6: CRoff vs Unedited, mc0.7\n",
    "plot_fracs(\n",
    "    {\"CRISPRoff (mc>0.995)\": filtered_CGs_D6_CRoff_mc0995_fracs,\n",
    "     \"Unedited  (mc>0.995)\": filtered_CGs_D6_unedited_mc0995_fracs},\n",
    "    title=\"Day 6 — CD55 CpG Methylation Fractions (Filtered, mC threshold 0.995)\",\n",
    "    colors=[\"skyblue\", \"royalblue\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31",
   "metadata": {},
   "outputs": [],
   "source": [
    "dif_filtered_D6_mCthresh_995"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32",
   "metadata": {},
   "source": [
    "# Reference genome plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# ── Genomic constants ──────────────────────────────────────────────────────────\n",
    "REGION_CHR   = \"chr1\"\n",
    "REGION_START = 206_583_354   # T2T v2.0, 1-based inclusive\n",
    "REGION_END   = 206_589_854   # T2T v2.0, 1-based inclusive\n",
    "OFFSET_HG38  = 734_704       # hg38_pos = T2T_pos + OFFSET_HG38\n",
    "TSS_T2T      = 206_586_974   # CD55 TSS (T2T v2.0)\n",
    "TSS_HG38     = 207_321_678   # CD55 TSS (hg38) — verified correct value\n",
    "PROMOTER_WIN = 1_000         # ±1,000 bp around TSS defines promoter\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from pathlib import Path\n",
    "from datetime import datetime\n",
    "\n",
    "# Add the folder containing 'my_functions.py' to sys.path\n",
    "sys.path.append(\"/home/michalula/code/epiCausality/epiCode/utils/\") # str(Path(__file__).parent / 'utils'))\n",
    "\n",
    "# Import the module or specific functions\n",
    "# from /home/michalula/code/epiCausality/epiCode/utils/funcs_extract_mC_profiles_from_BAMs.py\n",
    "# from funcs_extract_mC_profiles_from_BAMs import system_info, extract_from_bam\n",
    "from funcs_extract_mC_profiles_from_BAMs import (\n",
    "    system_info,\n",
    "    get_reference_sequence,\n",
    "    create_output_directory,\n",
    "    extract_from_bam,\n",
    "    process_extracted_reads,\n",
    "    visualize_data,\n",
    "    create_padded_reads,\n",
    "    plot_padded_reads,\n",
    "    save_padded_reads,\n",
    "    process_extracted_reads_no_fully_unmethylated,\n",
    "    create_padded_reads_no_fully_unmethylated\n",
    "    # main,\n",
    ")\n",
    "\n",
    "\n",
    "def find_cpgs(dna: str) -> list[int]:\n",
    "    \"\"\"Return the start positions of all CpG dinucleotides in a DNA string.\"\"\"\n",
    "    dna = dna.upper()\n",
    "    return [i for i in range(len(dna) - 1) if dna[i] == \"C\" and dna[i + 1] == \"G\"]\n",
    "\n",
    "\n",
    "def count_surrounding_cpgs(dna: str, n: int) -> dict[int, int]:\n",
    "    \"\"\"\n",
    "    For each CpG in dna, count how many *other* CpGs have their start position\n",
    "    within N bases (i.e. |pos_other - pos_self| <= N, excluding self).\n",
    "\n",
    "    Returns a dict mapping each CpG start position -> surrounding CpG count.\n",
    "    \"\"\"\n",
    "    cpg_positions = find_cpgs(dna)\n",
    "    result = {}\n",
    "    for pos in cpg_positions:\n",
    "        count = sum(\n",
    "            1\n",
    "            for other in cpg_positions\n",
    "            if other != pos and abs(other - pos) <= n\n",
    "        )\n",
    "        result[pos] = count\n",
    "    return result\n",
    "\n",
    "\n",
    "\n",
    "system_info()\n",
    "date_today = datetime.today().strftime('%Y-%m-%d')\n",
    "ref_genome_path = Path('/home/michalula/data/ref_genomes/t2t_v2_0/up_chm13v2.0.fasta')\n",
    "reg_genome_version = \"t2t_v2_0\"\n",
    "region_chr = 'chr1'\n",
    "\n",
    "# 6500bps: EXACT cutting region  \n",
    "# chr1:206,583,334-206,589,873\n",
    "region_start = 206583334    + 20\n",
    "region_end = 206589874      - 20 \n",
    "# CD55 TSS start: chr1:206586828-206606065 (+)\n",
    "# (before TSS: 3474, after TSS: 3026 bps)\n",
    "# region_start 206583354\n",
    "# region_end 206589854\n",
    "# chr1:206583354-206589854\n",
    "\n",
    "region_str = region_chr + \":\" + str(region_start) + \"-\" + str(region_end) #'chr1:206586162-206586192'\n",
    "region_length = region_end - region_start\n",
    "print(\"region_length\", region_length)\n",
    "\n",
    "\n",
    "motifs=['CG,0']\n",
    "ref_seq_list = get_reference_sequence(ref_genome_path, region_chr, region_start, region_end)\n",
    "\n",
    "print('region_chr', region_chr)\n",
    "print('region_start', region_start)\n",
    "print('region_end', region_end)\n",
    "print(f'{region_chr}:{region_start}-{region_end}') \n",
    "\n",
    "\n",
    "\n",
    "dna_6500_roi_str = \"\".join(ref_seq_list)         # \"abc\"\n",
    "\n",
    "dna = dna_6500_roi_str\n",
    "N = 50\n",
    "SEQ_LEN = len(dna)\n",
    "\n",
    "surrounding = count_surrounding_cpgs(dna, n=N)\n",
    "print(f\"Surrounding CpG counts within +/-{N} bases:\")\n",
    "print(f\"{'CpG pos':>10}  {'context':>20}  {'surrounding CpGs':>16}\")\n",
    "print(\"-\" * 52)\n",
    "for pos, count in surrounding.items():\n",
    "    # Show a small context window around the CpG\n",
    "    start = max(0, pos - 3)\n",
    "    end = min(SEQ_LEN, pos + 5)\n",
    "    context = dna[start:end]\n",
    "    print(f\"{pos:>10}  {context:>20}  {count:>16}\")\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Extract positions and counts from the surrounding dictionary\n",
    "positions =  list(range(0, len(surrounding) )) # CG_indexes #list(surrounding.keys())\n",
    "counts = list(surrounding.values())\n",
    "\n",
    "# Create a bar plot\n",
    "plt.figure(figsize=(14, 6))\n",
    "plt.bar(positions, counts, width=1.0, edgecolor='none')\n",
    "plt.xlabel('CpG Position')\n",
    "plt.ylabel('Surrounding CpG Count')\n",
    "plt.title(f'Surrounding CpG Counts within +/-{N} bases')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35",
   "metadata": {},
   "source": [
    "# SWITCH to HG38 coords:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36",
   "metadata": {},
   "outputs": [],
   "source": [
    "region_start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37",
   "metadata": {},
   "outputs": [],
   "source": [
    "207318058 - region_start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38",
   "metadata": {},
   "outputs": [],
   "source": [
    "OFFSET_HG38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39",
   "metadata": {},
   "outputs": [],
   "source": [
    "OFFSET_HG38 - region_end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40",
   "metadata": {},
   "outputs": [],
   "source": [
    "region_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41",
   "metadata": {},
   "outputs": [],
   "source": [
    "207324558 - region_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr1:207318058-207324558\n",
    "region_start=207318058\n",
    "region_end=207324558\n",
    "\n",
    "seq_str = ''.join(ref_seq_list)\n",
    "CG_pair_idx = [i for i in range(len(seq_str) - 1) if seq_str[i] == 'C' and seq_str[i + 1] == 'G']\n",
    "# print(\"CG Pair Indices:\", CG_pair_idx)\n",
    "# Calculate the genomic coordinates of the CGs\n",
    "CG_coordinates = [(region_start + idx) for idx in CG_pair_idx]\n",
    "\n",
    "CG_coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.ticker as ticker\n",
    "# Create a comprehensive plot of CpG methylation mapped to genomic coordinates\n",
    "fig, ax = plt.subplots(figsize=(22, 6))\n",
    "\n",
    "# Use genomic coordinates for x-axis instead of CpG indices\n",
    "ax.bar(CG_coordinates, dif_filtered_D6_mCthresh_995, snap=False, alpha=0.7, \n",
    "    color=\"turquoise\", edgecolor='none', width=10)\n",
    "\n",
    "\n",
    "# Add CpG index labels on top of each bar\n",
    "for idx, (coord, value) in enumerate(zip(CG_coordinates, dif_filtered_D6_mCthresh_995)):\n",
    "    ax.text(coord, value + 0.01 if value > 0 else value - 0.01, \n",
    "            str(idx), ha='center', va='bottom' if value > 0 else 'top',\n",
    "            fontsize=5, rotation=0)\n",
    "\n",
    "\n",
    "ax.axhline(0, color=\"black\", linewidth=0.8, linestyle=\"--\")\n",
    "ax.set_xlabel(\"Genomic Coordinate (HG38)\")\n",
    "ax.set_ylabel(\"ΔmC (CRISPRoff − Unedited)\")\n",
    "ax.set_title(f\"Day 6 Differential Methylation across {REGION_CHR}:{region_start}-{region_end}\")\n",
    "\n",
    "# Format x-axis with readable genomic coordinates\n",
    "ax.ticklabel_format(style='plain', axis='x')\n",
    "ax.xaxis.set_major_locator(ticker.MaxNLocator(nbins=10))\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.ticker as ticker\n",
    "\n",
    "\n",
    "# Zoom plot: CpG units 16..129 (inclusive)\n",
    "start_idx, end_idx = 0, 136\n",
    "sel_coords = CG_coordinates[start_idx:end_idx+1]\n",
    "sel_vals = dif_filtered_D6_mCthresh_995[start_idx:end_idx+1]\n",
    "\n",
    "# compute a reasonable bar width from local coordinate spacing\n",
    "if len(sel_coords) > 1:\n",
    "    median_spacing = int(np.median(np.diff(sel_coords)))\n",
    "    bar_width = max(10, int(median_spacing * 0.6))\n",
    "else:\n",
    "    bar_width = 50\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16, 4))\n",
    "ax.bar(sel_coords, sel_vals, width=bar_width, color=\"turquoise\", alpha=0.85, edgecolor=\"none\")\n",
    "ax.axhline(0, color=\"black\", linewidth=0.8, linestyle=\"--\")\n",
    "\n",
    "ax.set_xlabel(\"Genomic Coordinate (HG38)\")\n",
    "ax.set_ylabel(\"ΔmC (CRISPRoff − Unedited)\")\n",
    "ax.set_title(f\"Filtered Zoom: CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]})\")\n",
    "\n",
    "# Add CpG index labels on top of each bar\n",
    "for idx, (coord, value) in enumerate(zip(sel_coords, sel_vals), start=start_idx):\n",
    "    ax.text(coord, value + 0.01 if value > 0 else value - 0.01, \n",
    "            str(idx), ha='center', va='bottom' if value > 0 else 'top',\n",
    "            fontsize=5, rotation=0) \n",
    "\n",
    "# format x-axis and show fewer ticks to avoid crowding\n",
    "ax.ticklabel_format(style='plain', axis='x')\n",
    "ax.xaxis.set_major_locator(ticker.MaxNLocator(nbins=8))\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.ticker as ticker\n",
    "\n",
    "\n",
    "# Zoom plot: CpG units 23..103 (inclusive)\n",
    "start_idx, end_idx = 23, 103\n",
    "sel_coords = CG_coordinates[start_idx:end_idx+1]\n",
    "sel_vals = dif_filtered_D6_mCthresh_995[start_idx:end_idx+1]\n",
    "\n",
    "# compute a reasonable bar width from local coordinate spacing\n",
    "if len(sel_coords) > 1:\n",
    "    median_spacing = int(np.median(np.diff(sel_coords)))\n",
    "    bar_width = max(10, int(median_spacing * 0.6))\n",
    "else:\n",
    "    bar_width = 50\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16, 4))\n",
    "ax.bar(sel_coords, sel_vals, width=bar_width, color=\"turquoise\", alpha=0.85, edgecolor=\"none\")\n",
    "ax.axhline(0, color=\"black\", linewidth=0.8, linestyle=\"--\")\n",
    "\n",
    "ax.set_xlabel(\"Genomic Coordinate (HG38)\")\n",
    "ax.set_ylabel(\"ΔmC (CRISPRoff − Unedited)\")\n",
    "ax.set_title(f\"Zoom Differenced D6 CROff - Unedired Filtered : CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]})\")\n",
    "\n",
    "# Add CpG index labels on top of each bar\n",
    "for idx, (coord, value) in enumerate(zip(sel_coords, sel_vals), start=start_idx):\n",
    "    ax.text(coord, value + 0.01 if value > 0 else value - 0.01, \n",
    "            str(idx), ha='center', va='bottom' if value > 0 else 'top',\n",
    "            fontsize=6, rotation=0) \n",
    "\n",
    "# format x-axis and show fewer ticks to avoid crowding\n",
    "ax.ticklabel_format(style='plain', axis='x')\n",
    "ax.xaxis.set_major_locator(ticker.MaxNLocator(nbins=8))\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.ticker as ticker\n",
    "\n",
    "\n",
    "# Zoom plot: CpG units 23..103 (inclusive)\n",
    "start_idx, end_idx = 0, 136\n",
    "sel_coords = CG_coordinates[start_idx:end_idx+1]\n",
    "sel_vals = dif_filtered_D6_mCthresh_995[start_idx:end_idx+1]\n",
    "\n",
    "# compute a reasonable bar width from local coordinate spacing\n",
    "if len(sel_coords) > 1:\n",
    "    median_spacing = int(np.median(np.diff(sel_coords)))\n",
    "    bar_width = max(10, int(median_spacing * 0.8))\n",
    "else:\n",
    "    bar_width = 50\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16, 4))\n",
    "ax.bar(sel_coords, sel_vals, width=bar_width, color=\"turquoise\", alpha=0.85, edgecolor=\"none\")\n",
    "ax.axhline(0, color=\"black\", linewidth=0.8, linestyle=\"--\")\n",
    "\n",
    "ax.set_xlabel(\"Genomic Coordinate (HG38)\")\n",
    "ax.set_ylabel(\"ΔmC (CRISPRoff − Unedited)\")\n",
    "ax.set_title(f\"Zoom Differenced D6 CROff - Unedired Filtered : CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]})\")\n",
    "\n",
    "# Add CpG index labels on top of each bar\n",
    "for idx, (coord, value) in enumerate(zip(sel_coords, sel_vals), start=start_idx):\n",
    "    ax.text(coord, value + 0.01 if value > 0 else value - 0.01, \n",
    "            str(idx), ha='center', va='bottom' if value > 0 else 'top',\n",
    "            fontsize=6, rotation=0) \n",
    "\n",
    "# format x-axis and show fewer ticks to avoid crowding\n",
    "ax.ticklabel_format(style='plain', axis='x')\n",
    "ax.xaxis.set_major_locator(ticker.MaxNLocator(nbins=8))\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.ticker as ticker\n",
    "\n",
    "# Zoom plot: CpG units 22..103 (inclusive)\n",
    "start_idx, end_idx = 20, 103\n",
    "sel_coords = CG_coordinates[start_idx:end_idx+1]\n",
    "sel_vals = dif_filtered_D6_mCthresh_995[start_idx:end_idx+1]\n",
    "\n",
    "# compute a reasonable bar width from local coordinate spacing\n",
    "if len(sel_coords) > 1:\n",
    "    median_spacing = int(np.median(np.diff(sel_coords)))\n",
    "    bar_width = max(10, int(median_spacing * 0.6))\n",
    "else:\n",
    "    bar_width = 50\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16, 4))\n",
    "ax.bar(sel_coords, sel_vals, width=bar_width, color=\"turquoise\", alpha=0.85, edgecolor=\"none\")\n",
    "ax.axhline(0, color=\"black\", linewidth=0.8, linestyle=\"--\")\n",
    "\n",
    "ax.set_xlabel(\"Genomic Coordinate (HG38)\")\n",
    "ax.set_ylabel(\"ΔmC (CRISPRoff − Unedited)\")\n",
    "ax.set_title(f\"Zoom Differenced D6 CROff - Unedired Filtered : CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]})\")\n",
    "\n",
    "# Add CpG index labels on top of each bar\n",
    "for idx, (coord, value) in enumerate(zip(sel_coords, sel_vals), start=start_idx):\n",
    "    ax.text(coord, value + 0.01 if value > 0 else value - 0.01, \n",
    "            str(idx), ha='center', va='bottom' if value > 0 else 'top',\n",
    "            fontsize=6, rotation=0) \n",
    "\n",
    "# Set x-axis ticks every 100 units\n",
    "ax.xaxis.set_major_locator(ticker.MultipleLocator(100))\n",
    "ax.ticklabel_format(style='plain', axis='x')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.ticker as ticker\n",
    "\n",
    "# Zoom plot: CpG units 20..103 (inclusive)\n",
    "start_idx, end_idx = 20, 103\n",
    "sel_coords = CG_coordinates[start_idx:end_idx+1]\n",
    "sel_vals = dif_filtered_D6_mCthresh_995[start_idx:end_idx+1]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16, 4))\n",
    "\n",
    "# Create bar positions centered at coordinates\n",
    "bar_width = 3\n",
    "ax.bar(sel_coords, sel_vals, width=bar_width, color=\"turquoise\", alpha=0.85, edgecolor=\"none\", align='center')\n",
    "ax.axhline(0, color=\"black\", linewidth=0.8, linestyle=\"--\")\n",
    "\n",
    "ax.set_xlabel(\"Genomic Coordinate (HG38)\")\n",
    "ax.set_ylabel(\"ΔmC (CRISPRoff − Unedited)\")\n",
    "ax.set_title(f\"Zoom Differenced D6 CROff - Unedired Filtered : CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]})\")\n",
    "\n",
    "# Add CpG index labels on top of each bar\n",
    "for idx, (coord, value) in enumerate(zip(sel_coords, sel_vals), start=start_idx):\n",
    "    ax.text(coord, value + 0.01 if value > 0 else value - 0.01, \n",
    "            str(idx), ha='center', va='bottom' if value > 0 else 'top',\n",
    "            fontsize=6, rotation=0) \n",
    "\n",
    "# Set x-axis ticks every 100 units\n",
    "ax.xaxis.set_major_locator(ticker.MultipleLocator(100))\n",
    "ax.ticklabel_format(style='plain', axis='x')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.ticker as ticker\n",
    "\n",
    "# Zoom plot: CpG units 71..79 (inclusive)\n",
    "start_idx, end_idx = 71, 79\n",
    "sel_coords = CG_coordinates[start_idx:end_idx+1]\n",
    "sel_vals = dif_filtered_D6_mCthresh_995[start_idx:end_idx+1]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16, 4))\n",
    "\n",
    "# Create bar positions centered at coordinates\n",
    "bar_width = 3\n",
    "ax.bar(sel_coords, sel_vals, width=bar_width, color=\"turquoise\", alpha=0.85, edgecolor=\"none\", align='center')\n",
    "ax.axhline(0, color=\"black\", linewidth=0.8, linestyle=\"--\")\n",
    "\n",
    "ax.set_xlabel(\"Genomic Coordinate (HG38)\")\n",
    "ax.set_ylabel(\"ΔmC (CRISPRoff − Unedited)\")\n",
    "ax.set_title(f\"Zoom Differenced D6 CROff - Unedired Filtered : CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]})\")\n",
    "\n",
    "# Add CpG index labels on top of each bar\n",
    "for idx, (coord, value) in enumerate(zip(sel_coords, sel_vals), start=start_idx):\n",
    "    ax.text(coord, value + 0.01 if value > 0 else value - 0.01, \n",
    "            str(idx), ha='center', va='bottom' if value > 0 else 'top',\n",
    "            fontsize=6, rotation=0) \n",
    "\n",
    "# Set x-axis ticks every 100 units\n",
    "ax.xaxis.set_major_locator(ticker.MultipleLocator(100))\n",
    "ax.ticklabel_format(style='plain', axis='x')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.ticker as ticker\n",
    "\n",
    "# Zoom plot: CpG units 69..79 (inclusive)\n",
    "start_idx, end_idx = 69, 79\n",
    "sel_coords = CG_coordinates[start_idx:end_idx+1]\n",
    "sel_vals = dif_filtered_D6_mCthresh_995[start_idx:end_idx+1]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16, 4))\n",
    "\n",
    "# Create bar positions centered at coordinates\n",
    "bar_width = 3\n",
    "ax.bar(sel_coords, sel_vals, width=bar_width, color=\"turquoise\", alpha=0.85, edgecolor=\"none\", align='center')\n",
    "ax.axhline(0, color=\"black\", linewidth=0.8, linestyle=\"--\")\n",
    "\n",
    "ax.set_xlabel(\"Genomic Coordinate (HG38)\")\n",
    "ax.set_ylabel(\"ΔmC (CRISPRoff − Unedited)\")\n",
    "ax.set_title(f\"Zoom Differenced D6 CROff - Unedired Filtered : CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]})\")\n",
    "\n",
    "# Add CpG index labels on top of each bar\n",
    "for idx, (coord, value) in enumerate(zip(sel_coords, sel_vals), start=start_idx):\n",
    "    ax.text(coord, value + 0.01 if value > 0 else value - 0.01, \n",
    "            str(idx), ha='center', va='bottom' if value > 0 else 'top',\n",
    "            fontsize=6, rotation=0) \n",
    "\n",
    "# Set x-axis ticks every 100 units\n",
    "ax.xaxis.set_major_locator(ticker.MultipleLocator(1))\n",
    "ax.ticklabel_format(style='plain', axis='x')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.ticker as ticker\n",
    "\n",
    "# Zoom plot: CpG units 69..79 (inclusive)\n",
    "start_idx, end_idx = 69, 79\n",
    "sel_coords = CG_coordinates[start_idx:end_idx+1]\n",
    "sel_vals = dif_filtered_D6_mCthresh_995[start_idx:end_idx+1]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16, 4))\n",
    "\n",
    "# Create bar positions centered at coordinates\n",
    "bar_width = 1\n",
    "ax.bar(sel_coords, sel_vals, width=bar_width, color=\"turquoise\", alpha=0.85, edgecolor=\"none\", align='center')\n",
    "ax.axhline(0, color=\"black\", linewidth=0.8, linestyle=\"--\")\n",
    "\n",
    "ax.set_xlabel(\"Genomic Coordinate (HG38)\")\n",
    "ax.set_ylabel(\"ΔmC (CRISPRoff − Unedited)\")\n",
    "ax.set_title(f\"Zoom Differenced D6 CROff - Unedired Filtered : CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]})\")\n",
    "\n",
    "# Add CpG index labels on top of each bar\n",
    "for idx, (coord, value) in enumerate(zip(sel_coords, sel_vals), start=start_idx):\n",
    "    ax.text(coord, value + 0.01 if value > 0 else value - 0.01, \n",
    "            str(idx), ha='center', va='bottom' if value > 0 else 'top',\n",
    "            fontsize=6, rotation=0) \n",
    "\n",
    "# Set x-axis ticks every 100 units\n",
    "ax.xaxis.set_major_locator(ticker.MultipleLocator(1))\n",
    "ax.ticklabel_format(style='plain', axis='x')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.ticker as ticker\n",
    "\n",
    "# Zoom plot: CpG units 69..82 (inclusive)\n",
    "start_idx, end_idx = 68, 82\n",
    "sel_coords = CG_coordinates[start_idx:end_idx+1]\n",
    "sel_vals = dif_filtered_D6_mCthresh_995[start_idx:end_idx+1]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16, 4))\n",
    "\n",
    "# Create bar positions centered at coordinates\n",
    "bar_width = 1\n",
    "ax.bar(sel_coords, sel_vals, width=bar_width, color=\"turquoise\", alpha=0.85, edgecolor=\"none\", align='center')\n",
    "ax.axhline(0, color=\"black\", linewidth=0.8, linestyle=\"--\")\n",
    "\n",
    "ax.set_xlabel(\"Genomic Coordinate (HG38)\")\n",
    "ax.set_ylabel(\"ΔmC (CRISPRoff − Unedited)\")\n",
    "ax.set_title(f\"Zoom Differenced D6 CROff - Unedired Filtered : CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]})\")\n",
    "\n",
    "# Add CpG index labels on top of each bar\n",
    "for idx, (coord, value) in enumerate(zip(sel_coords, sel_vals), start=start_idx):\n",
    "    ax.text(coord, value + 0.01 if value > 0 else value - 0.01, \n",
    "            str(idx), ha='center', va='bottom' if value > 0 else 'top',\n",
    "            fontsize=6, rotation=0) \n",
    "\n",
    "# Set x-axis ticks every 100 units\n",
    "ax.xaxis.set_major_locator(ticker.MultipleLocator(2))\n",
    "ax.ticklabel_format(style='plain', axis='x')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.ticker as ticker\n",
    "\n",
    "# Zoom plot: CpG units 69..82 (inclusive)\n",
    "start_idx, end_idx = 69, 73\n",
    "sel_coords = CG_coordinates[start_idx:end_idx+1]\n",
    "sel_vals = dif_filtered_D6_mCthresh_995[start_idx:end_idx+1]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16, 4))\n",
    "\n",
    "# Create bar positions centered at coordinates\n",
    "bar_width = 1\n",
    "ax.bar(sel_coords, sel_vals, width=bar_width, color=\"turquoise\", alpha=0.85, edgecolor=\"none\", align='center')\n",
    "ax.axhline(0, color=\"black\", linewidth=0.8, linestyle=\"--\")\n",
    "\n",
    "ax.set_xlabel(\"Genomic Coordinate (HG38)\")\n",
    "ax.set_ylabel(\"ΔmC (CRISPRoff − Unedited)\")\n",
    "ax.set_title(f\"Zoom Differenced D6 CROff - Unedired Filtered : CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]})\")\n",
    "\n",
    "# Add CpG index labels on top of each bar\n",
    "for idx, (coord, value) in enumerate(zip(sel_coords, sel_vals), start=start_idx):\n",
    "    ax.text(coord, value + 0.01 if value > 0 else value - 0.01, \n",
    "            str(idx), ha='center', va='bottom' if value > 0 else 'top',\n",
    "            fontsize=6, rotation=0) \n",
    "\n",
    "# Set x-axis ticks every 100 units\n",
    "ax.xaxis.set_major_locator(ticker.MultipleLocator(2))\n",
    "ax.ticklabel_format(style='plain', axis='x')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.ticker as ticker\n",
    "\n",
    "# Zoom plot: CpG units 30..54 (inclusive)\n",
    "start_idx, end_idx = 30, 54\n",
    "sel_coords = CG_coordinates[start_idx:end_idx+1]\n",
    "sel_vals = dif_filtered_D6_mCthresh_995[start_idx:end_idx+1]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16, 4))\n",
    "\n",
    "# Create bar positions centered at coordinates\n",
    "bar_width = 1\n",
    "ax.bar(sel_coords, sel_vals, width=bar_width, color=\"turquoise\", alpha=0.85, edgecolor=\"none\", align='center')\n",
    "ax.axhline(0, color=\"black\", linewidth=0.8, linestyle=\"--\")\n",
    "\n",
    "ax.set_xlabel(\"Genomic Coordinate (HG38)\")\n",
    "ax.set_ylabel(\"ΔmC (CRISPRoff − Unedited)\")\n",
    "ax.set_title(f\"Zoom Differenced D6 CROff - Unedired Filtered : CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]}), length {sel_coords[-1]-sel_coords[0]+1} bp\")\n",
    "print(f\" CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]})\")\n",
    "print(f\" length {sel_coords[-1]-sel_coords[0]+1} bp\")\n",
    "\n",
    "# Add CpG index labels on top of each bar\n",
    "for idx, (coord, value) in enumerate(zip(sel_coords, sel_vals), start=start_idx):\n",
    "    ax.text(coord, value + 0.01 if value > 0 else value - 0.01, \n",
    "            str(idx), ha='center', va='bottom' if value > 0 else 'top',\n",
    "            fontsize=6, rotation=0) \n",
    "\n",
    "# Set x-axis ticks every 100 units\n",
    "ax.xaxis.set_major_locator(ticker.MultipleLocator(10))\n",
    "ax.ticklabel_format(style='plain', axis='x')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.ticker as ticker\n",
    "\n",
    "# Zoom plot: CpG units 30..54 (inclusive)\n",
    "start_idx, end_idx = 30, 60\n",
    "sel_coords = CG_coordinates[start_idx:end_idx+1]\n",
    "sel_vals = dif_filtered_D6_mCthresh_995[start_idx:end_idx+1]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16, 4))\n",
    "\n",
    "# Create bar positions centered at coordinates\n",
    "bar_width = 1\n",
    "ax.bar(sel_coords, sel_vals, width=bar_width, color=\"turquoise\", alpha=0.85, edgecolor=\"none\", align='center')\n",
    "ax.axhline(0, color=\"black\", linewidth=0.8, linestyle=\"--\")\n",
    "\n",
    "ax.set_xlabel(\"Genomic Coordinate (HG38)\")\n",
    "ax.set_ylabel(\"ΔmC (CRISPRoff − Unedited)\")\n",
    "ax.set_title(f\"Zoom Differenced D6 CROff - Unedired Filtered : CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]}), length {sel_coords[-1]-sel_coords[0]+1} bp\")\n",
    "print(f\" CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]})\")\n",
    "print(f\" length {sel_coords[-1]-sel_coords[0]+1} bp\")\n",
    "\n",
    "# Add CpG index labels on top of each bar\n",
    "for idx, (coord, value) in enumerate(zip(sel_coords, sel_vals), start=start_idx):\n",
    "    ax.text(coord, value + 0.01 if value > 0 else value - 0.01, \n",
    "            str(idx), ha='center', va='bottom' if value > 0 else 'top',\n",
    "            fontsize=6, rotation=0) \n",
    "\n",
    "# Set x-axis ticks every 100 units\n",
    "ax.xaxis.set_major_locator(ticker.MultipleLocator(10))\n",
    "ax.ticklabel_format(style='plain', axis='x')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.ticker as ticker\n",
    "\n",
    "# Zoom plot: CpG units 0..60 (inclusive)\n",
    "start_idx, end_idx = 0, 137\n",
    "sel_coords = CG_coordinates[start_idx:end_idx+1]\n",
    "sel_vals = dif_filtered_D6_mCthresh_995[start_idx:end_idx+1]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16, 4))\n",
    "\n",
    "# Create bar positions centered at coordinates\n",
    "bar_width = 1\n",
    "ax.bar(sel_coords, sel_vals, width=bar_width, color=\"turquoise\", alpha=0.85, edgecolor=\"none\", align='center')\n",
    "ax.axhline(0, color=\"black\", linewidth=0.8, linestyle=\"--\")\n",
    "\n",
    "ax.set_xlabel(\"Genomic Coordinate (HG38)\")\n",
    "ax.set_ylabel(\"ΔmC (CRISPRoff − Unedited)\")\n",
    "ax.set_title(f\"Zoom Differenced D6 CROff - Unedired Filtered : CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]}), length {sel_coords[-1]-sel_coords[0]+1} bp\")\n",
    "print(f\" CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]})\")\n",
    "print(f\" length {sel_coords[-1]-sel_coords[0]+1} bp\")\n",
    "\n",
    "# Add CpG index labels on top of each bar\n",
    "for idx, (coord, value) in enumerate(zip(sel_coords, sel_vals), start=start_idx):\n",
    "    ax.text(coord, value + 0.01 if value > 0 else value - 0.01, \n",
    "            str(idx), ha='center', va='bottom' if value > 0 else 'top',\n",
    "            fontsize=6, rotation=0) \n",
    "\n",
    "# Set x-axis ticks every 100 units\n",
    "# ax.xaxis.set_major_locator(ticker.MultipleLocator(10))\n",
    "ax.ticklabel_format(style='plain', axis='x')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.ticker as ticker\n",
    "\n",
    "# Zoom plot: CpG units 0..60 (inclusive)\n",
    "start_idx, end_idx = 0, 137\n",
    "sel_coords = CG_coordinates[start_idx:end_idx+1]\n",
    "sel_vals = dif_filtered_D6_mCthresh_995[start_idx:end_idx+1]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16, 4))\n",
    "\n",
    "# Create bar positions centered at coordinates\n",
    "bar_width = 1\n",
    "ax.bar(sel_coords, sel_vals, width=bar_width, color=\"turquoise\", alpha=0.85, edgecolor=\"none\", align='center')\n",
    "ax.axhline(0, color=\"black\", linewidth=0.8, linestyle=\"--\")\n",
    "\n",
    "ax.set_xlabel(\"Genomic Coordinate (HG38)\")\n",
    "ax.set_ylabel(\"ΔmC (CRISPRoff − Unedited)\")\n",
    "ax.set_title(f\"Zoom Differenced D6 CROff - Unedired Filtered : CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]}), length {sel_coords[-1]-sel_coords[0]+1} bp\")\n",
    "print(f\" CpG units {start_idx}–{end_idx} ({REGION_CHR}:{sel_coords[0]}-{sel_coords[-1]})\")\n",
    "print(f\" length {sel_coords[-1]-sel_coords[0]+1} bp\")\n",
    "\n",
    "# # Add CpG index labels on top of each bar\n",
    "# for idx, (coord, value) in enumerate(zip(sel_coords, sel_vals), start=start_idx):\n",
    "#     ax.text(coord, value + 0.01 if value > 0 else value - 0.01, \n",
    "#             str(idx), ha='center', va='bottom' if value > 0 else 'top',\n",
    "#             fontsize=6, rotation=0) \n",
    "\n",
    "# Set x-axis ticks every 100 units\n",
    "# ax.xaxis.set_major_locator(ticker.MultipleLocator(10))\n",
    "ax.ticklabel_format(style='plain', axis='x')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "60",
   "metadata": {},
   "source": [
    "# Create UCSC-compatible BED from annotated ΔmC table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61",
   "metadata": {},
   "outputs": [],
   "source": [
    "CG_coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Create UCSC-compatible BED from annotated ΔmC table\n",
    "# df_bed = pd.DataFrame({\n",
    "#     \"chrom\": \"chr1\",\n",
    "#     \"start\": df_dmc[\"Genomic_Position_hg38\"],\n",
    "#     \"end\": df_dmc[\"Genomic_Position_hg38\"] + 1,\n",
    "#     \"name\": \"CpG_\" + df_dmc[\"CpG Index\"].astype(str),\n",
    "#     \"score\": (df_dmc[\"ΔmC\"] * 1000).round(0).astype(int),  # UCSC expects int scores from 0–1000\n",
    "#     \"strand\": \"+\"\n",
    "# })\n",
    "\n",
    "# # Clip scores to UCSC-compliant range\n",
    "# df_bed[\"score\"] = df_bed[\"score\"].clip(lower=0, upper=1000)\n",
    "\n",
    "# # Save as plain BED file with no header\n",
    "# df_bed.to_csv(\"cd55_dmc_cpgs_INTscores.hg38.bed\", sep=\"\\t\", index=False, header=False)\n",
    "# df_bed\n",
    "# ✅ UCSC Track Line (Required for Upload)\n",
    "\n",
    "# When uploading this BED file to the UCSC Genome Browser, you'll need to add a \"track\" line manually in the custom tracks section:\n",
    "\n",
    "# track name=\"CD55 ΔmC CpGs\" description=\"CRISPRoff vs Unedited T cells\" visibility=2 itemRgb=\"On\" useScore=1\n",
    "\n",
    "\n",
    "# Then paste or upload your cd55_dmc_cpgs.hg38.bed file content below that line.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63",
   "metadata": {},
   "outputs": [],
   "source": [
    "CpG_indexes = np.arange(1, len(CG_coordinates) + 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"CpG_\" + CpG_indexes.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65",
   "metadata": {},
   "outputs": [],
   "source": [
    " \"CpG_\" + np.arange(1, len(CG_coordinates) + 1).astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.array(CG_coordinates) + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67",
   "metadata": {},
   "outputs": [],
   "source": [
    "dif_mc = dif_filtered_D6_mCthresh_995\n",
    "\n",
    "# Create UCSC-compatible BED from annotated ΔmC table\n",
    "df_bed = pd.DataFrame({\n",
    "    \"chrom\": \"chr1\",\n",
    "    \"start\": CG_coordinates,\n",
    "    \"end\": np.array(CG_coordinates) + 1,\n",
    "    \"name\": \"CpG_\" + np.arange(1, len(CG_coordinates) + 1).astype(str),\n",
    "    \"score\": (dif_mc * 1000).round(0).astype(int),  # UCSC expects int scores from 0–1000\n",
    "    \"strand\": \"+\"\n",
    "})\n",
    "\n",
    "# # Clip scores to UCSC-compliant range\n",
    "# df_bed[\"score\"] = df_bed[\"score\"] + 100 # .clip(lower=0, upper=1000)\n",
    "\n",
    "# # Save as plain BED file with no header\n",
    "# df_bed.to_csv(\"D6_CRof_unedit_Tcells_dif_cd55_cpgs_INTscores.hg38.bed\", sep=\"\\t\", index=False, header=False)\n",
    "df_bed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_bed['score'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69",
   "metadata": {},
   "outputs": [],
   "source": [
    "dif_mc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\"\"\"\n",
    "Scale dif_mc to integer scores in 0..1000 suitable for UCSC BED 'score'.\n",
    "Steps:\n",
    "    1. compute min (mn) and max (mx) of dif_mc\n",
    "    2. if mx == mn -> return constant mid score (500) array\n",
    "    3. otherwise: (dif_mc - mn) / (mx - mn) * 1000\n",
    "    4. round, clip to [0,1000], replace NaN with 0, cast to int\n",
    " \"\"\"\n",
    "\n",
    "# scale dif_mc to range 0..1000\n",
    "mn = np.nanmin(dif_mc)\n",
    "mx = np.nanmax(dif_mc)\n",
    "\n",
    "if np.isclose(mx, mn):\n",
    "    dif_mc_scaled = np.full_like(dif_mc, 500.0)  # constant -> mid score\n",
    "else:\n",
    "    dif_mc_scaled = (dif_mc - mn) / (mx - mn) * 1000.0\n",
    "\n",
    "# round, clip to [0,1000], replace NaN with 0 and convert to int\n",
    "dif_mc_scaled_round = np.round(dif_mc_scaled)\n",
    "dif_mc_scaled_clip = np.clip(dif_mc_scaled_round, 0, 1000)\n",
    "dif_mc_scaled_clip_nan0 = np.where(np.isnan(dif_mc_scaled_clip), 0, dif_mc_scaled_clip).astype(int)\n",
    "\n",
    "# assign back or use new variable\n",
    "dif_mc_scaled_int = dif_mc_scaled_clip_nan0\n",
    "\n",
    "print(\"orig_min, orig_max:\", mn, mx)\n",
    "print(\"scaled_min, scaled_max:\", dif_mc_scaled_int.min(), dif_mc_scaled_int.max())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd_scaling = pd.DataFrame()\n",
    "pd_scaling['dif_mc_scaled_int'] = dif_mc_scaled_int\n",
    "\n",
    "print(\"\\nScaled integer scores (0..1000) for UCSC BED 'score' field:\")\n",
    "print(pd_scaling['dif_mc_scaled_int'].describe())\n",
    "\n",
    "print(\"\\nScaled integer scores with NaN replaced by 0:\")\n",
    "pd_scaling['dif_mc_scaled_clip_nan0'] = dif_mc_scaled_clip_nan0\n",
    "print(pd_scaling['dif_mc_scaled_int'].describe())\n",
    "\n",
    "print(\"\\nScaled integer scores clipped to [0,1000] but with NaN still as NaN:\")\n",
    "pd_scaling['dif_mc_scaled_clip'] = dif_mc_scaled_clip\n",
    "pd_scaling['dif_mc_scaled_clip'].describe()\n",
    "\n",
    "\n",
    "print(\"\\nScaled integer scores:\")\n",
    "pd_scaling['dif_mc_scaled'] = dif_mc_scaled_clip\n",
    "pd_scaling['dif_mc_scaled'].describe()\n",
    "\n",
    "\n",
    "# [dif_mc_scaled_int]]).describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = np.arange(len(dif_mc))\n",
    "\n",
    "fig, axes = plt.subplots(2, 1, figsize=(14, 6), sharex=True)\n",
    "\n",
    "axes[0].bar(x, dif_mc, color=\"turquoise\", alpha=0.8)\n",
    "axes[0].set_ylabel(\"ΔmC (original)\")\n",
    "axes[0].set_title(\"Original dif_mc\")\n",
    "\n",
    "axes[1].bar(x, dif_mc_scaled_int, color=\"steelblue\", alpha=0.8)\n",
    "axes[1].set_ylabel(\"ΔmC (scaled 0–1000)\")\n",
    "axes[1].set_xlabel(\"CpG index\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name = \"UCSC_D6_CRof_unedit_Tcells_dif_cd55_cpgs_INTscores.hg38.bed\"\n",
    "\n",
    "def UCSC_scaled_shades_gray( dif_mc_scaled_int,CG_coordinates, file_name):\n",
    "\n",
    "    # Create UCSC-compatible BED from annotated ΔmC table\n",
    "    df_bed = pd.DataFrame({\n",
    "        \"chrom\": \"chr1\",\n",
    "        \"start\": CG_coordinates,\n",
    "        \"end\": np.array(CG_coordinates) + 1,\n",
    "        \"name\": \"CpG_\" + np.arange(1, len(CG_coordinates) + 1).astype(str),\n",
    "        \"score\": dif_mc_scaled_int,  # UCSC expects int scores from 0–1000\n",
    "        \"strand\": \"+\"\n",
    "    })\n",
    "\n",
    "    # # Clip scores to UCSC-compliant range\n",
    "    # df_bed[\"score\"] = df_bed[\"score\"] + 100 # .clip(lower=0, upper=1000)\n",
    "\n",
    "    # Save as plain BED file with no header\n",
    "    df_bed.to_csv(file_name, sep=\"\\t\", index=False, header=False)\n",
    "    df_bed\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74",
   "metadata": {},
   "source": [
    "# ✅ UCSC Track Line (Required for Upload)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75",
   "metadata": {},
   "source": [
    "\n",
    "When uploading this BED file to the UCSC Genome Browser, you'll need to add a \"track\" line manually in the custom tracks section:\n",
    "\n",
    "track name=\"CD55 ΔmC CpGs\" description=\"CRISPRoff vs Unedited T cells\" visibility=2 itemRgb=\"On\" useScore=1\n",
    "\n",
    "\n",
    "Then paste or upload your cd55_dmc_cpgs.hg38.bed file content below that line.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76",
   "metadata": {},
   "source": [
    "# Add RGB colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_track_difs_color_bed(dif_mc: np.ndarray, \n",
    "                              CG_coordinates: np.ndarray,\n",
    "                                    experiment_condition: str, \n",
    "                                    out_path: Path) -> None:\n",
    "    \"\"\"\n",
    "    BED9 with itemRgb, colors reflect dif_mc values:\n",
    "\n",
    "    - sign:   red = positive, blue = negative\n",
    "    - brightness: ~ |effect_size| (robustly scaled by 95th percentile)\n",
    "    \"\"\"\n",
    "\n",
    "    es = dif_mc.astype(float)\n",
    "\n",
    "    # robust scaling by |effect_size|\n",
    "    abs_es = np.abs(np.where(np.isinf(es), np.nan, es))\n",
    "    if np.isfinite(abs_es).any():\n",
    "        p95 = np.nanpercentile(abs_es[np.isfinite(abs_es)], 95)\n",
    "    else:\n",
    "        p95 = 1.0\n",
    "    if p95 == 0 or np.isnan(p95):\n",
    "        p95 = 1.0\n",
    "\n",
    "    norm = np.clip(abs_es / p95, 0, 1)\n",
    "\n",
    "    # intensity range (bright)\n",
    "    min_intensity = 150\n",
    "    intensity = (min_intensity + norm * (255 - min_intensity)).round().astype(int)\n",
    "\n",
    "    r = np.zeros(len(es), dtype=int)\n",
    "    g = np.zeros(len(es), dtype=int)\n",
    "    b = np.zeros(len(es), dtype=int)\n",
    "\n",
    "    pos_mask = es > 0\n",
    "    neg_mask = es < 0\n",
    "    zero_mask = ~(pos_mask | neg_mask)\n",
    "\n",
    "    # positive -> shades of red\n",
    "    r[pos_mask] = intensity[pos_mask]\n",
    "    g[pos_mask] = 0\n",
    "    b[pos_mask] = 0\n",
    "\n",
    "    # negative -> shades of blue\n",
    "    r[neg_mask] = 0\n",
    "    g[neg_mask] = 0\n",
    "    b[neg_mask] = intensity[neg_mask]\n",
    "\n",
    "    # zeros/NaNs -> neutral gray\n",
    "    r[zero_mask] = 180\n",
    "    g[zero_mask] = 180\n",
    "    b[zero_mask] = 180\n",
    "\n",
    "    itemRgb = [f\"{ri},{gi},{bi}\" for ri, gi, bi in zip(r, g, b)]\n",
    "\n",
    "    start_coords = CG_coordinates.astype(int)\n",
    "    end_coords = CG_coordinates.astype(int) + 2\n",
    "\n",
    "    bed_score = (norm * 1000).round().astype(int)\n",
    "\n",
    "    bed9 = pd.DataFrame(\n",
    "        {\n",
    "            \"chrom\": \"chr1\",\n",
    "            \"start\": CG_coordinates,\n",
    "            \"end\": np.array(CG_coordinates) + 2,\n",
    "            \"name\": \"CpG_\" + np.arange(1, len(CG_coordinates) + 1).astype(str),\n",
    "            \"score\": bed_score,\n",
    "            \"strand\": \"+\",\n",
    "            \"thickStart\": start_coords,\n",
    "            \"thickEnd\": end_coords,\n",
    "            \"itemRgb\": itemRgb,\n",
    "        }\n",
    "    )\n",
    "    track_line = (\n",
    "        'track name=\"dif_mCG_Red_Blue_'+experiment_condition+'\" '\n",
    "        'description=\"CG colored RB by dif_mCG (red=positive, blue=negative; brightness~|dif_mCG|) '+experiment_condition+'\" '\n",
    "        'visibility=2 itemRgb=\"On\"\\n'\n",
    "    )\n",
    "\n",
    "    with open(out_path, \"w\") as f:\n",
    "        f.write(track_line)\n",
    "    bed9.to_csv(out_path, sep=\"\\t\", header=False, index=False, mode=\"a\")\n",
    "    print(f\"Written effect_size-colored BED to: {out_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78",
   "metadata": {},
   "outputs": [],
   "source": [
    "dif_mc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('date_today id', date_today)\n",
    "# merge date_today by removing hyphens\n",
    "date_today_merged = date_today.replace(\"-\", \"\")\n",
    "date_today_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80",
   "metadata": {},
   "outputs": [],
   "source": [
    "dif_mc = dif_filtered_D6_mCthresh_995\n",
    "\n",
    "experiment_condition = \"D6_CROff_vs_Unedited_Tcells_0995thresh_filtered\"\n",
    "out_folder = \"/home/michalula/code/epiCausality/epiCode/analyze_ont_data/compare_conditions/final_visuals/UCSC/\"\n",
    "out_path = out_folder + date_today_merged +\"_UCSC_\" + experiment_condition + \"_dif_mCG_together_color_RedBlue.bed\"\n",
    "\n",
    "print(\"out_path:\", out_path)\n",
    "\n",
    "\n",
    "make_track_difs_color_bed(dif_mc, \n",
    "                              np.array(CG_coordinates),\n",
    "                                    experiment_condition, \n",
    "                                    out_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81",
   "metadata": {},
   "source": [
    "# Use this track line when uploading to UCSC Genome Browser:\n",
    "NOOO: track name=\"dif D6_CROff_Tcells_0995thr_filter\" description=\"T cells dif mCG D6 [CRISPRoff - Unedited] colored RB (red=positive, blue=negative; brightness~|dif_mCG|) FILTERED data 0.995 ONT conf threshold Mode > 10 Date:20260216\" visibility=2 itemRgb=\"On\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82",
   "metadata": {},
   "source": [
    "NOPE: track name=\"dif D6_CROff_Tcells_0995thr_filter\" description=\"T cells dif mCG D6 CRISPRoff (red=positive, blue=negative; brightness~|dif_mCG|) FILTERED data 0.995 thresh\" visibility=2 itemRgb=\"On\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83",
   "metadata": {},
   "source": [
    "track name=\"dif D6_CROff_Tcells_0995thr_filter\" description=\"T cells D6 CRISPRoff (red=positive, blue=neg; bright~|dif_mCG|) FILTERED 0.995thr\" visibility=2 itemRgb=\"On\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84",
   "metadata": {},
   "outputs": [],
   "source": [
    "track name=\"dif D6_CROff_Tcells_0995thr_filter\" description=\"T cells D6 CRISPR\" visibility=2 itemRgb=\"On\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85",
   "metadata": {},
   "outputs": [],
   "source": [
    "track name=\"T D6 CRoff Dif \" description=\"T cells D6 CRISPR\" visibility=2 itemRgb=\"On\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dimelo_v2_modkit_parsing",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
